{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ortholog Finding\n",
    "#### This module was originally created for cas9 but could extended to any protein.\n",
    "\n",
    "## Step 1: Profile the Protein\n",
    "    1. Mean, max, min length of already known orthologs\n",
    "## Step 2: HMM Seach\n",
    "    1. Align orthologs\n",
    "    2. build hmm profile\n",
    "    3. hmm search against a database\n",
    "## Step 3: Profile Results\n",
    "    1. Read hmm results\n",
    "    2. cluster results by percent identity\n",
    "    3. remove proteins without required domains\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Env"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<p>\n",
    "conda create --name casenv <br/>\n",
    "conda install -n casenv libgcc hmmer mafft biopython <br/>\n",
    "source activate casenv\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import re\n",
    "from sys import path as spath\n",
    "spath.append(\"scripts/\")\n",
    "\n",
    "from Bio.SeqIO import write, parse\n",
    "from collections import Counter\n",
    "from CRISPRtools import *\n",
    "from easyFunctions import *\n",
    "from HMMParser import *\n",
    "from IPython.display import display, HTML\n",
    "from matplotlib import pyplot as plt\n",
    "from os import chdir, listdir,system\n",
    "from pandas import Series\n",
    "from pickle import load #Dump has been wrapped aournd a function from easyFunctions\n",
    "\n",
    "chdir(\"data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Orthologs Fasta File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = \"Cas9\"\n",
    "geneProfile = \"proteins/DiverseCas9s.faa\" #File containing the seed proteins for the hmmsearch\n",
    "geneFile = \"proteins/%s.faa\" % gene\n",
    "alnName  = \"alignments/%s.aln\" % gene\n",
    "hmmName  = \"hmm/%s.hmm\" % gene\n",
    "hmmResultsDir = \"hmm/results\"\n",
    "baseDbsDir = \"/mnt/research/germs/shane/databases/assemblies/\"\n",
    "refDatabases = [\"NCBI/refseq/bacteria\",\"NCBI/refseq/archaea\",\"NCBI/genbank/bacteria\",\"NCBI/genbank/archaea\",\"PATRIC2/fastas\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Profile the gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count      84.000000\n",
      "mean     1243.547619\n",
      "std       160.535955\n",
      "min       984.000000\n",
      "25%      1099.750000\n",
      "50%      1207.500000\n",
      "75%      1376.500000\n",
      "max      1688.000000\n",
      "dtype: float64\n",
      "Finding proteins that match the protein profile and are between 673 and 1998\n"
     ]
    }
   ],
   "source": [
    "dists = []\n",
    "for rec in parse(geneProfile,\"fasta\"):dists.append(len(rec.seq))\n",
    "dists = Series(dists)\n",
    "minSeqLen = int(dists.min() - (dists.mean() * .25))\n",
    "maxSeqLen = int(dists.max() + (dists.mean() * .25))\n",
    "print (dists.describe());\n",
    "print (\"Finding proteins that match the protein profile and are between %i and %i\" % (minSeqLen, maxSeqLen))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEICAYAAABS0fM3AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAHYdJREFUeJzt3XuYHHWd7/H3B8IlYSCIkREjMuAiByTqksZFXXWGi4JAENcLbFSy4kbd42V3gwrihXNcFS+Ro497DkbNA4IyIHgFXQHXkXVF2QTBAYEFNUoCJCISmBCBge/5o6qTYpzurpnp6upOfV7PM0/q1r/6dHWlv12/rq5SRGBmZtW1XdkBzMysXC4EZmYV50JgZlZxLgRmZhXnQmBmVnEuBGZmFedCsA2TdK6kD5Sdw/KRtETSj9vcZtv2AUnPkDQmaft0fETSm9vRdtre9ySd0q72LD8Xgh4laY2kzZIelHS/pJ9IequkLa9pRLw1Ij5cZs5WJO0o6SxJt0valD6vlZIGZtjugZL+XdJGSXdIOrHJsm1/A86Rb0BSSJo1gzbatg+kbR3ZbJmI+F1E9EXEY9PNnFnfWZIunND+MRFx/kzbtqlzIehtx0fErsA+wNnAe4EvFblCJdq531wKLAL+FpgLPBdYDRwx3QbTN9dvAZcDewBLgQslPWvGabtPR/aBmRQs6wER4b8e/APWAEdOmPZ84HHg4HT8POBf0uFbgOMyy84C7gUOSccPA34C3A/cCAxmlh0BPgL8J7AZ+AtgCfBr4EHgN8DizPJvStf3R+D7wD4NnsORaXt7N3mef5e29WC6vrdk5s0jebO/H7gP+A+SDzcHA2OAMsteCXy4wTqWAD9uMG8uyRvr3cA64F+A7bOPAz6VPtffAMdkHrsvcE2a/WrgX4EL03m/AyLNOQa8oFV7bdgHGm2vC9LHbE6zvAcYSPOdmma9JjNtVma/+BhwHbCRpPjukc4bBNZOlhc4GngEeDRd342Z9t6cDm8HvB/4LbAB+DIwN51Xz3FKmu1e4Myy/0/28p+PCLYhEXEdsBZ48SSzLwJOzoy/HLg3Iq6XNB+4guRNbg/gNOAySU/JLP8Gkk/WuwK/Bz5L8ia1K/BC4AYASa8E3ge8CngKyZvNRQ0iHwlcFxF3NnlaG4DjgN1IisI5kg5J5y1Ln+9TgP50vQFoknZEUiCm6nxgnKT4/SXwMiDbL/5XwG0kb7KfAL4kqb7+r5K8ST4ZOItkG9a9JP1390i6W67N0V5LLfaBSbdXRLyB5A31+DTLJzKPeSlwIMn+Mpk3khT+p5Fsp8/myPhvwEeBi9P1PXeSxZakf0PAfkAf8LkJy/w1cADJ0eMHJR3Yat02OReCbc9dJG/mE30VWCRpTjr+t+k0gNcD342I70bE4xFxFbAKeEXm8edFxM0RMU7yH/5x4GBJsyPi7oi4OV3uLcDHIuKWdNmPAs+TtM8kmZ5M8km7oYi4IiJ+FYkfkXyyr7/JPQrsRXLE8WhE/EckHxlvJSkg75a0g6SXkbyhzZlsHY1I6geOAf4xIjZFxAbgHOCkzGK/jYgvRNJvfn6ap1/SM4BDgQ9GxCMR8WPg2zlWO2l7U8lN432g0fZq5qz0uW9uMP+CiLgpIjYBHwBeW/8yeYYWA5+OiF9HxBhwBnDShC6q/xURmyPiRpKj2MkKiuXgQrDtmU9y2P8EEXEHSRfL8WkxWMTWQrAP8Jr0C8f7Jd1P8mlrr0wTd2ba2gS8DngrcLekKyT9j0xbn8m0cx/Jp/H5k2T9w4R1/BlJx0j6qaT70vZeQfJpGeCTwB3AlZJ+Len0NN+jwCuBY4F7SD4JX0LyaXgq9gF2SJ9j/fl8Htgzs8w99YGIeCgd7CP5hHxfZhpktmETjdqbikn3ARpsrxZaZc7O/y3J9prXYNmpeFraXrbtWTyxKN6TGX6IqW8nS7kQbEMkHUryJtDoDJh699AJwC/T4gDJf+YLImL3zN8uEXF25rFP+OQYEd+PiKNI3shvBb6QaestE9qaHRE/mSTP1cDzJT29wfPZCbiMpM+8PyJ2B75L2vUTEQ9GxLKI2A84HvhnSUek834RES+NiCdHxMtJuheua7BdGrkTeBiYl3kuu0XEs3M89m5gj8wRGMDemeFCLvvbbB9otr2a5GmVM/ucnkFy1HEvsInMEVh6lJDtamzV7l0khTjb9jiwvsXjbBpcCLYBknaTdBwwTPJl5GiDRYdJ+rjfxtajAYALSY4UXi5pe0k7Sxps8gbdL2mRpF1I3ijHgPophecCZ0h6drrsXEmvmaydiLgauAr4hqSFkmZJ2jU9BfJNwI7ATiTfSYxLOibNX89xnKS/SPvQH0gzPJbOe076POZIOo2kYJ3XeCuidPktfxFxN0lX1PJ0G28n6ZmSXtqknfpz+y1J99pZ6SmyLyB58637PUn32n6t2sojzz7QbHuRvMFOJ8vrJR2UFrz/DVyadmv9N7CzpGMl7UDyxe9OmcetBwaanIF2EfBPkvaV1MfW7xTGp5HRWnAh6G3fkfQgySfXM4FPk3yhOqn0je1aki93L85Mv5PkKOF9JG9QdwLvpvH+sR1Jd8tdJF0QLwX+IW3rG8DHgWFJDwA3kfSzN/Jqkk/5F5OceXITUAOujogHgXeSdOv8keR7jWw/+/4kRxVj6fP6vxExks57A8mn8g0kXyYeFREPN8nxQpKzZrb8pf3RbyQpSL9MM1xKi+6sjMUkZwP9geSL+ItJCme92+cjwH+m3U6H5WxzoqnsA82218eA96dZTpvC+i8gKbD3ADuTvF5ExEaSfeKLJGdbbeKJXXNfS//9g6TrJ2l3Zdr2NSRnT/0JeMcUctkUqPV3RWbWDpIuBm6NiA+VncUsy0cEZgWRdGjalbSdpKNJjrq+WXYus4n8a0Gz4jwV+DrJabJrgbdFxM/LjWT259w1ZGZWce4aMjOruJ7oGpo3b14MDAy0vd1Nmzaxyy67tL3dIvRSVuitvM5aDGctTt68q1evvjcintJywZlerKgTfwsXLowi/PCHPyyk3SL0UtaI3srrrMVw1uLkzQusCl90zszMWnEhMDOrOBcCM7OKcyEwM6s4FwIzs4pzITAzqzgXAjOzinMhMDOrOBcCM7OK64lLTPSqgdOvaFtbyxaMsyRne2vOPrZt67Xm2vka55HdD/w6W7v4iMDMrOJcCMzMKs6FwMys4lwIzMwqzoXAzKziXAjMzCrOhcDMrOJcCMzMKq6wQiBppaQNkm6aMP0dkm6TdLOkTxS1fjMzy6fII4LzgKOzEyQNAScAz4mIZwOfKnD9ZmaWQ2GFICKuAe6bMPltwNkR8XC6zIai1m9mZvkoudF9QY1LA8DlEXFwOn4D8C2SI4U/AadFxH81eOxSYClAf3//wuHh4bbnGxsbo6+vr+3t1o2u29i2tvpnw/rN+ZZdMH9u29Y7XUVv23aaSdZ2vsZ5ZPeDbnidm6nKPlCGvHmHhoZWR0St1XKdvujcLOBJwGHAocAlkvaLSapRRKwAVgDUarUYHBxse5iRkRGKaLcu70Xi8li2YJzlo/lerjWLB9u23ukqetu200yytvM1ziO7H3TD69xMVfaBMrQ7b6fPGloLfD0S1wGPA/M6nMHMzDI6XQi+CRwOIOlZwI7AvR3OYGZmGYV1DUm6CBgE5klaC3wIWAmsTE8pfQQ4ZbJuITMz65zCCkFEnNxg1uuLWqeZmU2df1lsZlZxLgRmZhXnQmBmVnEuBGZmFedCYGZWcS4EZmYV1+lLTFgHDHT4sgd1a84+tpT1mtnM+IjAzKziXAjMzCrOhcDMrOJcCMzMKs6FwMys4lwIzMwqzoXAzKziXAjMzCqusEIgaaWkDelNaCbOO01SSPJtKs3MSlbkEcF5wNETJ0raGzgK+F2B6zYzs5wKKwQRcQ1w3ySzzgHeA/gWlWZmXUBF3jJY0gBweUQcnI4vAo6IiHdJWgPUImLSm9dLWgosBejv7184PDzc9nxjY2P09fW1vd260XUb29ZW/2xYv7ltzRViwfy5W4aL3rbtNJOs7XyN88juB9nt3Y2qsg+UIW/eoaGh1RFRa7Vcxy46J2kOcCbwsjzLR8QKYAVArVaLwcHBtmcaGRmhiHbrlrTx4m/LFoyzfLS7rxG4ZvHgluGit207zSRrO1/jPLL7QXZ7d6Oq7ANlaHfeTp419ExgX+DG9Gjg6cD1kp7awQxmZjZBxz5iRsQosGd9vFXXkJmZdUaRp49eBFwLHCBpraRTi1qXmZlNX2FHBBFxcov5A0Wt28zM8vMvi83MKs6FwMys4lwIzMwqzoXAzKziXAjMzCrOhcDMrOJcCMzMKs6FwMys4lwIzMwqzoXAzKziXAjMzCrOhcDMrOJcCMzMKs6FwMys4lwIzMwqrsgb06yUtEHSTZlpn5R0q6RfSPqGpN2LWr+ZmeVT5BHBecDRE6ZdBRwcEc8B/hs4o8D1m5lZDoUVgoi4BrhvwrQrI2I8Hf0pyQ3szcysRGV+R/Am4Hslrt/MzABFRHGNSwPA5RFx8ITpZwI14FXRIICkpcBSgP7+/oXDw8Ntzzc2NkZfX1/b260bXbexbW31z4b1m9vWXCEWzJ+7ZbjobdtOM8naztc4j+x+kN3e3agq+0AZ8uYdGhpaHRG1VssVdvP6RiSdAhwHHNGoCABExApgBUCtVovBwcG2ZxkZGaGIduuWnH5F29patmCc5aMdf7mmZM3iwS3DRW/bdppJ1na+xnlk94Ps9u5GVdkHytDuvB19Z5F0NPBe4KUR8VAn121mZpMr8vTRi4BrgQMkrZV0KvA5YFfgKkk3SDq3qPWbmVk+hR0RRMTJk0z+UlHrMzOz6fEvi83MKs6FwMys4lwIzMwqzoXAzKziXAjMzCrOhcDMrOJcCMzMKs6FwMys4nIVAkkHt17KzMx6Ud4jgnMlXSfpH3xXMTOzbUuuQhARfw0sBvYGVkn6qqSjCk1mZmYdkfs7goi4HXg/6dVDgc+m9x9+VVHhzMyseHm/I3iOpHOAW4DDgeMj4sB0+JwC85mZWcHyXn30c8AXgPdFxJb7ZEXEXZLeX0gyMzPriLyF4BXA5oh4DEDSdsDOEfFQRFxQWDozMytc3u8IrgZmZ8bnpNPMzKzH5S0EO0fEWH0kHZ7T7AGSVkraIOmmzLQ9JF0l6fb03ydNL7aZmbVL3kKwSdIh9RFJC4HNTZYHOA84esK004EfRMT+wA/ScTMzK1He7wj+EfiapLvS8b2A1zV7QERcI2lgwuQTgMF0+HxghOR0VDMzK4kiIt+C0g7AAYCAWyPi0RyPGQAuj4iD0/H7I2L3zPw/RsSk3UOSlgJLAfr7+xcODw/nyjkVY2Nj9PX1tb3dutF1G9vWVv9sWN/qGKxkC+bP3TJc9LZtp5lkbedrnEc37AfZ17mZquwDZcibd2hoaHVE1FotN5Wb1x8KDKSP+UtJRMSXp/D4KYmIFcAKgFqtFoODg21fx8jICEW0W7fk9Cva1tayBeMsH53Ky9V5axYPbhkuetu200yytvM1zqMb9oPs69xMVfaBMrQ7b649StIFwDOBG4DH0skBTLUQrJe0V0TcLWkvYMMUH29mZm2W96NFDTgo8vYjNfZt4BTg7PTfb82wPTMzm6G8Zw3dBDx1Kg1Lugi4FjhA0lpJp5IUgKMk3Q4clY6bmVmJ8h4RzAN+Kek64OH6xIhY1OgBEXFyg1lH5I9nZmZFy1sIzioyhJmZlSdXIYiIH0naB9g/Iq6WNAfYvthoZmbWCXkvQ/33wKXA59NJ84FvFhXKzMw6J++Xxf8TeBHwAGy5Sc2eRYUyM7POyVsIHo6IR+ojkmaR/I7AzMx6XN5C8CNJ7wNmp/cq/hrwneJimZlZp+Q9a+h04FRgFHgL8F3gi0WFaqeBJpcAWLZgvOOXCDAz6zZ5zxp6nORWlV8oNo6ZmXVa3msN/YZJvhOIiP3ansjMzDpqKtcaqtsZeA2wR/vjmJlZp+X6sjgi/pD5WxcR/wc4vOBsZmbWAXm7hg7JjG5HcoSwayGJzMyso/J2DS3PDI8Da4DXtj2NmZl1XN6zhoaKDmJmZuXI2zX0z83mR8Sn2xPHzMw6Le8vi2vA20guNjcfeCtwEMn3BFP+rkDSP0m6WdJNki6StPNU2zAzs/aYyo1pDomIBwEknQV8LSLePNUVSpoPvJPk1pebJV0CnAScN9W2zMxs5vIeETwDeCQz/ggwMIP1ziK5btEsYA5w1wzaMjOzGVCe+9FLOpPkLKFvkPzC+ETgkoj46LRWKr0L+AiwGbgyIhZPssxSYClAf3//wuHh4emsitF1GxvO658N6zdPq9mO64WsC+bP3TI8NjZGX19fiWnym0nWZvtXEbphP8i+zs1UZR8oQ968Q0NDqyOi1mq5XIUAtvyW4MXp6DUR8fNcD/zzdp4EXAa8Drif5Eqml0bEhY0eU6vVYtWqVdNZXcuLzi0fzds7Vq5eyLrm7GO3DI+MjDA4OFhemCmYSdZm+1cRumE/yL7OzVRlHyhD3rySchWCvF1DkHThPBARnwHWStp3Co/NOhL4TUT8PiIeBb4OvHCabZmZ2QzlvVXlh4D3Amekk3YAGn6Cb+F3wGGS5kgScARwyzTbMjOzGcp7RHAisAjYBBARdzHNS0xExM9I7n98Pcn9DbYDVkynLTMzm7m8nY2PRERICgBJu8xkpRHxIeBDM2nDzMzaI+8RwSWSPg/sLunvgavxTWrMzLYJea819Kn0XsUPAAcAH4yIqwpNZmZmHdGyEEjaHvh+RBwJ+M3fzGwb07JrKCIeAx6SlO9XJGZm1lPyfln8J2BU0lWkZw4BRMQ7C0llZmYdk7cQXJH+mZnZNqZpIZD0jIj4XUSc36lAZmbWWa2OCL4JHAIg6bKI+JviI5lN3Uyu+bNswThLOnzNILNu0urLYmWG9ysyiJmZlaNVIYgGw2Zmto1o1TX0XEkPkBwZzE6HSccjInYrNJ2ZmRWuaSGIiO07FcTMzMoxlfsRmJnZNsiFwMys4lwIzMwqzoXAzKziSikEknaXdKmkWyXdIukFZeQwM7P81xpqt88A/xYRr5a0IzCnpBxmZpXX8UIgaTfgJcASgIh4BHik0znMzCyhiM7+YFjS80huVv9L4LnAauBdEbFpwnJLgaUA/f39C4eHh6e1vtF1GxvO658N6zdPq9mO64WsC+ZvvWXF2NgYfX19HVt3s9e5lV7YtnXdkDX7OjfT6X1gJnopK+TPOzQ0tDoiaq2WK6MQ1ICfAi+KiJ9J+gzwQER8oNFjarVarFq1alrra3YxsmULxlk+Wlbv2NT0QtY1Zx+7ZXhkZITBwcGOrXumF53r9m1b1w1Zs69zM53eB2ail7JC/rySchWCMr4sXgusjYifpeOXkl7h1MzMOq/jhSAi7gHulHRAOukIkm4iMzMrQVnHmO8AvpKeMfRr4O9KymFmVnmlFIKIuAFo2W9lZmbF8y+LzcwqzoXAzKziXAjMzCrOhcDMrOJcCMzMKs6FwMys4nrjd/XWE7KXeVi2YJwlM7jsg5l1jo8IzMwqzoXAzKziXAjMzCrOhcDMrOJcCMzMKs6FwMys4lwIzMwqzoXAzKziSisEkraX9HNJl5eVwczMyj0ieBdwS4nrNzMzSioEkp4OHAt8sYz1m5nZVoqIzq9UuhT4GLArcFpEHDfJMkuBpQD9/f0Lh4eHp7Wu0XUbG87rnw3rN0+r2Y7rpazQW3mddWoWzJ+ba7mxsTH6+voKTtMevZQV8ucdGhpaHREtbwvc8YvOSToO2BARqyUNNlouIlYAKwBqtVoMDjZctKlmFz5btmCc5aO9cd29XsoKvZXXWadmzeLBXMuNjIww3f+3ndZLWaH9ecvoGnoRsEjSGmAYOFzShSXkMDMzSigEEXFGRDw9IgaAk4B/j4jXdzqHmZkl/DsCM7OKK7WzMSJGgJEyM5iZVZ2PCMzMKs6FwMys4lwIzMwqzoXAzKziXAjMzCrOhcDMrOJcCMzMKs6FwMys4lwIzMwqzoXAzKziXAjMzCrOhcDMrOJcCMzMKs6FwMys4lwIzMwqruOFQNLekn4o6RZJN0t6V6czmJnZVmXcmGYcWBYR10vaFVgt6aqI+GUJWczMKq+MexbfHRHXp8MPArcA8zudw8zMEoqI8lYuDQDXAAdHxAMT5i0FlgL09/cvHB4entY6RtdtbDivfzas3zytZjuul7JCb+V11mIUkXXB/LntbTA1NjZGX19fw/nN3keKNtlzbpW3bmhoaHVE1FotV1ohkNQH/Aj4SER8vdmytVotVq1aNa31DJx+RcN5yxaMs3y01Ns259ZLWaG38jprMYrIuubsY9vaXt3IyAiDg4MN5zd7HynaZM+5Vd46SbkKQSlnDUnaAbgM+EqrImBmZsUq46whAV8CbomIT3d6/WZm9kRlHBG8CHgDcLikG9K/V5SQw8zMKOH00Yj4MaBOr9fMzCbnXxabmVWcC4GZWcW5EJiZVZwLgZlZxbkQmJlVnAuBmVnFuRCYmVWcC4GZWcW5EJiZVZwLgZlZxbkQmJlVnAuBmVnFuRCYmVWcC4GZWcW5EJiZVZwLgZlZxZV1z+KjJd0m6Q5Jp5eRwczMEmXcs3h74F+BY4CDgJMlHdTpHGZmlijjiOD5wB0R8euIeAQYBk4oIYeZmQGKiM6uUHo1cHREvDkdfwPwVxHx9gnLLQWWpqMHALcVEGcecG8B7Rahl7JCb+V11mI4a3Hy5t0nIp7SaqGO37yeyW9c/2fVKCJWACsKDSKtiohaketol17KCr2V11mL4azFaXfeMrqG1gJ7Z8afDtxVQg4zM6OcQvBfwP6S9pW0I3AS8O0ScpiZGSV0DUXEuKS3A98HtgdWRsTNnc6RKrTrqc16KSv0Vl5nLYazFqeteTv+ZbGZmXUX/7LYzKziXAjMzCpumyoEklZK2iDppsy0PSRdJen29N8nZeadkV7m4jZJL89MXyhpNJ33WUmTnfJaVN7XSLpZ0uOSahOWLy1vg6yflHSrpF9I+oak3bs464fTnDdIulLS07o1a2beaZJC0rxuyNoor6SzJK1Lt+0Nkl7RDXkbbVtJ70jz3CzpE92aVdLFmW26RtINhWWNiG3mD3gJcAhwU2baJ4DT0+HTgY+nwwcBNwI7AfsCvwK2T+ddB7yA5DcP3wOO6WDeA0l+QDcC1DLTS83bIOvLgFnp8Me7Zds2yLpbZvidwLndmjWdvjfJCRW/BeZ1Q9Ym2/Ys4LRJlu26bQsMAVcDO6Xje3Zr1gnzlwMfLCrrNnVEEBHXAPdNmHwCcH46fD7wysz04Yh4OCJ+A9wBPF/SXiRvGtdGsmW/nHlM4Xkj4paImOxX1KXmbZD1yogYT0d/SvKbkG7N+kBmdBe2/oix67KmzgHewxN/bNmV+2wT3bht3wacHREPp8ts6OKsAKSf6l8LXFRU1m2qEDTQHxF3A6T/7plOnw/cmVlubTptfjo8cXrZuj3vm0g+gUCXZpX0EUl3AouBD6aTuy6rpEXAuoi4ccKsrsua8fa0622ltna/dmPeZwEvlvQzST+SdGg6vRuz1r0YWB8Rt6fjbc9ahULQSKNLXeS6BEYJujavpDOBceAr9UmTLFZ61og4MyL2JslZv7ZVV2WVNAc4k62F6gmzJ5lW+nYF/h/wTOB5wN0k3RjQnXlnAU8CDgPeDVySfuLuxqx1J7P1aAAKyFqFQrA+PWQi/bd+KNjoUhdr2drFkZ1etq7MK+kU4DhgcXo4Cl2aNeOrwN+kw92W9Zkk/b43SlqTrvd6SU/twqwARMT6iHgsIh4HvkByhWHozrxrga9H4jrgcZILuHVjViTNAl4FXJyZ3PasVSgE3wZOSYdPAb6VmX6SpJ0k7QvsD1yXdh89KOmw9JPCGzOPKVPX5ZV0NPBeYFFEPNTlWffPjC4Cbu3GrBExGhF7RsRARAyQ/Oc+JCLu6basdfUPWqkTgfqZL92Y95vA4WnuZwE7klzFsxuzAhwJ3BoR2S6f9mdt97ffZf6RHD7dDTxK8h/oVODJwA+A29N/98gsfybJN+63kfl2HaiR7My/Aj5H+gvsDuU9MR1+GFgPfL8b8jbIegdJX+UN6d+5XZz1snS9vwC+A8zv1qwT5q8hPWuo7KxNtu0FwGi6bb8N7NUNeRtk3RG4MF339cDh3Zo1nX4e8NZJlm9rVl9iwsys4qrQNWRmZk24EJiZVZwLgZlZxbkQmJlVnAuBmVnFuRCYmVWcC4GZWcX9f+BHtChQcsKiAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "dists.plot(kind='hist',grid=True,title=\"Diverse Cas9 Length Distribution\",figsize=(6,4))\n",
    "plt.savefig(\"images/DiverseCas9Lengths.png\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make HMM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# hmmbuild :: profile HMM construction from multiple sequence alignments\n",
      "# HMMER 3.2.1 (June 2018); http://hmmer.org/\n",
      "# Copyright (C) 2018 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# input alignment file:             alignments/Cas9.aln\n",
      "# output HMM file:                  hmm/Cas9.hmm\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "# idx name                  nseq  alen  mlen eff_nseq re/pos description\n",
      "#---- -------------------- ----- ----- ----- -------- ------ -----------\n",
      "1     Cas9                    84  4084  1039     7.01  0.590 \n",
      "\n",
      "# CPU time: 2.13u 0.00s 00:00:02.13 Elapsed: 00:00:02.16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "outputhat23=16\n",
      "treein = 0\n",
      "compacttree = 0\n",
      "stacksize: -1 kb\n",
      "All-to-all alignment.\n",
      "tbfast-pair (aa) Version 7.407\n",
      "alg=L, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "15 thread(s)\n",
      "\n",
      "outputhat23=16\n",
      "Loading 'hat3.seed' ... \n",
      "done.\n",
      "Writing hat3 for iterative refinement\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "tbutree = 1, compacttree = 0\n",
      "Constructing a UPGMA tree ... \n",
      "   80 / 84\n",
      "done.\n",
      "\n",
      "Progressive alignment ... \n",
      "STEP    36 /83 (thread    7) \n",
      "Reallocating (by thread 12) ..done. *alloclen = 4507\n",
      "STEP    55 /83 (thread    6) \n",
      "Reallocating (by thread 14) ..done. *alloclen = 5536\n",
      "STEP    79 /83 (thread   12) \n",
      "Reallocating (by thread 2) ..done. *alloclen = 6590\n",
      "STEP    83 /83 (thread    4) \n",
      "done.\n",
      "tbfast (aa) Version 7.407\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "15 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 8\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 16\n",
      "sueff_global = 0.100000\n",
      "nadd = 16\n",
      "Loading 'hat3' ... done.\n",
      "\n",
      "   80 / 84\n",
      "Segment   1/  1    1-4441\n",
      "016-0164-1 (thread    3) worse      \n",
      "Reached 16\n",
      "done\n",
      "dvtditr (aa) Version 7.407\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "8 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " L-INS-i (Probably most accurate, very slow)\n",
      " Iterative refinement method (<16) with LOCAL pairwise alignment information\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash -s $geneProfile $alnName $hmmName\n",
    "mafft --thread 15 --maxiterate 1000 --retree 100 --localpair --reorder --treeout $1 > $2 #Assuming there are <200 sequences in the profile\n",
    "hmmbuild $3 $2 2>&1 >/dev/null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 453958 HMM files\n",
      "Loaded 453958 CRISPR files\n",
      "\n",
      "Dumping pickles/HMMFiles_Cas9.p\n"
     ]
    }
   ],
   "source": [
    "from os import listdir\n",
    "hmmFiles = listdir(hmmResultsDir)\n",
    "print(\"Loaded %i HMM files\" % (len(hmmFiles)))\n",
    "results = set()\n",
    "for fname in hmmFiles: results.add(\"%s/%s\" % (hmmResultsDir,fname))\n",
    "hmmFiles = results\n",
    "assembliesWCrisprs = load(open(\"pickles/CRISPRs.p\",\"rb\")) #From ProcessCRISPRs.ipynb\n",
    "print(\"Loaded %i CRISPR files\" % (len(assembliesWCrisprs)))\n",
    "dump(hmmFiles, \"pickles/HMMFiles_%s.p\" % gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 442759 HMM files\n",
      "Loaded CRISPR files\n"
     ]
    }
   ],
   "source": [
    "hmmFiles = load(open(\"pickles/HMMFiles_%s.p\" % gene, \"rb\"))\n",
    "print(\"Loaded %i HMM files\" % (len(hmmFiles)))\n",
    "assembliesWCrisprs = load(open(\"pickles/CRISPRs.p\",\"rb\")) #From ProcessCRISPRs.ipynb\n",
    "print(\"Loaded %i CRISPR files\" % (len(assembliesWCrisprs)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build hpc scripts for all assemblies with a CRISPR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 453958 CRISPR related assemblies\n",
      "There are 61 files for 6005 assemblies\n"
     ]
    }
   ],
   "source": [
    "system(\"rm ../scripts/hpc/hmmSearch/*\")\n",
    "print(\"There are %i CRISPR related assemblies\" % (len(assembliesWCrisprs)))\n",
    "fh = open(\"../scripts/hpc/hmmSearch/HMMSearch_0.sb\",\"w\")\n",
    "header = \"\"\n",
    "for line in open(\"../scripts/hpc/header.sb\"): header+=line\n",
    "fh.write(header)\n",
    "cmdCount, fileCounter = 0,1\n",
    "getOrfsCMD = \"python /mnt/research/germs/shane/transActRNA/scripts/GetOrfs.py %s sequences/orfs/%s.orfs 673 >/dev/null\"\n",
    "hmmSearchCMD = \"hmmsearch hmm/%s.hmm sequences/orfs/%s.orfs >%s/%s.hmmout\"\n",
    "\n",
    "for asmName, operon in assembliesWCrisprs.items():\n",
    "    asmFilePath = operon.assembly\n",
    "    if \"%s/%s.hmmout\" % (hmmResultsDir,asmName) in hmmFiles: continue\n",
    "    cmdCount += 1\n",
    "    fh.write(\"if [ ! -f %s/%s.hmmout ]; then\\n\" % (hmmResultsDir,asmName)) #Check to see if the hmm profile already exists\n",
    "    fh.write(\"\\t\"+getOrfsCMD % (asmFilePath,asmName) + \"\\n\")\n",
    "    fh.write(\"\\t\"+hmmSearchCMD % (gene,asmName,hmmResultsDir,asmName)+\"\\n\")\n",
    "    fh.write(\"\\thmmsize=$(wc -l <\\\"%s/%s.hmmout\\\") \\n\" % (hmmResultsDir,asmName))\n",
    "    fh.write(\"\\tif [ $hmmsize -le 40 ]; then \\n\")\n",
    "    fh.write(\"\\t\\trm sequences/orfs/%s.orfs \\n\" % (asmName))\n",
    "    fh.write(\"\\tfi\\n\")\n",
    "    fh.write(\"fi\\n\\n\")\n",
    "    if cmdCount % 50 == 0:\n",
    "        fh.close()\n",
    "        fh = open(\"../scripts/hpc/hmmSearch/HMMSearch_%i.sb\" % (fileCounter),\"w\")\n",
    "        fh.write(header)\n",
    "        fileCounter += 1\n",
    "fh.close()\n",
    "print(\"There are %i files for %i assemblies\" % (fileCounter,cmdCount))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Do HMM search on each assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "bash ../scripts/hpc/LaunchHMMSearch.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read the HMM Results and create fasta files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 24156860\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "sbatch ../scripts/hpc/ReadHMM_Results.sb # This script does the same thing as the cell below\n",
    "#cat ../scripts/hpc/ReadHMM_Results.sb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from CRISPRtools import CasOperons\n",
    "crisprFiles = load(open(\"pickles/CRISPRs.p\",\"rb\")) \n",
    "casOperons = CasOperons(gene)\n",
    "casOperons.hasCas9(hmmResultsDir+\"/\",crisprFiles)\n",
    "dump(casOperons, \"pickles/%s_Operons.p\" % gene)\n",
    "\n",
    "# Sort through all the results and make a file with all of the Cas proteins of interest. \n",
    "# If there are duplicates only keep the duplicates that come from a unique chromosome\n",
    "casOperons = load(open(\"pickles/%s_Operons.p\" % gene,\"rb\"))\n",
    "allCasAsmFile = \"assemblies/All_%s_Representative_Assemblies.fasta\" % (gene)\n",
    "allCasAAsFile = \"proteins/All_%s-Like.faa\" % (gene)\n",
    "casOperons.uniqueNukeSeqs(allCasAsmFile,allCasAAsFile,gene)\n",
    "dump(casOperons, \"pickles/casOperonDataStructureW%s.p\" % gene) \n",
    "system(\"sbatch ../scripts/hpc/DomainSearch.sb\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Profile domains of the Cas protein of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of unique proteins: 5269\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAbcAAAEWCAYAAADl19mgAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3XuYXFWZ7/HvjwRDQoCAkRaSQCOCCgQVmogjzjQCgugYnBFPMGqiaMRBRY2jQTkjzjEHZhSd4SBqBlBuQ4yIwoAol7FBRyASRENAJJIIgUi4kyAGgu/5Y62GTVHVXd1V1dW1+X2ep5+uWnvtvd+1b+++1d6KCMzMzMpks3YHYGZm1mxObmZmVjpObmZmVjpObmZmVjpObmZmVjpObmZmVjodmdwkrZZ0cIvH0SXpWknrJZ3SguHPlnRFs4c72knaIOlloyCOXklrCt9XSOptY0g2AiR9R9KXBug+YsunpD5JHxyJcb0QDSm5jURSqTLOARfGFpoHPABsHRHzKztWWzArN5gDiYjzI+LNzQm1fpJOlPRUXokfkfQLSa8f5rC6JYWksfX2ExETI+LOYY7vRTn+OyQ9npfHsyR1D2d4FXHtGRF9jQ6nUo75FElr8jRfJelrzR7PC1kzt0uNLJ8Dycvtec0ebsU43i3pxrycrZV0uaQDGhzmJElnS1qX/04coG7V7UFxG16oc1NFncmSnpS0ulC2OpdNrqh7cx5G90Cxd+SR2wjZGbg1yvkr9+9GxETgJcDPgYskqbLSUJLWCLkQeDvwbmAb4NXAMuCgdgY1iOOBHmAGsBVwIPCrtkZkpSPpU8C/Af8X6AJ2Ak4HZjY46K8BE4Bu0jL8Xknvb3CYAFtK2qvw/d3Aqir1VgFH9X+RNB0YX9cYIqLuP2A1cHCNbm8DbgYeAX4B7F3R36eB3wCPAt8Ftih0/wywFrgX+CAQwMtJR09PAU8CG4D/Gmx4wGTg0hzHQ8DPgM1qxPxXwC/zMH4J/FUu/07FeJ/XZqAP+GBFWS+wpvA9gGOAO4CHga8Dyt3mAj8v1D0E+G2O5TTgmv7hAycC5xXqdudhj83ftwHOzNPwHuBLwJgaba4c1p55WJNzTP9DWqAfysPZDDgB+AOwDjgH2Cb3e1fud0P+e30u/wBwW27zT4CdK6bJywvT+evAZcB64AZg1xpxHww8AUwbYPncEbgkx74S+FCh2/g8voeBW4F/rJhXq/vnc55GS3Jb1wMrgJ5C3X1ICWo98D3S8velGjFdCnxikJi/D9xPWpE/PoSYn5mWhen5pcL3RtbJmbnfx4DfA4cNtqyR1tlr8vAeIO1EVWtzd479/cDduX3HAPvleB4BTivU3xX4b+DBPNzzgUm527nAX/KysQH4TC4/ILf5kTyOufUscwxh+QTeDNye23s6hXW2or2HkbYlT+UYf13Yhvwf0jq3HrgCmFzob/9CG34N9NaYntvk4R45wHI2A7guD2staRvzotxNpHV+XW7Lb4C9crcHgP0Kw/kc8LNB5uvYivLvkJfLQp0TgC8X6twIfB5YXbGMngD8slD2lVwvgO5a7Y2I5iQ30sq+DngdMAaYk+uOK/S3lLQib0fa8B1TmPF/JG1kJ5AW1soF7EtV4qg1vJOAbwKb5783khNKxTC2I61U7wXGkvYOHgZeXGu8Ff33UV9yuxSYRNqTup9nNxJzycmNlFgeA96ZY/4ksIn6k9sPgW8BWwLb52nz4RpxPzMsYBzwZeDuQkybgI/laTKelKhWAi8DJgIXAefWWpiBI3L9V+VhnAD8YoCNx0OkFW8saaO1uEbcJwPXDLJ8XkPayGwBvCZP74MK/f8sz/dpwC0MnNz+DBxOWp5PAq7P3V5ESvTH5Xn1d6QNV63kdgJpJ+AfgOkUlkXSjsMy4J/ycF8G3AkcWmfMNZMbja2TM0gbuUNyjFOAVw62rAEXkDY8m+V5cMAgG8Fv5npvztP7h3mYU3Lsf5PrvzzHMo50tuFa4N9qbZdI69p60jq9OfBi4DX1LHPUuXzy7Dr7d7nbcaTk9bzkVm0dLmxDfg/sTlrX+oCTc7cppGR+eJ6eh+TvL6ky7MNI6+3YauPOdfYlJcuxefrfRt7pAg4lLYeTSInuVcAOudsDwIzCcD4PPDzIfK0nuXWTdjrG5PHdTtqBrUxuB+dur8p17yadVRs0uTXrtOSHgG9FxA0R8XREnA1szBOz36kRcW9EPAT8F2njA/Au4NsRsSIi/gR8sc5x1hreU8AOpKOFpyLiZ5GnVIW3AndExLkRsSkiLiAdOf3tENpdj5Mj4pGIuAv4aSHOosNJp0AvjIinSKcX/ljPwCV1AW8hLaiPR8Q60l7YrAF6e5ek/j3afUkJqd+9EfH/8jR5ApgNfDUi7oyIDaTTbLMGOGX5YeCkiLgtIjaRTpO8RtLONepfFBFLc93zqT59IG2g1tZqkKRppL31z0bEnyPiZuAM0s4LpOVsYUQ8FBF3A6fWGlb284j4UUQ8TdrhenUu799AnJqXr4tIG/haTgL+hTQdbwTukTQnd9uPtLH654h4MtK1nv/g2Xk31JiLGlknjwbOiogrI+IvEXFPRPy2jmXtKdKGZ8c8D34+SIz/J9e7AngcuCAi1kXEPaSk/lqAiFiZY9kYEfcDXwX+ZoDhzgauiogL8jx6MC8P/epd5gaqeziwIiIuyt1Opc51tsK3I+J3eV1bUhj+e4Af5WXwLxFxJWn5ObzKMF4MPJDjqCoilkXE9Xm9Xk3aQemfhk+RTpm/krTzdVtE9K9rPwYWSNpK0stJO7sTBmnTA/l6/iN5O/PuKnXW8GxCm0M6S1LLucD7ePbs1j2DjB9o3jW3nYH5FQ2aRtor7Fec8X8iHQWQ69xd6Fb8PJBaw/sy6cjhCkl3SlpQo/8dSXvgRX8g7THVYxNpr7Boc9KCUk+clbE80+6cjOudDjvn8a4tTPtvkfaAa1kSEZMiYvuIeFNELCt0qxxv5XT6A2nj3jVAPP9eiOUh0t5grelaz/SBtNe6Q41u/XE+FBHrK2KdUuh+d0W3gVTGtUVO6DsC91TsMNWcVzmxfD0i3kDaM14InCXpVeREULHefI5np+1QYy5qZJ2cRjqiqDbMgZa1z5Dm9dJ89+kHBonxvsLnJ6p8nwggaXtJiyXdI+kx4DzSkVMtteLvV+8yN1DdautsXTeT1Tn8nYEjK+bfAVRfBx4EJg90jVzS7pIulfTHPA3/L3kaRsR/k05Tfh24T9IiSVvnXj9Omhd3ABeTjs4Ha+fkvH2ZFBGTgP+sUe8c0tmio0jztJZzSQlyLgMnwedoVnK7m7SHOanwNyEfDQ1mLTC18H1aRfdqR101RcT6iJgfES8jHYV9SlK1Gw7uJS1ARTtR514B6VRTd0XZLgxtA9RvLYV255s7itPhcZ67t/TSwue7SXvkxQVq64jYcxhxwPOnd+V02omU2O+rUrc/ng9XLAvjI+IXw4yn31XADElTa3S/F9hO0lYVsfbPz+dM49xtONYCUypuwKlcZquKiCci4uuk0997kKbVqopptVVE9O+dDxbznxh4uRjuOnk36TpXtfKay1pE/DEiPhQRO5KO4E/Pe/uNOom0rO0dEVuTjmqK079yOawVfzM9Z7uVl4dayyYMcTtGasO5FfNvy4g4uUrd60indY+o0q3fN0hHPbvlafg5CtMwIk6NiH1Jl4d2J13fJZ81mB0RL83zeTMGPlMxFN8nnUG7MyJqbjdzt1Wko9aL6h34cJLb5pK2KPyNJZ1KOUbS65RsKemtFRuaWpYA75f0KkkTSNcfiu4jXYuoi6S3SXp5XtgeA57Of5V+BOyeb58dK+l/kTY4l9Y5qu/muGfkNu9Oula2uN5YCy4D9pT0d3l6fpznbqhuBv5a0k6StiGdGgQgnz64AjhF0taSNpO0q6SBTtsMxQXAJyXtImkiaY/vu/kUyP2ki/nF+fNN4HhJewJI2kbSkY0GERFXAVcCP5C0b55nW0k6RtIH8mm7XwAn5eVyb9LptfPzIJbkuLbNCfJjwwzlOtLy9NEcw0zSNZmqJH1C6Sci43P9OaRTQL8ibSQek/TZ3H2MpL0k7VdnzDcD7879HcZzT9U1sk6eSVq2D8rL0xRJrxxsWZN0ZGHn42HSBr3aujdUW5FumHhE0hTyhregchtxPnCwpHflaf5iSQOdehyOy4Dpko7I6+yxPHedrXQf0C2p3m3uecDfSjo0z98t8nL0vAQaEY+Stptfz/FMkLS5pLdI+tdcbSvS9nCDpFcCH+nvX9J+eTnZnLQj/WfyfMvz98U5hreQbvJryk+zIuJx4E2kmwgHczTwptxPXYaT3H5EOkzt/zsxIm4kneM/jbRQryQdQg4qIi4nna/+ae7vutxpY/5/JrCH0qH5D+sY5G6kvfwNeVinR5XfL0XEg6S7yeaTDus/A7wtIh6oM+6fAAuAb5Muvv8IOBtYVE//FcN6ADiSdAPBg7kN/1PofiUpmf6GdOG3MgG/j3RDwq2k6X8hA5/CG4qzSKcFriXtPf2ZvJGNdI10IfA/ef7sHxE/IF1jWqx0+uMW0nWaZngnaTp/lzTNbyHdZn9V7n4U6Wj6XuAHwBfytIN0Lbd/D/CK3KYhi4gnSTcRHE268+w9pPmxsUYvTwCnkE4/PUDaCP59pGuYT5POLrwmx/UA6TrhNnXGfFzu/xHSdaZn1o8G18mlpDsZv0aaztfw7NH7QMvafsANkjaQ7lo9LiJW1TPOQXyRdIPMo6SkUrn3fhJwQl4GPx3p+vbhpHX7IdJOwKtposI6+6+kdXYP0jWxWsvB9/L/B1XxG68aw7+bdMfq50g7kXeTknrVbXZEfBX4FOkGpv76H+XZZeLTpFN760k7Pt8t9L51LnuYtLw9SLorEdI1+eW5v5OA2RGxYrD46xURN0bEQKeQ++v9Pi/Tdeu/LX3UULoWcQvprq6aF0jLTlIf6e6qM9odiw1M0g3ANyPi2y0eTy9pmRjo9Je1QT4iW0Pa+P+03fHYKPkRt6R3KD3JYVvSXv9/vZATm41ukv5G0ksLpxn3Jt1VZi8g+ZThJEnjePYa1vVtDsuyUZHcSBef7yfd4fQ0hfPBZqPQK0g/qn2UdOrrnYVbp+2F4/WkbdYDpNPDR+Rb+m0UGHWnJc3MzBo1Wo7czMzMmma0PRi3aSZPnhzd3d3tDqNujz/+OFtuuWW7w2iJsrbN7eosbld9li1b9kBEvKRpA2yT0ia37u5ubrxxSHeOtlVfXx+9vb3tDqMlyto2t6uzuF31kTScB1GMOj4taWZmpePkZmZmpePkZmZmpePkZmZmpePkZmZmpePkZmZmpePkZmZmpdOy5CbpLEnrJN1SpdunJYWkyYWy4yWtlHS7pEML5ftKWp67nSo95yWRZmZmz9PKI7fvAIdVFkqaBhxCepN1f9kewCzSW2API73Bd0zu/A3SC/J2y3/PG6aZmVlRy55QEhHXSuqu0ulrpBeDXlwomwksjoiNwCpJK4EZklYDW0fEdQCSziG9Sv3yVsVtVmbdCy4btM786ZuYW0e9oVh98lubOjyzwYzo47ckvR24JyJ+XXF2cQrPfQ/Smlz2VP5cWV5r+PNIR3l0dXXR19fXnMBHwIYNGzoq3qEoa9s6sV3zpw/+msSu8fXVG4rRMJ06cX7Vo6ztatSIJTdJE4DPA2+u1rlKWQxQXlVELAIWAfT09EQnPUeurM+9g/K2rRPbVc8R2fzpmzhleXM3Datn9zZ1eMPRifOrHmVtV6NG8shtV2AXoP+obSpwk6QZpCOyaYW6U4F7c/nUKuVmZmY1jdhPASJieURsHxHdEdFNSlz7RMQfgUuAWZLGSdqFdOPI0vx24/WS9s93Sb6P516rMzMze55W/hTgAuA64BWS1kg6ulbdiFgBLAFuBX4MHBsRT+fOHwHOAFaSXunum0nMzGxArbxb8qhBundXfF8ILKxS70Zgr6YGZ2ZmpeYnlJiZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWek4uZmZWem0LLlJOkvSOkm3FMq+LOm3kn4j6QeSJhW6HS9ppaTbJR1aKN9X0vLc7VRJalXMZmZWDq08cvsOcFhF2ZXAXhGxN/A74HgASXsAs4A9cz+nSxqT+/kGMA/YLf9VDtPMzOw5WpbcIuJa4KGKsisiYlP+ej0wNX+eCSyOiI0RsQpYCcyQtAOwdURcFxEBnAMc0aqYzcysHNp5ze0DwOX58xTg7kK3NblsSv5cWW5mZlbT2HaMVNLngU3A+f1FVarFAOW1hjuPdAqTrq4u+vr6Ggt0BG3YsKGj4h2KsratE9s1f/qmQet0ja+v3lCMhunUifOrHmVtV6NGPLlJmgO8DTgon2qEdEQ2rVBtKnBvLp9apbyqiFgELALo6emJ3t7e5gXeYn19fXRSvENR1rZ1YrvmLrhs0Drzp2/ilOXN3TSsnt3b1OENRyfOr3qUtV2NGtHTkpIOAz4LvD0i/lTodAkwS9I4SbuQbhxZGhFrgfWS9s93Sb4PuHgkYzYzs87TsiM3SRcAvcBkSWuAL5DujhwHXJnv6L8+Io6JiBWSlgC3kk5XHhsRT+dBfYR05+V40jW6yzEzMxtAy5JbRBxVpfjMAeovBBZWKb8R2KuJoZmZWcn5CSVmZlY6Tm5mZlY6Tm5mZlY6Tm5mZlY6bfkRt1m/7jp+d9Uqq09+a9vGbWat5SM3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrHSc3MzMrnUGTm6Rx9ZRVqXOWpHWSbimUbSfpSkl35P/bFrodL2mlpNslHVoo31fS8tztVEmqr2lmZvZCVc+R23V1llX6DnBYRdkC4OqI2A24On9H0h7ALGDP3M/pksbkfr4BzAN2y3+VwzQzM3uOsbU6SHopMAUYL+m1QP8R09bAhMEGHBHXSuquKJ4J9ObPZwN9wGdz+eKI2AiskrQSmCFpNbB1RFyXYzoHOAK4fPCmmZnZC1XN5AYcCswFpgJfLZSvBz43zPF1RcRagIhYK2n7XD4FuL5Qb00ueyp/riyvStI80lEeXV1d9PX1DTPMkbdhw4aOincoBmrb/OmbRjaYgkandyfOs3qmd9f45s+X0TCdOnF+1aOs7WpUzeQWEWcDZ0v6+4j4fovjqHYdLQYoryoiFgGLAHp6eqK3t7cpwY2Evr4+2hlv94LLWjbs+dOf5pSfP16j60D7V621enZvQ/23e54Nx9w65vP86Zs4ZXlz50uj07oZOnF+1aOs7WpUPUvwpZLeDXQX60fEPw9jfPdJ2iEfte0ArMvla4BphXpTgXtz+dQq5WZmZjXVc0PJxaRrYpuAxwt/w3EJMCd/npOH3V8+S9I4SbuQbhxZmk9hrpe0f75L8n2FfszMzKqq58htakQM+Q5FSReQbh6ZLGkN8AXgZGCJpKOBu4AjASJihaQlwK2kJHpsRDydB/UR0p2X40k3kvhmEjMzG1A9ye0XkqZHxPKhDDgijqrR6aAa9RcCC6uU3wjsNZRxm5nZC1s9ye0AYK6kVcBG0k0eERF7tzQyMzOzYaonub2l5VGYmZk1UT3Jreat92ZmZqNRPcntMp79zdkWwC7A7aRHZZmZmY06gya3iJhe/C5pH+DDLYvIzMysQUN+5U1E3ATs14JYzMzMmmLQIzdJnyp83QzYB7i/ZRGZmZk1qJ5rblsVPm8iXYNr9bMmzczMhq2ea25fBJC0VfoaG1oelVmJtfIh1WaW1PMm7r0k/Qq4BVghaZkkPzHEzMxGrXpuKFkEfCoido6InYH5uczMzGxUqie5bRkRP+3/EhF9wJYti8jMzKxB9dxQcqek/w2cm7+/B1jVupDMzMwaU8+R2weAlwAX5b/JwPtbGZSZmVkjah65SdoC2Coi7gc+XijvAp4YgdjMzMyGZaAjt1OBN1YpPxj4WmvCMTMza9xAye2AiLiosjAizgf+unUhmZmZNWag5KZh9mdmZtZWAyWpdZJmVBZK2g8/W9LMzEaxgX4K8I/AEknfAZblsh7gfcCsFsdlZmY2bDWP3CJiKTCDdHpybv4T8LqIuKGRkUr6pKQVkm6RdIGkLSRtJ+lKSXfk/9sW6h8vaaWk2yUd2si4zcys/Ab8EXdErAO+0MwRSppC+mnBHhHxhKQlpCPBPYCrI+JkSQuABcBnJe2Ru+8J7AhcJWn3iHi6mXGZmVl5tOvGkLHAeEljgQnAvcBM4Ozc/WzgiPx5JrA4IjZGxCpgJemI0szMrCpFxMiPVDoOWEj6MfgVETFb0iMRMalQ5+GI2FbSacD1EXFeLj8TuDwiLqwy3HnAPICurq59Fy9ePBLNaYoNGzYwceLEto1/+T2PtmzYXePhvlH4s//pU7ZpqP/hzrNWTutmaMX8anRaN0O717FWaXa7DjzwwGUR0dO0AbZJPW/i3isibmnWCPO1tJnALsAjwPckvWegXqqUVc3IEbGI/MaCnp6e6O3tbSzYEdTX10c7453bwneMzZ++iVOW1/MY05G1enZvQ/0Pd561clo3QyvmV6PTuhnavY61Slnb1ah6Tkt+U9JSSf8gadLg1Qd1MLAqIu6PiKdIz6v8K+A+STsA5P/rcv01wLRC/1NJpzHNzMyqGjS5RcQBwGxSgrlR0n9KOqSBcd4F7C9pgiQBBwG3AZcAc3KdOcDF+fMlwCxJ4yTtAuwGLG1g/GZmVnJ1nXuIiDsknQDcSHrm5GtzYvpctUd0DTKsGyRdCNwEbAJ+RTqVOJH0u7qjSQnwyFx/Rb6j8tZc/1jfKWlmZgOp55rb3qRX3LwVuBL424i4SdKOwHWk04pDEhFf4Pk/MdhIOoqrVn8h6QYUMzOzQdVz5HYa8B+ko7Rn7qGKiHvz0ZyZmdmoUk9yOxx4ov9UoKTNgC0i4k8Rce7AvZqZmY28eu6WvAoYX/g+IZeZmZmNSvUkty0iYkP/l/x5QutCMjMza0w9ye1xSfv0f5G0L+nJImZmZqNSPdfcPkF6ikj/D6d3AP5X60IyMzNrzKDJLSJ+KemVwCtIj8L6bX6yiJmZ2ahU7wPk9gO6c/3XSiIizmlZVGZmZg2o50fc5wK7AjcD/U8GCcDJzczMRqV6jtx6SC8WHfl345iZmQ1DPXdL3gK8tNWBmJmZNUs9R26TgVslLSU9/xGAiHh7y6IyMzNrQD3J7cRWB2FmZtZM9fwU4BpJOwO7RcRVkiYAY1ofmpmZ2fAMes1N0oeAC4Fv5aIpwA9bGZSZmVkj6rmh5FjgDcBjkF5cCmzfyqDMzMwaUU9y2xgRT/Z/kTSW9Ds3MzOzUame5HaNpM8B4yUdAnwP+K/WhmVmZjZ89SS3BcD9wHLgw8CPAL+B28zMRq167pb8C/Af+c/MzGzUq+fZkquoco0tIl7WkojMzMwaVO+zJfttARwJbNfISCVNAs4A9iIlzg8AtwPfJb19YDXwroh4ONc/Hjia9ODmj0fETxoZv5mZldug19wi4sHC3z0R8W/Amxoc778DP46IVwKvBm4jXdu7OiJ2A67O35G0BzAL2BM4DDhdkn9EbmZmNdVzWnKfwtfNSEdyWw13hJK2Bv4amAuQf2bwpKSZQG+udjbQB3wWmAksjoiNwCpJK4EZwHXDjcHMzMpNg73JRtJPC183kU4ZfiUibh/WCKXXAIuAW0lHbcuA44B7ImJSod7DEbGtpNOA6yPivFx+JnB5RFxYZdjzgHkAXV1d+y5evHg4IbbFhg0bmDhxYtvGv/yeR1s27K7xcN8TLRv8sE2fsk1D/Q93nrVyWjdDK+ZXo9O6Gdq9jrVKs9t14IEHLouInsFrjm713C15YAvGuQ/wsYi4QdK/k09B1qBqYVWrGBGLSImTnp6e6O3tbTDUkdPX10c745274LKWDXv+9E2csrzel76PnNWzexvqf7jzrJXTuhlaMb8andbN0O51rFXK2q5G1XNa8lMDdY+Irw5xnGuANRFxQ/5+ISm53Sdph4hYK2kHYF2h/rRC/1OBe4c4TjMzewGp50fcPcBHSA9MngIcA+xBuu425GtvEfFH4G5Jr8hFB5FOUV4CzMllc4CL8+dLgFmSxknaBdgNWDrU8ZqZ2QtHvS8r3Sci1gNIOhH4XkR8sIHxfgw4X9KLgDuB95MS7RJJRwN3kX5yQESskLSElAA3AcdGxNMNjNvMzEqunuS2E/Bk4fuTpN+iDVtE3Mxzfz/X76Aa9RcCCxsZp5mZvXDUk9zOBZZK+gHpRo53AOe0NCozM7MG1HO35EJJlwNvzEXvj4hftTYsMzOz4avnhhKACcBjEfHvwJp8Y4eZmdmoNGhyk/QF0pNCjs9FmwPntTIoMzOzRtRz5PYO4O3A4wARcS8NPH7LzMys1epJbk9GekZXAEjasrUhmZmZNaae5LZE0reASZI+BFyFX1xqZmajWD13S35F0iHAY8ArgH+KiCtbHpmZmdkwDZjc8nvTfhIRBwNOaGZm1hEGPC2ZH3P1J0ntf1+FmZlZnep5QsmfgeWSriTfMQkQER9vWVRmZmYNqCe5XZb/zMzMOkLN5CZpp4i4KyLOHsmAzMzMGjXQNbcf9n+Q9P0RiMXMzKwpBkpuKnx+WasDMTMza5aBklvU+GxmZjaqDXRDyaslPUY6ghufP5O/R0Rs3fLozMzMhqFmcouIMSMZiJmZWbPU+z43MzOzjuHkZmZmpdO25CZpjKRfSbo0f99O0pWS7sj/ty3UPV7SSkm3Szq0XTGbmVlnaOeR23HAbYXvC4CrI2I34Or8HUl7ALOAPYHDgNPzA53NzMyqaktykzQVeCtwRqF4JtD/NJSzgSMK5YsjYmNErAJWAjNGKlYzM+s8Si/ZHuGRShcCJwFbAZ+OiLdJeiQiJhXqPBwR20o6Dbg+Is7L5WcCl0fEhVWGOw+YB9DV1bXv4sWLR6I5TbFhwwZ5Q4y7AAAMvUlEQVQmTpzYtvEvv+fRlg27azzc90TLBj9s06c09rKL4c6zVk7rZmjF/Gp0WjdDu9exVml2uw488MBlEdHTtAG2ST0PTm4qSW8D1kXEMkm99fRSpaxqRo6IRcAigJ6enujtrWfwo0NfXx/tjHfugtY9G3v+9E2csnzEF7VBrZ7d21D/w51nrZzWzdCK+dXotG6Gdq9jrVLWdjWqHVucNwBvl3Q4sAWwtaTzgPsk7RARayXtAKzL9dcA0wr9TwXuHdGIrZS6G0wy86dvGvWJyuyFasSvuUXE8RExNSK6STeK/HdEvAe4BJiTq80BLs6fLwFmSRonaRdgN2DpCIdtZmYdZDSdKzoZWCLpaOAu4EiAiFghaQlwK7AJODa/IdzMzKyqtia3iOgD+vLnB4GDatRbCCwcscDMzKyj+QklZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOqPpd25mVlKNPg1muFaf/Na2jNfaz0duZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOiOe3CRNk/RTSbdJWiHpuFy+naQrJd2R/29b6Od4SSsl3S7p0JGO2czMOks7jtw2AfMj4lXA/sCxkvYAFgBXR8RuwNX5O7nbLGBP4DDgdElj2hC3mZl1iBFPbhGxNiJuyp/XA7cBU4CZwNm52tnAEfnzTGBxRGyMiFXASmDGyEZtZmadRBHRvpFL3cC1wF7AXRExqdDt4YjYVtJpwPURcV4uPxO4PCIurDK8ecA8gK6urn0XL17c+kY0yYYNG5g4cWLbxr/8nkdbNuyu8XDfEy0bfNu4XaPf9CnbPPO53etYqzS7XQceeOCyiOhp2gDbpG1v4pY0Efg+8ImIeExSzapVyqpm5IhYBCwC6Onpid7e3iZEOjL6+vpoZ7xzW/im5PnTN3HK8vK99N3tGv1Wz+595nO717FWKWu7GtWWuyUlbU5KbOdHxEW5+D5JO+TuOwDrcvkaYFqh96nAvSMVq5mZdZ523C0p4Ezgtoj4aqHTJcCc/HkOcHGhfJakcZJ2AXYDlo5UvGZm1nnace7hDcB7geWSbs5lnwNOBpZIOhq4CzgSICJWSFoC3Eq60/LYiHh65MM2M7NOMeLJLSJ+TvXraAAH1ehnIbCwZUGZmVmp+AklZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOk5uZmZWOu14E7fV0L3gsnaHYGZWCk5uVbQjycyfvgnPDjOz5vBpSTMzKx0nNzMzKx0nNzMzK52OSW6SDpN0u6SVkha0Ox4zMxu9OuIOBkljgK8DhwBrgF9KuiQibm1vZGY2mhVvDps/fRNzS3hHcmW7Vp/81jZGM3p0ypHbDGBlRNwZEU8Ci4GZbY7JzMxGKUVEu2MYlKR3AodFxAfz9/cCr4uIj1bUmwfMy19fAdw+ooE2ZjLwQLuDaJGyts3t6ixuV312joiXNHF4bdERpyUBVSl7XlaOiEXAotaH03ySboyInnbH0QplbZvb1VncrheWTjktuQaYVvg+Fbi3TbGYmdko1ynJ7ZfAbpJ2kfQiYBZwSZtjMjOzUaojTktGxCZJHwV+AowBzoqIFW0Oq9k68nRqncraNrers7hdLyAdcUOJmZnZUHTKaUkzM7O6ObmZmVnpOLmNEEmflLRC0i2SLpC0haTtJF0p6Y78f9tC/ePzo8Zul3RoO2OvJOksSesk3VIoG3JbJO0raXnudqqkaj/5GDE12vVlSb+V9BtJP5A0qdCtY9tV6PZpSSFpcqGso9sl6WM59hWS/rVQ3hHtyjFVWxZfI+l6STdLulHSjEK3jmnbiIkI/7X4D5gCrALG5+9LgLnAvwILctkC4F/y5z2AXwPjgF2A3wNj2t2OQnv+GtgHuKVQNuS2AEuB15N+x3g58JZR2K43A2Pz538pS7ty+TTSTVp/ACaXoV3AgcBVwLj8fftOa9cAbbuiPzbgcKCvE9s2Un8+chs5Y4HxksYCE0i/05sJnJ27nw0ckT/PBBZHxMaIWAWsJD2CbFSIiGuBhyqKh9QWSTsAW0fEdZHWwnMK/bRFtXZFxBURsSl/vZ70G0vo8HZlXwM+w3MfiNDp7foIcHJEbMx11uXyjmkX1GxbAFvnz9vw7G99O6ptI8XJbQRExD3AV4C7gLXAoxFxBdAVEWtznbXA9rmXKcDdhUGsyWWj2VDbMiV/riwfzT5A2vuFDm+XpLcD90TErys6dXS7gN2BN0q6QdI1kvbL5Z3eLoBPAF+WdDdpe3J8Li9D25rOyW0E5OtPM0mnDHYEtpT0noF6qVLWqb/ZqNWWjmqjpM8Dm4Dz+4uqVOuIdkmaAHwe+KdqnauUdUS7srHAtsD+wD8CS/J1pk5vF6Sj0k9GxDTgk8CZubwMbWs6J7eRcTCwKiLuj4ingIuAvwLuy6cOyP/7T6F04uPGhtqWNTx7iq9YPupImgO8DZidT+9AZ7drV9KO1q8lrSbFeJOkl9LZ7YIU50WRLAX+QnqwcKe3C2AOadsB8D2evVRRhrY1nZPbyLgL2F/ShLwXeRBwG+kRYnNynTnAxfnzJcAsSeMk7QLsRrowPJoNqS351OV6SfvnafK+Qj+jhqTDgM8Cb4+IPxU6dWy7ImJ5RGwfEd0R0U3aCO4TEX+kg9uV/RB4E4Ck3YEXkZ6Y3+ntgpSY/iZ/fhNwR/5chrY1X7vvaHmh/AFfBH4L3AKcS7qz6cXA1aSF9Gpgu0L9z5PuerqdUXaHE3AB6drhU6QN49HDaQvQk6fH74HTyE/MGWXtWkm6nnFz/vtmGdpV0X01+W7JTm8XKZmdl+O8CXhTp7VrgLYdACwj3Rl5A7BvJ7ZtpP78+C0zMysdn5Y0M7PScXIzM7PScXIzM7PScXIzM7PScXIzM7PScXKzjiPpHflJ9q8cRr/HSHpfE2P5pKQ/S9qmjrpnSNqjSvlcSafVW94skiZJ+ofC915Jl7ZqfGYjycnNOtFRwM+BWUPtMSK+GRHnNDmWXwLvqGPcH4yIW5s47kZNAv5h0FpmHcjJzTqKpInAG0g/ap1VKO/ND8pdIul3kk6WNFvS0vw+q11zvRMlfTp/7pP0L7nO7yS9MZdvIenbub9fSTqwRiy7AhOBE0hJrr98jKSv5P5/I+ljhfH15M/vz+O8JrdnKNPgzZKuk3STpO/laYKk1ZK+mMuX9x/ZSnqJ0jv2bpL0LUl/UHp/28nArkrvB/tyHvxESRcqvcPu/PxkC/L0vDW35ytDidesHZzcrNMcAfw4In4HPCRpn0K3VwPHAdOB9wK7R8QM4AzgYzWGNzbX+QTwhVx2LEBETCclrbMlbVGl36NIT5L4GfAKSf1vQphHenbjayNib5592DLwzLM3v0hKaoeQ3sdVl5yUTgAOjoh9gBuBTxWqPJDLvwF8Opd9AfjvXP4DYKdcvgD4fUS8JiL+MZe9Nk+LPYCXAW+QtB3pyHTP3J4v1RuvWbs4uVmnOQpYnD8vpnDEBPwyItZGepfX70kvdwRYDnTXGF7/g2iXFeocQHpEGhHxW9LLPHev0u8s0nu0/pKHc2QuP5j0mK5NeRiV7+V6HelFk/dHxJPAd2s1tor9SYnnfyTdTHqO5851tGdxjuXHwMMDDH9pRKzJbbo5D+Mx4M/AGZL+DvjTAP2bjQpj2x2AWb0kvZj0wNi9JAUwBghJn8lVNhaq/6Xw/S/UXtb76zxdqFPtVSGVsexNekDtlfnM3YuAO4Gv5/4He67dcJ97J+DKiDiqRvdhtadK/88MIyI2SZpBeuD3LOCj5IcTm41WPnKzTvJO4JyI2DnSE+2nAatIRybNdC0wG555svxOpAfSFh0FnJjj6I6IHYEpknYmHTEeo/TWdfJpvaIbgF5JL5a0Oc8e8dXjetKpwpfnYU/IMQ7k58C7cv03k953BrAe2GqwEeZrettExI9IpyxfM4R4zdrCyc06yVGka0ZF3wfe3eTxnA6MkbScdMpwbj7VWTSrSiw/yOVnkF5z9BtJv66ML9KrSE4ErgOuIj29vpa5ktb0/5HeJjEXuEDSb0jJbrCfRHwReLOkm4C3kJ42vz4iHiSd3rylcENJNVsBl+bxXUN6UabZqOa3ApiVnKRxwNP59OLrgW9EhI++rNR8zc2s/HYClkjaDHgS+FCb4zFrOR+5mZlZ6fiam5mZlY6Tm5mZlY6Tm5mZlY6Tm5mZlY6Tm5mZlc7/B/cIrMAk0XiTAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dist,prots = [],set()\n",
    "protSeqs = dict(fasta_index(\"proteins/All_Cas9-Like-filtered.faa\",'fasta')) #All_%s-Like.faa\" % gene))\n",
    "seqCounter = {}\n",
    "for rec in parse(\"proteins/All_Cas9-Like-filtered.faa\",\"fasta\"): \n",
    "    seq = str(rec.seq).strip().upper()\n",
    "#     if len(seq)>2000:continue\n",
    "    if seq not in prots:\n",
    "        prots.add(seq)\n",
    "        dist.append(len(seq))\n",
    "        seqCounter[seq] = [rec.id]\n",
    "    else:\n",
    "        seqCounter[seq].append(rec.id)\n",
    "dist = Series(dist)\n",
    "plt.title(\"Lengths of Unique Protein Coding Sequences matching the %s HMM\" % gene)\n",
    "plt.xlabel(\"Amino Acid Lengths\")\n",
    "plt.ylabel(\"Frequency Count\")\n",
    "hist = dist.hist(grid=True,figsize=(6,4))\n",
    "print (\"Total number of unique proteins:\", len(dist))\n",
    "plt.savefig(\"images/Unique%sLengths.png\" %(gene))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/bin/bash --login\n",
      "########## Define Resources Needed with SBATCH Lines ##########\n",
      " \n",
      "#SBATCH --time=6:00:00             # limit of wall clock time - how long the job will run (same as -t)\n",
      "#SBATCH --ntasks=1                  # number of tasks - how many tasks (nodes) that you require (same as -n)\n",
      "#SBATCH --cpus-per-task=20           # number of CPUs (or cores) per task (same as -c)\n",
      "#SBATCH --mem=20G                    # memory required per node - amount of memory (in bytes)\n",
      "#SBATCH --job-name DomainSearch      # you can give your job a name for easier identification (same as -J)\n",
      "#SBATCH --account shadeash-colej \n",
      "#SBATCH --mail-type=END                      # Mail type:\n",
      "#SBATCH --mail-user=dooley.shanek@gmail.com  # Mail user:\n",
      "\n",
      "\n",
      "MY_PATH=/mnt/research/germs/shane/transActRNA/data\n",
      "cd $MY_PATH\n",
      "\n",
      "hmmsearch --max --cpu 20 --pfamtblout $MY_PATH/hmm/Cas9-Like_phi.pfamtbl --tblout $MY_PATH/hmm/Cas9-Like_phi.tbl --domtblout $MY_PATH/hmm/Cas9-Like_phi.faa.domtbl $MY_PATH/hmm/phi_domains.hmm $MY_PATH/proteins/All_Cas9-Like.faa > $MY_PATH/hmm/Cas9-Like_phi.faa.log\n",
      "\n",
      "\n",
      "##hmmsearch --max --cpu 20 --pfamtblout $MY_PATH/domains/DiverseCas9s_uniref.pfamtbl --tblout $MY_PATH/domains/DiverseCas9s_uniref.tbl --domtblout $MY_PATH/domains/DiverseCas9s_uniref.domtbl $MY_PATH/DiverseCas9s.faa.hmm $MY_PATH/domains/uniref90.fasta > $MY_PATH/domains/DiverseCas9s_uniref.log\n",
      "\n",
      "#hmmsearch --max --cpu 20 --pfamtblout $MY_PATH/domains/DiverseCas9s_uniref.pfamtbl --tblout $MY_PATH/domains/DiverseCas9s_uniref.tbl --domtblout $MY_PATH/domains/DiverseCas9s_uniref.domtbl $MY_PATH/DiverseCas9s.faa.hmm $MY_PATH/domains/uniref90.fasta > $MY_PATH/domains/DiverseCas9s_uniref.log\n",
      "\n",
      "##hmmsearch --max --cpu 20 --pfamtblout $MY_PATH/domains/Cas9-Like-clustered.pfamtbl --tblout $MY_PATH/domains/Cas9-Like-clustered.tbl --domtblout $MY_PATH/domains/Cas9-Like-clustered.faa.domtbl $MY_PATH/domains/Pfam-A.full.hmm $MY_PATH/Cas9-Like-clustered.faa > $MY_PATH/domains/Cas9-Like-clustered.faa.log\n",
      "\n",
      "##hmmsearch --max --cpu 20 --pfamtblout $MY_PATH/domains/DiverseCas9s1.pfamtbl       --tblout $MY_PATH/domains/DiverseCas9s1.tbl       --domtblout $MY_PATH/domains/DiverseCas9s1.domtbl           $MY_PATH/domains/Pfam-A.full.hmm $MY_PATH/DiverseCas9s.faa        > $MY_PATH/domains/DiverseCas9s_Log_1.log\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cat ../scripts/hpc/DomainSearch.sb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Profile the domains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "sbatch ../scripts/hpc/DomainSearch.sb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading Table Definitions\n",
      "28671\n",
      "\n",
      "Dumping pickles/Cas9_HMM_DOMAIN_Search_Results.p\n"
     ]
    }
   ],
   "source": [
    "spacePat = re.compile( r'^[\\t ]*$')\n",
    "hits=SamplesDict()\n",
    "print(\"Reading Table Definitions\")\n",
    "for line in open(\"hmm/Cas9-Like_phi.faa.domtbl\"):\n",
    "    if line.startswith( '#' ):continue\n",
    "    fields = re.split( r'(\\[[^\\[]*[^\\S]+[^\\]]*\\]|[^\\t ]+)', line.strip() )\n",
    "    fields = DomainHit(list(filter(lambda i: not spacePat.search(i), fields)))\n",
    "    hits[fields.hit]=fields\n",
    "print(len(hits.samples))\n",
    "dump(hits,\"pickles/%s_HMM_DOMAIN_Search_Results.p\" % (gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Newest Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table align='left'>\n",
       "    <tr style='background-color:#9B7E46;color:white'><td>Number of Proteins:</td><td>28671</td></tr><tr style='background-color:#373D20;color:white'><td>Has all domains</td><td>27026</td></tr><tr style='background-color:#BCBD8B;color:white'><td>Number of sequences with no detected RuvCI domain</td><td>921</td></tr><tr style='background-color:#717744;color:white'><td>Number of sequences with no detected RuvCII domain</td><td>971</td></tr><tr style='background-color:#766153;color:white'><td>Number of sequences with no detected RuvCIII domain</td><td>432</td></tr><tr style='background-color:#F4B266;color:white'><td>Number of sequences with no detected HnH</td><td>186</td></tr>\n",
       "    </table>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<table align='left'>\n",
      "    <tr style='background-color:#9B7E46;color:white'><td>Number of Proteins:</td><td>28671</td></tr><tr style='background-color:#373D20;color:white'><td>Has all domains</td><td>27026</td></tr><tr style='background-color:#BCBD8B;color:white'><td>Number of sequences with no detected RuvCI domain</td><td>921</td></tr><tr style='background-color:#717744;color:white'><td>Number of sequences with no detected RuvCII domain</td><td>971</td></tr><tr style='background-color:#766153;color:white'><td>Number of sequences with no detected RuvCIII domain</td><td>432</td></tr><tr style='background-color:#F4B266;color:white'><td>Number of sequences with no detected HnH</td><td>186</td></tr>\n",
      "    </table>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "hits = load(open(\"pickles/%s_HMM_DOMAIN_Search_Results.p\" % (gene),\"rb\"))\n",
    "hasAllDomains = hits[\"RuvC_1_Cas9\"].intersection(hits[\"RuvC_2_Cas9\"].intersection(hits[\"RuvC_3_Cas9\"].intersection(hits[\"HNH_Cas\"])))\n",
    "allSamples = set(hits.samples.keys())\n",
    "nSamples = len(allSamples)\n",
    "noRuvC1 = allSamples.difference(hits[\"RuvC_1_Cas9\"])\n",
    "noRuvC2 = allSamples.difference(hits[\"RuvC_2_Cas9\"])\n",
    "noRuvC3 = allSamples.difference(hits[\"RuvC_3_Cas9\"])\n",
    "noHNH   = allSamples.difference(hits[\"HNH_Cas\"])\n",
    "\n",
    "htmlString =\"\"\"<table align='left'>\n",
    "    \"\"\"\\\n",
    "    \"<tr style='background-color:#9B7E46;color:white'><td>Number of Proteins:</td><td>%i</td></tr>\" % (nSamples) +\\\n",
    "    \"<tr style='background-color:#373D20;color:white'><td>Has all domains</td><td>%i</td></tr>\" % (len(hasAllDomains)) +\\\n",
    "    \"<tr style='background-color:#BCBD8B;color:white'><td>Number of sequences with no detected RuvCI domain</td><td>%i</td></tr>\" % (len(noRuvC1)) +\\\n",
    "    \"<tr style='background-color:#717744;color:white'><td>Number of sequences with no detected RuvCII domain</td><td>%i</td></tr>\" % (len(noRuvC2)) +\\\n",
    "    \"<tr style='background-color:#766153;color:white'><td>Number of sequences with no detected RuvCIII domain</td><td>%i</td></tr>\" % (len(noRuvC3)) +\\\n",
    "    \"<tr style='background-color:#F4B266;color:white'><td>Number of sequences with no detected HnH</td><td>%i</td></tr>\" % (len(noHNH)) +\\\n",
    "    \"\"\"\n",
    "    </table>\n",
    "\"\"\"\n",
    "display(HTML(htmlString))\n",
    "#    \"<tr style='background-color:#766153;color:white'><td>Number of proteins with neither an RuvCI nor a RuvCIII</td><td>%i</td></tr>\" % (len(neitherDomain)) +\\\n",
    "print(htmlString)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Old Run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table align='left'>\n",
       "    <tr style='background-color:#9B7E46;color:white'><td>Number of Proteins:</td><td>30782</td></tr><tr style='background-color:#373D20;color:white'><td>Has all domains</td><td>29050</td></tr><tr style='background-color:#BCBD8B;color:white'><td>Number of sequences with no detected RuvCI domain</td><td>964</td></tr><tr style='background-color:#717744;color:white'><td>Number of sequences with no detected RuvCII domain</td><td>1014</td></tr><tr style='background-color:#766153;color:white'><td>Number of sequences with no detected RuvCIII domain</td><td>450</td></tr><tr style='background-color:#F4B266;color:white'><td>Number of sequences with no detected HnH</td><td>181</td></tr>\n",
       "    </table>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<table align='left'>\n",
      "    <tr style='background-color:#9B7E46;color:white'><td>Number of Proteins:</td><td>30782</td></tr><tr style='background-color:#373D20;color:white'><td>Has all domains</td><td>29050</td></tr><tr style='background-color:#BCBD8B;color:white'><td>Number of sequences with no detected RuvCI domain</td><td>964</td></tr><tr style='background-color:#717744;color:white'><td>Number of sequences with no detected RuvCII domain</td><td>1014</td></tr><tr style='background-color:#766153;color:white'><td>Number of sequences with no detected RuvCIII domain</td><td>450</td></tr><tr style='background-color:#F4B266;color:white'><td>Number of sequences with no detected HnH</td><td>181</td></tr>\n",
      "    </table>\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#hits = load(open(\"pickles/%s_HMM_DOMAIN_Search_Results.p\" % (gene),\"rb\"))\n",
    "hasAllDomains = hits[\"RuvC_1_Cas9\"].intersection(hits[\"RuvC_2_Cas9\"].intersection(hits[\"RuvC_3_Cas9\"].intersection(hits[\"HNH_Cas\"])))\n",
    "allSamples = set(hits.samples.keys())\n",
    "nSamples = len(allSamples)\n",
    "noRuvC1 = allSamples.difference(hits[\"RuvC_1_Cas9\"])\n",
    "noRuvC2 = allSamples.difference(hits[\"RuvC_2_Cas9\"])\n",
    "noRuvC3 = allSamples.difference(hits[\"RuvC_3_Cas9\"])\n",
    "noHNH   = allSamples.difference(hits[\"HNH_Cas\"])\n",
    "\n",
    "htmlString =\"\"\"<table align='left'>\n",
    "    \"\"\"\\\n",
    "    \"<tr style='background-color:#9B7E46;color:white'><td>Number of Proteins:</td><td>%i</td></tr>\" % (nSamples) +\\\n",
    "    \"<tr style='background-color:#373D20;color:white'><td>Has all domains</td><td>%i</td></tr>\" % (len(hasAllDomains)) +\\\n",
    "    \"<tr style='background-color:#BCBD8B;color:white'><td>Number of sequences with no detected RuvCI domain</td><td>%i</td></tr>\" % (len(noRuvC1)) +\\\n",
    "    \"<tr style='background-color:#717744;color:white'><td>Number of sequences with no detected RuvCII domain</td><td>%i</td></tr>\" % (len(noRuvC2)) +\\\n",
    "    \"<tr style='background-color:#766153;color:white'><td>Number of sequences with no detected RuvCIII domain</td><td>%i</td></tr>\" % (len(noRuvC3)) +\\\n",
    "    \"<tr style='background-color:#F4B266;color:white'><td>Number of sequences with no detected HnH</td><td>%i</td></tr>\" % (len(noHNH)) +\\\n",
    "    \"\"\"\n",
    "    </table>\n",
    "\"\"\"\n",
    "display(HTML(htmlString))\n",
    "#    \"<tr style='background-color:#766153;color:white'><td>Number of proteins with neither an RuvCI nor a RuvCIII</td><td>%i</td></tr>\" % (len(neitherDomain)) +\\\n",
    "print(htmlString)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Profile the lengths of where the domains are located in the protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Mean distance from RuvC_1 to start of protein: 11 std: 9\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAakAAAEWCAYAAADcsGj7AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3XucHFWZ//HPFwjhFu4QQ4IEIaiAC0rACCsEUEFUQBckKBKQNcrCAgqrwM8VEFFZCLqsgkZhCRcTEJWboiA4sCi3JFzDNQJKTEgAA8kglyQ8vz/OaahpenoqM9Mzlcz3/Xr1a6pP1al66nRVPV2naqoVEZiZmVXRSv0dgJmZWWecpMzMrLKcpMzMrLKcpMzMrLKcpMzMrLKcpMzMrLKW6yQlaaaksf0dR3+S9ElJT0tql/TeEtO3SfrXPPxZSTe0Psq+JWl1SddKelHSz/s7Hus57+uNVX1bl3SypJ/2ZB6VTVKSnpL0obqywyTdVnsfEdtERFsX8xkpKSSt0qJQ+9vZwNERsVZE3LMsFSPisoj4SFfTSbpI0re6HWHfOwAYCmwQEQf29cIlnSppcf7i8IKkP0n6QC/OfytJP5f0XD443S/pK5JW7qLeJEmPSnpd0mFdTHuRpNckLcqvByV9R9I6vbUey6LMvt6ZvP9v2cshdfjC14N5nCrp0h7Mok+39WU9FkTEtyOiR21U2SS1vKhA8tsMmNnPMVTNZsBjEbGk0cg++swuj4i1gA2BPwC98i1X0hbAncDTwHsiYh3gQGA0MKSL6vcB/wbMKLm4/4qIIcBGwOHAGOCPktbsTuzWUS9th0239X6Ip/dFRCVfwFPAh+rKDgNuazQNsBMwDVgIzAPOyeV/BQJoz68PkJLz14G/APOBi4F1CvM9NI97HvjPuuWcClwJXJqX9a952bcDLwBzgR8AqxbmF6SDw+PAIuB0YItcZyFwRXH6unVuGCswOK9PAC8Bf+6k/oeBR4AXc1y3AP9a356AgO/lZbwI3A9sC0wAFgOv5eVdm6c/EfhzXp+HgE/Wf06ks7wFwJPARwvj1wf+F5iTx19VGPdx4N7cln8C/qkw7mvA3/IyHwX2bLC+p+VYF+d4j8jx/DGv39+BbzXbBoCRuV0PJyWDBcCXgB1zu7wA/KDJtnsqcGnh/dZ5fhs12o4L28iWpETwDLByYdwngfvz8KXAr3u4b90GHNbFNBcB36orG0Lavo9utm12pw1J+8PNpH3uOeAyYN1O9vVTSfvMxXlbmAmM7mQ9buXNfaQdOCiXfwGYlbeHa4BNOqm/Wm7z53PMd5POXM4AlgKv5Pn+IE//33l9FwLTgQ/WbRfFY8fRdNxW7+skhncDbXn5M4F9O9vWO9kWrwQuz201A9iurl2/lj+TV4FVmiyvs2PBJsAvgGdJ+/oxjfaFwjYxnnRcfg74f11urz3Z2Fv5YtmT1O3A5/LwWsCYuoZZpVDv83kDfUee9pfAJYUDSjvwz8CqpAPtYjruIIuB/Uk76erADqSDyyp5eQ8Dx9UdgK4B1ga2yRvDTXn565AO8uM7aYdOYy3Me8tO6m5I2hkOAAYBXwaW0DhJ7UXaqdYlJax3A8OaHLAOzBvnSsBBpIPAsMJ8F5MOBCsDR5ISkvL4X5N2mvVyXLvl8veRDnbvz/XG5894MPBO0s6/SeFz3aKT9T6VjknisLze/54/o9W72AZG5nb9Eekg9RHSwegqYGNgeI5zt66WT9qGvkvaIVdptB3Xf46k5P/hwrifAyfm4WeAw3u4b3UrSeXyi0lnifRmG5IS9IfzZ70RKbl8v5N9/dQ8r33ydvId4I4m69JhHwH2yJ/H+/Ly/ge4tZO6XwSuBdbIy9oBWDuPayPvS4XpDwE2yNvZ8fnzWq3JseONbaWT5Q/KbXxy3pb2ICWbdzba1jvZFhfz5jHgBFIiGVRo13uBTXM8XS2vw3aR12M68I08/TuAJ4C9GuwLtW3iJ3lZ25GOhe9uti1Wvbvvqtyn/4KkF4Dzmky7GNhS0oYR0R4RdzSZ9rOkM60nIqIdOAkYl093DyB9Q7gtIl4jNX79Aw5vj4irIuL1iHg5IqZHxB0RsSQingJ+DOxWV+fMiFgYETOBB4Eb8vJfBK4HOrvpoVmsXdkHeCgiroyIxcD3STtNI4tJ35TfRUomD0fE3M5mHBE/j4g5uQ0uJ50l7lSY5C8R8ZOIWApMBoYBQyUNAz4KfCkiFkTE4oi4Jdf5AvDjiLgzIpZGxGTSRjyG9K11MLC1pEER8VRE/LlEG9TMiYj/yZ/Ry5Rr19Mj4pWIuIGUhKdExPyI+Bvwf3T+mQF8Om+zL+f1OiDKd8lMAQ4GkDSE9DlOyeM2IJ3N9Jc5pDNh6MU2jIhZEXFjRLwaEc8C5/DWfajotoj4Td6+LiEd8Mr6LHBhRMyIiFdz3B+QNLLBtItJbb5l3ianR8TCzmYcEZdGxPN5O5vIm1+wajocO0rEOob0BeC7EfFaRNwMXEfePkqaXjgGnEP60jCmMP7ciHg6x7Osy9uR1EPwzTz9E6QkNK5JPKfl4+Z9pC7opp9d1ZPU/hGxbu1F6jLrzBHAVsAjku6W9PEm025C6qKo+Qvpm8/QPO7p2oiI+AfpVL/o6eKbfCH7OknPSFoIfJt0FlM0rzD8coP3a3Uj1q7Ur0vUx14YdzOpO/CHwLx8kX3tzmYs6VBJ9xa+QGxLx3V+IxnmNoS0jpsCf4+IBQ1muxlwfN0Xk01JZ0+zgONI38zmS5oqaZMu1r+ofr3LtGt3PzOAK/I2O5T0pWSHZYj1Z8CnJA0GPgXMiIharM+TEn5/GU7qIoNebENJG+fP9G95H7qUt+5DRcUvW/8AVluGayod4s4J9nnSutW7BPgdMFXSHEn/JWlQZzOWdLykh/MNLS+QekqK69Fw/+si1qcj4vVC2V86ibUzxWPA68DsPN9GMS3r8jYDNqnbZ0+m+fGp/rNrth9VPkmVFhGPR8TBpK6EM4Er8wXe+rMgSN8GNyu8fzupO2ge6VvqiNoISauTvkl1WFzd+/NJ131GRcTapA9J3V+b0rF2ZS7pIA+AJBXf14uIcyNiB1KX5FbAf9RGFaeTtBnp29LRpLuK1iUdiMus89PA+pLW7WTcGcUvJhGxRkRMyfH9LCL+mdQeQfqcy6r/zHrSruUXGvEcqcvo1HwWCemMYo3aNJLeVlfnIdKB4aPAZ0hJq+b3wL/0ZoxlSVoL+BDpDAh6tw2/Q/qM/invQ4fQe/tQvQ5x5+PEBqTrnR3kM/3TImJrYGfSNdNDa6OL00r6IOn6zqeB9fJ+8SId16N+O2x0fKqPdVNJxWP12xvF2kTxGLAS6fg2p5MYulpefbxPA0/W7bNDImKfZYivqRUmSUk6RNJG+RvAC7l4Keli3uukvtKaKcCXJW2ed7xvk/rZl5AuMn5C0s6SViVdnOxqZxlCuvbTLuldpGswvaVZrF35NbCNpE/lb5nHAG9rNKGkHSW9P39LfInU5780j55Hx/arJf9nc93DSWdSXcpdiNcD50laT9IgSbvm0T8BvpTjkKQ1JX1M0hBJ75S0Rz67eIX0LXxpJ4spoyftukwi4hHSt/Gv5qL7SJ/L9pJWI50d1vsZ6fPalY53Bp4C7CzprFpyk7SlpEs7SfxvkLRqXp6AQZJWqzsYdVZvsKQdSNeTFpBueoHebcMhpGvBL0gazptfkHpD/fb7M+Dw3P6DSXHfmbvqO5C0u6T3KN3ev5DU/dfZfjGElKSfBVaR9A3SdeiuYhvZ5HO4k7Q/fjXvK2OBTwBTu5hv0Q6FY8BxpC70zi6HdLW8+nW+C1go6WtK/7O1sqRtJe24DPE1tcIkKWBvYKakdtIdNuNyX/g/SHfi/DGfjo4BLiSdxt9Kuoj4CumiOvma0b+TPpS5pIuG80kfbGdOIH3jXUQ60F7ei+vVaaxdyd/iDyRduH8eGEW6y62RtUmxL+DNOxvPzuMuIF0LekHSVfmb/kTSzSrzgPc0mW8jnyPt7I+Q2va4HO800vWbH+Q4ZpFuMoDUt1+7AeEZ0hnzycuwzHrdbtduOguYIGnjiHgM+CbprOhx0o0M9aYAY4Gb8+cIQL4O9wHSReiZkl4k3Vk1jbT9NXMDKbnvDEzKw7s2mf6rkhaRuvcuJl0g3zkiXsrje7MNTyPdyPAi6cvVL7s5n0ZOBSbn7ffTEXET6a7dX5D28S3o/BrK20hfXBeSboi6hdQVCek4c4CkBZLOJX0RuR54jLQPvULX3Xu1LyDPS3rLvwZEui6+L+ms+jnSdflD8xefsq4m3dy0gLTvfSpfn3qLEsurPxYsJSWx7UnbwHPAT0ndnL2idreVdSJ/Q3yB1JX3ZH/HY2ZWlqRTSTd9HNLfsXTXinQm1WskfULSGrmv+mzgAdKtmmZm1oecpBrbj3QBcQ6pi2xc+JTTKk7pWYztDV5+Ioktt9zdZ2ZmleUzKTMzq6xqPlCwhTbccMMYOXJkt+q+9NJLrLmmn63pdnAb1LgdkoHQDtOnT38uIjbq6+UOuCQ1cuRIpk2b1q26bW1tjB07tncDWg65HdwGNW6HZCC0g6S/dD1V73N3n5mZVZaTlJmZVZaTlJmZVZaTlJmZVZaTlJmZVZaTlJmZVZaTlJmZVZaTlJmZVZaTlJmZVdaAe+JET0yfO53dT9u9X5Ydp/hBwGY28PhMyszMKstJyszMKstJyszMKstJyszMKstJyszMKstJyszMKstJyszMKstJyszMKstJyszMKqtlSUrSppL+IOlhSTMlHZvLT5X0N0n35tc+hTonSZol6VFJexXKd5D0QB53riTl8sGSLs/ld0oa2ar1MTOzvtfKM6klwPER8W5gDHCUpK3zuO9FxPb59RuAPG4csA2wN3CepJXz9OcDE4BR+bV3Lj8CWBARWwLfA85s4fqYmVkfa1mSioi5ETEjDy8CHgaGN6myHzA1Il6NiCeBWcBOkoYBa0fE7RERwMXA/oU6k/PwlcCetbMsMzNb/vXJNancDfde4M5cdLSk+yVdKGm9XDYceLpQbXYuG56H68s71ImIJcCLwAYtWAUzM+sHLX8KuqS1gF8Ax0XEQknnA6cDkf9OBD4PNDoDiibldDGuGMMEUnchQ4cOpa2tbRnXIhkxeARnb3V2t+r2VHdjboX29vZKxdMf3AaJ2yFxO7ROS5OUpEGkBHVZRPwSICLmFcb/BLguv50NbFqoPgKYk8tHNCgv1pktaRVgHeDv9XFExCRgEsDo0aNj7Nix3VqfiVMmcsJjJ3Srbk/FwdX5qY62tja624YrCrdB4nZI3A6t08q7+wRcADwcEecUyocVJvsk8GAevgYYl+/Y25x0g8RdETEXWCRpTJ7nocDVhTrj8/ABwM35upWZma0AWnkmtQvwOeABSffmspOBgyVtT+qWewr4IkBEzJR0BfAQ6c7AoyJiaa53JHARsDpwfX5BSoKXSJpFOoMa18L1MTOzPtayJBURt9H4mtFvmtQ5AzijQfk0YNsG5a8AB/YgTDMzqzA/ccLMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCqryyQl6VhJayu5QNIMSR/pi+DMzGxgK3Mm9fmIWAh8BNgIOBz4bkujMjMzo1ySUv67D/C/EXFfoczMzKxlyiSp6ZJuICWp30kaArzeVSVJm0r6g6SHJc2UdGwuX1/SjZIez3/XK9Q5SdIsSY9K2qtQvoOkB/K4cyUplw+WdHkuv1PSyGVbfTMzq7IySeoI4ERgx4j4B7AqqcuvK0uA4yPi3cAY4ChJW+d53RQRo4Cb8nvyuHHANsDewHmSVs7zOh+YAIzKr70LsS2IiC2B7wFnlojLzMyWE2WSVABbA8fk92sCq3VZKWJuRMzIw4uAh4HhwH7A5DzZZGD/PLwfMDUiXo2IJ4FZwE6ShgFrR8TtERHAxXV1avO6EtizdpZlZmbLv1VKTHMeqXtvD+CbwCLgF8COZReSu+HeC9wJDI2IuZASmaSN82TDgTsK1WbnssV5uL68VufpPK8lkl4ENgCeq1v+BNKZGEOHDqWtra1s6B2MGDyCs7c6u1t1e6q7MbdCe3t7peLpD26DxO2QuB1ap0ySen9EvE/SPQARsUDSqmUXIGktUlI7LiIWNjnRaTQimpQ3q9OxIGISMAlg9OjRMXbs2C6ibmzilImc8NgJ3arbU3HwW1ar37S1tdHdNlxRuA0St0PidmidMt19i/O1oQCQtBElbpzI0w4iJajLIuKXuXhe7sIj/52fy2cDmxaqjwDm5PIRDco71JG0CrAO8PcysZmZWfWVSVLnAr8CNpZ0BnAb8O2uKuVrQxcAD0fEOYVR1wDj8/B44OpC+bh8x97mpBsk7spdg4skjcnzPLSuTm1eBwA35+tWZma2Auiyuy8iLpM0HdiT1L22f0Q8XGLeuwCfAx6QdG8uO5n0j8BXSDoC+CtwYF7OTElXAA+R7gw8KiKW5npHAhcBqwPX5xekJHiJpFmkM6hxJeIyM7PlRJdJStIYYGZE/DC/HyLp/RFxZ7N6EXEbnf/T756d1DkDOKNB+TRg2wblr5CTnJmZrXjKdPedD7QX3r+Uy8zMzFqq1GORitd5IuJ1yt0VaGZm1iNlktQTko6RNCi/jgWeaHVgZmZmZZLUl4Cdgb+Rbvl+P/kfY83MzFqpzN198/Fdc2Zm1g/K3N23EfAFYGRx+oj4fOvCMjMzK3cDxNXA/wG/B5Z2Ma2ZmVmvKZOk1oiIr7U8EjMzszplbpy4TtI+LY/EzMysTpkkdSwpUb0iaaGkRZIWtjowMzOzMnf3DemLQMzMzOp1eSal5BBJ/5nfbyppp9aHZmZmA12Z7r7zgA8An8nv24EftiwiMzOzrOW/zGtmZtZdLf1lXjMzs55o2S/zmpmZ9VQrf5nXzMysR8o8u+/twD+Aa4tlEfHXVgZmZmZW5saJX5OuRwlYDdgceBTYpoVxmZmZlerue0/xvaT3AV9sWURmZmZZmRsnOoiIGcCOLYjFzMysgzLXpL5SeLsS8D7g2ZZFZGZmlpW5JlV8dt8S0jWqX7QmHDMzszeVuSZ1Wl8EYmZmVq9Md981zcZHxL69F46ZmdmbynT3PQm8Dbg0vz8YeAr4XYtiMjMzA8olqfdGxK6F99dKujUiTm5VUGZmZlDuFvSNJL2j9kbS5sBGrQvJzMwsKXMm9WWgTdIT+f1I/M+8ZmbWB8rc3fdbSaOAd+WiRyLi1daGZWZmVu7n49cA/gM4OiLuA94u6eMl6l0oab6kBwtlp0r6m6R782ufwriTJM2S9KikvQrlO0h6II87V5Jy+WBJl+fyOyWNXKY1NzOzyitzTep/gddIPyEPMBv4Vol6FwF7Nyj/XkRsn1+/AZC0NTCO9NDavYHz8g8tApwPTABG5VdtnkcACyJiS+B7wJklYjIzs+VImSS1RUT8F7AYICJeJj0RvamIuBX4e8k49gOmRsSrEfEkMAvYSdIwYO2IuD0iArgY2L9QZ3IevhLYs3aWZWZmK4YyN068Jml13vz5+C2AnlyTOlrSocA04PiIWAAMB+4oTDM7ly3Ow/Xl5L9PA0TEEkkvAhsAz9UvUNIE0tkYQ4cOpa2trVuBjxg8grO3OrtbdXuquzG3Qnt7e6Xi6Q9ug8TtkLgdWqdMkjoF+C2wqaTLgF2Aw7q5vPOB00kJ73RgIvB5Gp+ZRZNyuhjXsTBiEjAJYPTo0TF27NhlCrpm4pSJnPDYCd2q21NxcMNV6xdtbW10tw1XFG6DxO2QuB1ap8zdfTdKmgGMISWGYyPiLWcrZUTEvNqwpJ8A1+W3s4FNC5OOAObk8hENyot1ZktaBViH8t2LZma2HChzd98uwCsR8WtgXeBkSZt1Z2H5GlPNJ4HanX/XAOPyHXubk26QuCsi5gKLJI3J15sOBa4u1Bmfhw8Abs7XrczMbAVRprvvfGA7SduRbkW/kHQDw27NKkmaAowFNpQ0m9RtOFbS9qRuuafI/xQcETMlXQE8RPo5kKMiYmme1ZGkOwVXB67PL4ALgEskzSKdQY0rsS5mZrYcKZOklkRESNoPODciLpA0vqtKEXFwg+ILmkx/BnBGg/JpwLYNyl8BDuwqDjMzW36VSVKLJJ0EHALsmv9/aVBrwzIzMyv3f1IHkW45PyIiniHd+n1WS6MyMzOj3N19zwDnFN7/lXRNyszMrKXKnEmZmZn1CycpMzOrrE6TlKSb8l8/uNXMzPpFs2tSwyTtBuwraSp1jyGKiBktjczMzAa8ZknqG8CJpEcRnVM3LoA9WhWUmZkZNElSEXElcKWk/4yI0/swJjMzM6DcLeinS9oX2DUXtUXEdc3qmJmZ9YYyD5j9DnAs6bl6DwHH5jIzM7OWKvNYpI8B20fE6wCSJgP3ACe1MjAzM7Oy/ye1bmF4nVYEYmZmVq/MmdR3gHsk/YF0G/qu+CzKzMz6QJkbJ6ZIagN2JCWpr+Xn+ZmZmbVUmTMp8i/kXtPiWMzMzDrws/vMzKyynKTMzKyymiYpSStJerCvgjEzMytqmqTy/0bdJ+ntfRSPmZnZG8rcODEMmCnpLuClWmFE7NuyqMzMzCiXpE5reRRmZmYNlPk/qVskbQaMiojfS1oDWLn1oZmZ2UBX5gGzXwCuBH6ci4YDV7UyKDMzMyh3C/pRwC7AQoCIeBzYuJVBmZmZQbkk9WpEvFZ7I2kV0i/zmpmZtVSZJHWLpJOB1SV9GPg5cG1rwzIzMyuXpE4EngUeAL4I/Ab4eiuDMjMzg3J3972ef+jwTlI336MR4e4+MzNruTJ3930M+DNwLvADYJakj5aod6Gk+cXHKklaX9KNkh7Pf9crjDtJ0ixJj0raq1C+g6QH8rhzJSmXD5Z0eS6/U9LIZVlxMzOrvjLdfROB3SNibETsBuwOfK9EvYuAvevKTgRuiohRwE35PZK2BsYB2+Q650mq/S/W+cAEYFR+1eZ5BLAgIrbM8ZxZIiYzM1uOlElS8yNiVuH9E8D8ripFxK3A3+uK9wMm5+HJwP6F8qkR8WpEPAnMAnaSNAxYOyJuz12MF9fVqc3rSmDP2lmWmZmtGDq9JiXpU3lwpqTfAFeQrkkdCNzdzeUNzT+gSETMlVT7f6vhwB2F6WbnssV5uL68VufpPK8lkl4ENgCe62ZsZmZWMc1unPhEYXgesFsefhZY762T90ijM6BoUt6szltnLk0gdRkydOhQ2trauhEijBg8grO3OrtbdXuquzG3Qnt7e6Xi6Q9ug8TtkLgdWqfTJBURh7dgefMkDctnUcN4s9twNrBpYboRwJxcPqJBebHO7PwPxuvw1u5FACJiEjAJYPTo0TF27NhuBT9xykROeOyEbtXtqTi4OjdUtrW10d02XFG4DRK3Q+J2aJ0yd/dtLukcSb+UdE3t1c3lXQOMz8PjgasL5ePyHXubk26QuCt3DS6SNCZfbzq0rk5tXgcAN/vWeDOzFUuZn+q4CriA9JSJ18vOWNIUYCywoaTZwCnAd4ErJB0B/JV0fYuImCnpCuAhYAlwVEQszbM6knSn4OrA9flFjukSSbNIZ1DjysZmZmbLhzJJ6pWIOHdZZxwRB3cyas9Opj8DOKNB+TRg2wblr5CTnJmZrZjKJKn/lnQKcAPwaq0wIma0LCozMzPKJan3AJ8D9uDN7r7I783MzFqmTJL6JPCO4s91mJmZ9YUyT5y4D1i31YGYmZnVK3MmNRR4RNLddLwmtW/LojIzM6Nckjql5VGYmZk1UOb3pG7pi0DMzMzqdZmkJC3izWfirQoMAl6KiLVbGZiZmVmZM6khxfeS9gd2allEZmZmWZm7+zqIiKvw/0iZmVkfKNPd96nC25WA0XTykxhmZma9qczdfcXflVoCPEX6VVwzM7OWKnNNqhW/K2VmZtalZj8f/40m9SIiTm9BPGZmZm9odib1UoOyNYEjgA0AJykzM2upZj8fP7E2LGkIcCxwODAVmNhZPTMzs97S9JqUpPWBrwCfBSYD74uIBX0RmJmZWbNrUmcBnwImAe+JiPY+i8rMzIzm/8x7PLAJ8HVgjqSF+bVI0sK+Cc/MzAayZteklvlpFGZmZr3JicjMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCrLScrMzCqrX5KUpKckPSDpXknTctn6km6U9Hj+u15h+pMkzZL0qKS9CuU75PnMknSuJPXH+piZWWv055nU7hGxfUSMzu9PBG6KiFHATfk9krYGxgHbAHsD50laOdc5H5gAjMqvvfswfjMza7EqdfftR/rNKvLf/QvlUyPi1Yh4EpgF7CRpGLB2RNweEQFcXKhjZmYrgKY/ethCAdwgKYAfR8QkYGhEzAWIiLmSNs7TDgfuKNSdncsW5+H68reQNIF0xsXQoUNpa2vrVtAjBo/g7K3O7lbdnupuzK3Q3t5eqXj6g9sgcTskbofW6a8ktUtEzMmJ6EZJjzSZttF1pmhS/tbClAQnAYwePTrGjh27jOEmE6dM5ITHTuhW3Z6KgxuuWr9oa2uju224onAbJG6HxO3QOv3S3RcRc/Lf+cCvgJ2AebkLj/x3fp58NrBpofoIYE4uH9Gg3MzMVhB9nqQkrSlpSG0Y+AjwIHANMD5PNh64Og9fA4yTNFjS5qQbJO7KXYOLJI3Jd/UdWqhjZmYrgP7o7hsK/CrfLb4K8LOI+K2ku4ErJB0B/BU4ECAiZkq6AngIWAIcFRFL87yOBC4CVgeuzy8zM1tB9HmSiogngO0alD8P7NlJnTOAMxqUTwO27e0YzcysGqp0C7qZmVkHTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlLz5jigAAAJAklEQVRmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZTlJmZlZZy32SkrS3pEclzZJ0Yn/HY2ZmvWeV/g6gJyStDPwQ+DAwG7hb0jUR8VD/Rtb7dJr6ZblxSvTLcs3MYPk/k9oJmBURT0TEa8BUYL9+jsnMzHrJcn0mBQwHni68nw28v34iSROACfltu6RHu7m8DYHnull3uaRTG57BDbh2aMBtkLgdkoHQDpv1x0KX9yTV6Aj6lv6piJgETOrxwqRpETG6p/NZ3rkd3AY1bofE7dA6y3t332xg08L7EcCcforFzMx62fKepO4GRknaXNKqwDjgmn6OyczMesly3d0XEUskHQ38DlgZuDAiZrZwkT3uMlxBuB3cBjVuh8Tt0CKK8C3GZmZWTct7d5+Zma3AnKTMzKyynKRKGEiPXpJ0oaT5kh4slK0v6UZJj+e/6xXGnZTb5VFJe/VP1L1P0qaS/iDpYUkzJR2bywdMW0haTdJdku7LbXBaLh8wbVAkaWVJ90i6Lr8fkO3Q15ykulB49NJHga2BgyVt3b9RtdRFwN51ZScCN0XEKOCm/J7cDuOAbXKd83J7rQiWAMdHxLuBMcBReX0HUlu8CuwREdsB2wN7SxrDwGqDomOBhwvvB2o79Cknqa4NqEcvRcStwN/rivcDJufhycD+hfKpEfFqRDwJzCK113IvIuZGxIw8vIh0cBrOAGqLSNrz20H5FQygNqiRNAL4GPDTQvGAa4f+4CTVtUaPXhreT7H0l6ERMRfSwRvYOJcPiLaRNBJ4L3AnA6wtchfXvcB84MaIGHBtkH0f+CrweqFsILZDn3OS6lqpRy8NUCt820haC/gFcFxELGw2aYOy5b4tImJpRGxPeprLTpK2bTL5CtkGkj4OzI+I6WWrNChb7tuhvzhJdc2PXoJ5koYB5L/zc/kK3TaSBpES1GUR8ctcPCDbIiJeANpI11gGWhvsAuwr6SlSd/8eki5l4LVDv3CS6pofvZTWd3weHg9cXSgfJ2mwpM2BUcBd/RBfr5Mk4ALg4Yg4pzBqwLSFpI0krZuHVwc+BDzCAGoDgIg4KSJGRMRI0v5/c0QcwgBrh/6yXD8WqS/0w6OX+pWkKcBYYENJs4FTgO8CV0g6AvgrcCBARMyUdAXwEOluuKMiYmm/BN77dgE+BzyQr8kAnMzAaothwOR8Z9pKwBURcZ2k2xk4bdDMQNoW+o0fi2RmZpXl7j4zM6ssJykzM6ssJykzM6ssJykzM6ssJykzM6ssJynrd5KWSro3P2n7PklfkbRSHjda0rlN6o6U9Jm+i7bDsteV9G+9NK/jJK3RjXqHSdpkGeu8K7f3PZK2WNZldjHv/cs8gFnSlyQd2pvLthWTk5RVwcsRsX1EbAN8GNiH9P9ZRMS0iDimSd2RQL8kKWBdoFeSFHAcsExJKv//0mHAMiUp0oNQr46I90bEn5exLpKa/X/l/qRfC2gqIn4UERcv67JtAIoIv/zq1xfQXvf+HcDzpGegjQWuy+W7Affm1z3AEOAO4MVc9mVS0vo/YEZ+7ZzrjiU91udK0lMTLuPN/xPcEfgTcB/pyQBDSP+4fRbpiSP3A19sEPdU4OW87LNyvGcBDwIPAAc1qLMm8Ou8rAeBg4BjgNdynT/k6c4HpgEzgdMK9Z8CvgHcBhwCtAOP5hhWr1vW9rl97gd+BaxH+gLwDPC32rLqPwtgYm67m4CNcnkb8G3gFuB4YLM8/v789+3AzqQn6D+Z49kiv34LTM+fy7vy/E4FTijM+8zc9o8BH+zvbdKv6rz6PQC//KpPUrlsATC0LkldC+ySh9ciPTHljfG5fA1gtTw8CpiWh8eSktkIUg/C7cA/A6sCTwA75unWzvOdAHw9lw3OCWPzuhhHAg8W3v8LcCMpwQ0lPYVgWF2dfwF+Uni/Tv77FLBhoXz9/HflfBD/p8J0Xy1M1waM7qRd7wd2y8PfBL6fh99IEA3qBPDZPPwN4AeF5ZxXmO5aYHwe/jxwVR6+CDigMN1NwKg8/H7SI4UaJamJeXgf4Pf9vU36VZ2Xu/usqho9SfqPwDmSjgHWjYglDaYZBPxE0gPAz+nY9XRXRMyOiNdJ3/RHAu8E5kbE3QARsTDP9yPAofmRSHcCG5CSXjP/DEyJ9OTweaSzjh3rpnkA+JCkMyV9MCJe7GRen5Y0g3TGuE3delzeRRxIWofURrfkosnArl3VI/0URW3+l+Z1arTcDwA/y8OX1E1Xi2Et0tnVz3M7/pj0qKVGag/wnU76XMwAP7vPKkjSO4ClpKdKv7tWHhHflfRr0rftOyR9qEH1LwPzgO1IZ0yvFMa9WhheStr+ReOfURDw7xHxu2UJvasJIuIxSTuQ1uE7km6IiG92mEl6KOkJpLO7BZIuAlYrTPLSMsTUU8W2abbcRm24EvBCpJ/66Erts6l9LmaAb5ywipG0EfAjUjdT1I3bIiIeiIgzSd1v7wIWka4h1axDOjN6nfSA2K5+tvsRYBNJO+ZlDMk3BvwOODL/XAeStpK0Zl3d+mXfChyUfyhwI9KZS4enX+c78f4REZcCZwPvazCvtUkJ4UVJQ4GPNom/PgYA8hnaAkkfzEWfI53ZdWUl4IA8/BnSta9G/kR6IjjAZwvTvRFPpN/felLSgZCeLC9puxIxmL3B31isClbP3UGDSE+NvgQ4p8F0x0nanfRt+yHgelL31BJJ95Guh5wH/CIfGP9AF2cdEfGapIOA/8k/R/Ey6ScpfkrqdpqRf7bjWd78efBa3ecl/VHSgzmWr5K6we4jnVl8NSKeqVvke4CzJL0OLAaOzOWTgOslzY2I3SXdQ7pp4glSN2dnLgJ+JOll4AMR8XJh3Pg8bo08n8ObtUX2ErCNpOmka3gHdTLdMcCFkv6D1Da1eU8ldbceQ0p2nwXOl/R10uc7ldQ+ZqX4Kehm9gZJ7RGxVn/HYVbj7j4zM6ssn0mZmVll+UzKzMwqy0nKzMwqy0nKzMwqy0nKzMwqy0nKzMwq6/8DPSU9/V9yFJMAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "from pandas import Series\n",
    "ruvC1Coords = []\n",
    "ruvC3Coords = []\n",
    "samples = []\n",
    "difs = []\n",
    "for sample in hasAllDomains:\n",
    "    ruc1Start = hits.samples[sample][\"RuvC_1_Cas9\"].start\n",
    "    ruc3End = hits.samples[sample][\"RuvC_3_Cas9\"].dend\n",
    "    ruvC1Coords.append(ruc1Start)\n",
    "    ruvC3Coords.append(ruc3End)\n",
    "    difs.append(ruc3End-ruc1Start)\n",
    "    samples.append(sample)\n",
    "    \n",
    "ruvC1Dists = Series(ruvC1Coords,index=samples)\n",
    "ruvC3Dists = Series(ruvC3Coords,index=samples)\n",
    "distBetween = Series(difs,index=samples)\n",
    "\n",
    "print (\"\\nMean distance from RuvC_1 to start of protein: %.0f std: %.0f\" % (ruvC1Dists.mean(), ruvC1Dists.std()))\n",
    "\n",
    "# the histogram of the data\n",
    "n, bins, patches = plt.hist(ruvC1Dists,facecolor='green')\n",
    "plt.xlabel('Distance to start of protein')\n",
    "plt.ylabel('Number of sequences')\n",
    "plt.title(\"Histogram of distances from RuvC_1 Domain to start of protein\")\n",
    "plt.grid(True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Mean distance from RuvC_3 to end of protein: 256 std: 57\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZgAAAEWCAYAAABbgYH9AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3XmYHFW9//H3B8KeQFgjMmhYIl7gCpLI8nNLZAuIhOvVK6gQuVzjgooisiqowBV1riguKEoEZIkYFyJbiJgRF7aEPSyXELgwEoiQBBhWA9/fH+cMKYbumZ7uqZnu4fN6nnqm6tQ5Vef0VPe3z6nqKkUEZmZmA22Voa6AmZkNTw4wZmZWCgcYMzMrhQOMmZmVwgHGzMxK4QBjZmalcIAZQJJ+LOkrQ10Ps54kdUnacqjr0ShJD0jao86yp0h6TNIjA10vq8wBpkb5wH5W0lOSlkv6m6RPSnr5NYyIT0bEyTVuq643SVkkTZT0Uv4g6p5+P0R1OUfSC7kOSyXNkfTmAdz+zpIuz//HpZJukHRoH2W2lTRP0rI8/UHStr3k75D0XD5enpQ0X9KxktYYqHb0R0SMjIhF/S0naaykkDSinv1WOa66JO1Wz/bqJWlz4IvAthHxuip5jpd0f65fp6RfDmYdhyMHmP55X0SMAt4InAYcA5w9tFUaUA/nD6Lu6X2VMtX7YdNP34qIkcBmwN8ZoNc5f7D9EfgTsDWwIfApYJ8+ij4MfADYANgImAXM6KPMZ/Lxsinpw+1A4HJJqrsBranncTUyIq4d5Dq8EXg8IpZUWilpKnAwsEc+7iYAVw9i/YaniPBUwwQ8QDr4imk7Ay8B2+flc4BT8vxGwKXAcmAp8GdSQP9FLvMs0AUcnfP/CngEeAK4BtiusJ9zgB8ClwFPAdcDWxXWbwfMyft5FDg+p68CHAvcBzwOXAxsUKV9E4HOKuu+CswEzgeeBP4LWAP4LumD9+E8v0ZxW8DRwBJgMXAAsC/wv7mex/fyWr/8OublfYGne9Tn/MLyWCCAEaQP8Xk9tvcFYFae/wvwwwaPhRHA4cAzveTpAP6rR9obgGeA/fLygL2GpGPx2ny8LQZ+AKxeWB/A1rUcTz3q/GAu25Wn3fJx9WXg/3LdzgPW6+9xVXidTgb+mutyFbBRYf3BeT+PAydQ4X1YyLterss/cpkv57ruQXq/vZTbcE6Fsj8AvttLPdcjfclZTPrCcwqwal63KtAOPAYsysdGACMqfXbw6uN3V+Bv+X93KzCxH6/POwplHwI+Vji22vP/71Hgx8BajRz39UzuwTQgIm4gfQi8s8LqL+Z1GwNjgONTkTiY9E9/X6Rvct/K+a8AxgGbADcBF/TY3kHA14D1gYXAqQCSRgF/AK4EXk/6Vt79zetzpA+ld+d1y0gfLPWYQgoyo3PdTiC9MXYEdiB9wH25kP91wJqkHsiJwE+BjwLjSa/XibWcE5C0DqntC2us5yxgG0njCmkfBi6UtDbpA3JmjduqVJ/lwHPA94H/7k/ZiHgQmMfK42UgX8MXSYF0I1Ibdwc+3Ut1Kh5PFbwr/x0dK3seH8vTJGBLYCTpA7peHwYOJR37qwNHQRqWBM4kBZnXk3qbbb1s5/ukQLAl6Zg/BDg0Iv5A6qF296Q+VqHsdcAhkr4kaYKkVXusPxdYQXp/vRXYi/RFC+DjwH45fQKpp1sTSZuRAv0ppN7xUcCvJW1cyFbt9XkD6XPj+6TPmR2BW3KZbwJvymlbs/IYGlyDHdFadaLKNyfSgXlCnj+HlT2YrwOXkL811rKtwvrRpG9A6xW2+7PC+n2Bu/P8QcDNVbZzF7B7YXlT4J/kb1Y98k4kfcNbXpj+I6/7KnBNj/z3AfsWlvcGHihs61lWfsMblduzSyH/fOCAKvU+h/QhvjzX6X7gLYX1X6VKDyYvnw+cmOfHkb75rU16kwXw5gaPhXVIH97v7SVPBz16MDl9BvDTQXgNPw/8trDcswdT8XiqsJ1XvLY57Wrg04Xlbfp5XC0H1im8Tl8u5P80cGWePxGY0eN1f4HK78NVgedJ51i60z4BdBTqUbUnlfN8hPRl7WlSj+nYnD4mb3utQt6DgLl5/o/AJwvr9qLGHgxpmP0XPeoxG5haw+tzXPF/XMij3IbiKMduwP2NHPf1TO7BNG4z0nBFT98mfTO8StIiScdW24CkVSWdJuk+SU+SDkhI30a7Fa98eYb0rRFgc9IHVSVvBH6bT2YvJwWcF0lvmEoejojRheniwrqHeuR9PWkYotv/5bRuj0fEi3n+2fz30cL6ZwttqKQ9IkaTPuCeJX2I1epC0gcApG9/v4uIZ0g9uJdIgbZuEfE0acjhPEmb9LN48XgZsNdQ0pskXSrpkXwM/TevPH56qnY81aJSvUdQ+3E1Or+GfdXl9RSOu1zm8Sr72Ij07b5nvTbrqzGF7V8QEXuQvuB9Evi6pL1J76PVgMWF99JPSD2KV9WzRx368kbgg93bzdt+B688Rvv73t+Y9IVqfmGbV+b0QeUA0wBJbyMdwH/puS4inoqIL0bElsD7gCMl7d69ukf2D5OGoPYgdfHHdu+ihmo8BGzVy7p9eryx14yIv9ew3Z561vlh0puj2xty2oCKNKx0BPA9SWvl5KdJb6BuPa8KugrYSNKOpEBzYd7WM6TzFP8+AFVbhZW9oprkK5nGk87HwcC+hmcCdwPjImJd0pDsQFxMUOl265XqvYJXBr+BsJj0IQpAHuLcsErex0i9qJ716vexHhH/jIhfAbcB25PeR8+Tzn10v4/WjYjtKtUz77eot+P1IVIPpvgeXSciTquhqtXe+4+RvnxsV9jmepEuXhhUDjB1kLSupP1Iwx3nR8TtFfLsJ2nrfMXQk6SeQ/e30UdJ48TdRpEO4MdJB2J/xvYvBV4n6fOS1pA0StIued2PgVMlvTHXaWNJU/qx7d5cBHw5b3Mj0nDG+QO07VeIiDmkD7VpOekW4F2S3iBpPdJQQTH/CtJ5lm+TxrXnFFYfDXwsj7VvCCBpB0m9XhEmaU9Jb829zXWB75B6RHf1VX9Ja0t6N2nI9Abg8rxqIF/DUaTjrCtf0v2pOrfT0z9Ivb7i8XoR8AVJW0gaSTpef5lf94E0E9hP0jskrU4adq74mZV7eheTjvdR+Zg/khpfT0kfk/TeXHYVSfuQLp65PiIWk760/E9+768iaav8PyXv93OS2iStT7qwpugW4EBJq0nqeY7mfOB9kvbOx9aa+dLu3s41dbsA2EPSf0gaIWlDSTtGxEuk83Wnd/ewJW2We2ODygGmf34v6SnSN4cTSB8y1X4/MY40nttF+tb8o4joyOu+QfpgWS7pKNKVL/9H+rZ1J+m8Tk0i4ilgT1Iv6RHgXtLJV4DvkU56X5XrfR2wS6Xt1OEU0gnr24DbSRcmnDJA267k28DRktbIAeeXed/zSUG2pwtJPcJfFT/4IuJvwHvytEjSUuAsVn7oVzOa9MH6BGlYYmtgckQ810uZH+TX/VHSFWK/zmVeyusH8jU8itQTfor04TIgv+HIvb5Tgb/m43VXYDrpashrSOfHngM+28tmXl/hdzB99iIjYgHpiqwLSb2EZaQLZ6r5LKm3sIg0qnBhrmstniT1+h4knSP6FvCpiOgenTiENAR3Z67HTFYOY/2UdN7kVtL/8Dc9tv0VUk9jGenCigsLbXyINHpxPCmYPwR8iRo+m3Pvfl/SBUVLSYFsh7z6GNIQ/XV5yPQP9G+YeUAonwAyM7MBIGksKfCuVkKvrqW4B2NmZqVwgDHLlG4V0nMop0vSFUNdN7NW5CEyMzMrhXswZmZWisG4aWFT2WijjWLs2LF1lX366adZZ511BrZCQ8xtan7DrT3gNrWKYpvmz5//WET068ear7kAM3bsWObNm1dX2Y6ODiZOnDiwFRpiblPzG27tAbepVRTbJKk/dygAPERmZmYlcYAxM7NSOMCYmVkpHGDMzKwUDjBmZlYKBxgzMyuFA4yZmZXCAcbMzErhAGNmZqVwgLGmJNU/zZ/fWHkzGxgOMGZmVgoHGDMzK4UDjJmZlcIBxszMSuEAY2ZmpXCAMTOzUpQWYCRNl7RE0h0V1h0lKSRtlJcl6QxJCyXdJmmnQt6pku7N09RC+nhJt+cyZ0i+wNTMrJmU2YM5B5jcM1HS5sCewIOF5H2AcXmaBpyZ824AnATsAuwMnCRp/VzmzJy3u9yr9mVmZkOntAATEdcASyusOh04GohC2hTgvEiuA0ZL2hTYG5gTEUsjYhkwB5ic160bEddGRADnAQeU1RYzM+u/EYO5M0n7A3+PiFt7jGhtBjxUWO7Mab2ld1ZIr7bfaaTeDmPGjKGjo6Ou+nd1ddVdtlk1a5va2+sv29bWRXt7R93lm+3laNb/USPcptbQaJsGLcBIWhs4Adir0uoKaVFHekURcRZwFsCECRNi4sSJfVW3oo6ODuot26yatU2TJtVftr29g6OOmlh3+ah6JA2NZv0fNcJtag2NtmkwryLbCtgCuFXSA0AbcJOk15F6IJsX8rYBD/eR3lYh3czMmsSgBZiIuD0iNomIsRExlhQkdoqIR4BZwCH5arJdgSciYjEwG9hL0vr55P5ewOy87ilJu+arxw4BLhmstpiZWd/KvEz5IuBaYBtJnZIO6yX75cAiYCHwU+DTABGxFDgZuDFPX89pAJ8CfpbL3AdcUUY7zMysPqWdg4mIg/pYP7YwH8DhVfJNB6ZXSJ8HbN9YLc3MrCz+Jb+ZmZXCAcbMzErhAGNmZqVwgDEzs1I4wJiZWSkcYMzMrBQOMGZmVgoHGDMzK4UDjJmZlcIBxszMSuEAY2ZmpXCAMTOzUjjAmJlZKRxgzMysFA4wZmZWCgcYMzMrhQOMmZmVwgHGzMxK4QBjZmalKC3ASJouaYmkOwpp35Z0t6TbJP1W0ujCuuMkLZR0j6S9C+mTc9pCSccW0reQdL2keyX9UtLqZbXFzMz6r8wezDnA5B5pc4DtI+ItwP8CxwFI2hY4ENgul/mRpFUlrQr8ENgH2BY4KOcF+CZwekSMA5YBh5XYFjMz66fSAkxEXAMs7ZF2VUSsyIvXAW15fgowIyKej4j7gYXAznlaGBGLIuIFYAYwRZKA9wAzc/lzgQPKaouZmfXfUJ6D+U/gijy/GfBQYV1nTquWviGwvBCsutPNzKxJjBiKnUo6AVgBXNCdVCFbUDkARi/5q+1vGjANYMyYMXR0dPSnui/r6uqqu2yzatY2tbfXX7atrYv29o66yzfby9Gs/6NGuE2todE2DXqAkTQV2A/YPSK6g0InsHkhWxvwcJ6vlP4YMFrSiNyLKeZ/lYg4CzgLYMKECTFx4sS66t7R0UG9ZZtVs7Zp0qT6y7a3d3DUURPrLh9Vv6oMjWb9HzXCbWoNjbZpUIfIJE0GjgH2j4hnCqtmAQdKWkPSFsA44AbgRmBcvmJsddKFALNyYJoLfCCXnwpcMljtMDOzvpV5mfJFwLXANpI6JR0G/AAYBcyRdIukHwNExALgYuBO4Erg8Ih4MfdOPgPMBu4CLs55IQWqIyUtJJ2TObustpiZWf+VNkQWEQdVSK4aBCLiVODUCumXA5dXSF9EusrMzMyakH/Jb2ZmpXCAMTOzUvQZYCQdIWldJWdLuknSXoNROTMza1219GD+MyKeBPYCNgYOBU4rtVZmZtbyagkw3T9q3Bf4eUTcSuUfOpqZmb2slgAzX9JVpAAzW9Io4KVyq2VmZq2ulsuUDwN2BBZFxDOSNiQNk5mZmVVVSw8mSLfK/1xeXgdYs7QamZnZsFBLgPkRsBvQ/cPJp0jPaDEzM6uqliGyXSJiJ0k3A0TEMj890szM+lJLD+af+cmSASBpY3yS38zM+lBLgDkD+C2wiaRTgb8A/11qrczMrOX1OUQWERdImg/sTvr9ywERcVfpNTMzs5bWZ4CRtCuwICJ+mJdHSdolIq4vvXZmZtayahkiOxPoKiw/ndPMzMyqqulWMYVHGxMRLzEEj1o2M7PWUkuAWSTpc5JWy9MRwKKyK2ZmZq2tlgDzSeD/AX8HOoFdgGllVsrMzFpfLVeRLQEOHIS6mJnZMFLLA8c2lnS8pLMkTe+eaig3XdISSXcU0jaQNEfSvfnv+jldks6QtFDSbZJ2KpSZmvPfK2lqIX28pNtzmTMk+RECZmZNpJYhskuA9YA/AJcVpr6cA0zukXYscHVEjAOuzssA+wDj8jSNfJWapA2Ak0jDcjsDJ3UHpZxnWqFcz32ZmdkQquVqsLUj4pj+bjgirpE0tkfyFGBinj8X6ACOyenn5avVrpM0WtKmOe+ciFgKIGkOMFlSB7BuRFyb088DDgCu6G89zcysHLX0YC6VtO8A7W9MRCwGyH83yembAQ8V8nXmtN7SOyukm5lZk6ilB3MEcLykF4AXSLeLiYhYdwDrUen8SdSRXnnj0jTylW9jxoyho6OjjipCV1dX3WWbVbO2qb29/rJtbV20t3fUXb7ZXo5m/R81wm1qDY22qZaryEbVvfVXe1TSphGxOA+BLcnpncDmhXxtwMM5fWKP9I6c3lYhf0URcRZwFsCECRNi4sSJ1bL2qqOjg3rLNqtmbdOkSfWXbW/v4KijJtZdPqp+VRkazfo/aoTb1BoabVMtV5FJ0kclfSUvby5p5zr3NwvovhJsKukCgu70Q/K+dgWeyENos4G9JK2fT+7vBczO656StGu+euyQwrbMzKwJ9OeJlh/Oy13U8ERLSRcB1wLbSOqUdBhwGrCnpHuBPfMywOWkuwMsBH4KfBogn9w/GbgxT1/vPuEPfAr4WS5zHz7Bby1OqjzNn1993UBNZmUo7YmWEXFQlVW7V8gbwOFVtjMdeNXvbiJiHrB9X/UwM7Oh4SdamplZKfxESzMzK4WfaGlmZqWo5YmWbwCeAX5fTIuIB8usmJmZtbZaTvJfxsofN64JbAHcA2xXYr3MzKzF1TJE9q/F5Xyn40+UViMzMxsWajnJ/woRcRPwthLqYmZmw0gt52COLCyuAuwE/KO0GpmZ2bBQyzmY4r3IVpDOyfy6nOqYmdlwUcs5mK8NRkXMzGx4qWWIbFZv6yNi/4GrjpmZDRe1DJHdD7wOOD8vHwQ8QLrTsZmZWUW1BJi3RsS7Csu/l3RNRBxfVqXMzKz11XKZ8saStuxekLQFsHF5VTIzs+Gglh7MF4AOSYvy8lj8Q0szM+tDLVeRXSlpHPDmnHR3RDxfbrXMzKzV1fLI5LWBLwGfiYhbgTdI2q/0mpmZWUur5RzMz4EXSI9NBugETimtRmZmNizUEmC2iohvAf8EiIhnSXdWNjMzq6qWAPOCpLVY+cjkrYCGzsFI+oKkBZLukHSRpDUlbSHpekn3SvqlpNVz3jXy8sK8fmxhO8fl9Hsk7d1InczMbGDVEmBOAq4ENpd0AXA1cHS9O5S0GfA5YEJEbA+sChwIfBM4PSLGAcuAw3KRw4BlEbE1cHrOh6Rtc7ntgMnAjyStWm+9zMxsYPUZYCJiDvB+4GPARaTA0NHgfkcAa0kaAawNLAbeA8zM688FDsjzU/Iyef3ukpTTZ0TE8xFxP7AQ2LnBepmZ2QCp5SqytwPPRcRlwGjgeElvrHeHEfF3oB14kBRYngDmA8sjYkXO1glsluc3Ax7KZVfk/BsW0yuUMTOzIaaI6D2DdBuwA/AW4DxgOvD+iHh3XTuU1ifd7v9DwHLgV3n5pDwMhqTNgcsj4l8lLQD2jojOvO4+Uk/l68C1EXF+Tj87l3nVowQkTQOmAYwZM2b8jBkz6qk6XV1djBw5sq6yzapZ2zR/fv1l29q66Oysv03jx9e/70ZUa3Oj7anFYLe5WY+7Rgz3Nk2aNGl+REzo1wYiotcJuCn/PRE4rJhWzwR8EDi7sHwIcCbwGDAip+0GzM7zs4Hd8vyInE/AccBxhe28nK+3afz48VGvuXPn1l22WTVrm6D+qb19bkPlm63NjbanGdvcrMddI4Z7m4B50c/P+1pO8j8l6Tjgo8Bl+UT6av2KYq/0ILCrpLXzuZTdgTuBucAHcp6pwCV5flZeJq//Y27sLODAfJXZFsA44IYG6mVmZgOolgDzIdJlyYdFxCOk8xzfrneHEXE96WT9TcDtuQ5nAccAR0paSDrHcnYucjawYU4/Ejg2b2cBcDEpOF0JHB4RL9ZbLzMzG1i13IvsEeA7heUHSedi6hYRJ5Eufy5aRIWrwCLiOdKwWqXtnAqc2khdzMysHLX0YMzMzPrNAcbMzEpRNcBIujr//ebgVcfMzIaL3s7BbCrp3cD+kmbQ4waXEXFTqTUzM7OW1luAOZF0xVYbhZP8WZBu7WJmZlZR1QATETOBmZK+EhEnD2KdzMxsGKjlMuWTJe0PvCsndUTEpeVWy8zMWl0tN7v8BnAE6QeNdwJH5DQzM7Oq+uzBAO8FdoyIlwAknQvcTLoXmJmZWUW1/g5mdGF+vTIqYmZmw0stPZhvADdLmku6VPlduPdiZmZ9qOUk/0WSOoC3kQLMMfn+ZGZmZlXV0oMhIhaTbo9vZmZWE9+LzMzMSuEAY2Zmpeg1wEhaRdIdg1UZMzMbPnoNMPm3L7dKesMg1cfMzIaJWk7ybwoskHQD8HR3YkTsX1qtzMys5dUSYL5Wei3MzGzY6fMkf0T8CXgAWC3P3wg09CwYSaMlzZR0t6S7JO0maQNJcyTdm/+un/NK0hmSFkq6TdJOhe1MzfnvlTS1kTqZmdnAquVmlx8HZgI/yUmbAb9rcL/fA66MiDcDOwB3kZ49c3VEjAOuzssA+wDj8jQNODPXawPgJGAXYGfgpO6gZGZmQ6+Wy5QPB94OPAkQEfcCm9S7Q0nrkm43c3be3gsRsRyYApybs50LHJDnpwDnRXIdMFrSpsDewJyIWBoRy4A5wOR662VmZgOrlgDzfES80L0gaQTpiZb12hL4B/BzSTdL+pmkdYAx+Y4B3XcO6A5imwEPFcp35rRq6WZm1gRqOcn/J0nHA2tJ2hP4NPD7Bve5E/DZiLhe0vdYORxWiSqkRS/pr96ANI00vMaYMWPo6OjoV4W7dXV11V22WTVrm9rb6y/b1tZFe3tH3eWH6uWo1uZG21OLwW5zsx53jXCbKoiIXidSL+fjwK9I52I+Dqivcr1s73XAA4XldwKXAfcAm+a0TYF78vxPgIMK+e/J6w8CflJIf0W+atP48eOjXnPnzq27bLNq1jZB/VN7+9yGyjdbmxttTzO2uVmPu0YM9zYB86Kfn/e1XEX2EumcyMmkS5bPzTurN6A9AjwkaZuctDvpSZmzgO4rwaYCl+T5WcAh+WqyXYEnIg2hzQb2krR+Prm/V04zM7Mm0OcQmaT3Aj8G7iMNS20h6RMRcUUD+/0scIGk1YFFwKGkntLFkg4DHgQ+mPNeDuwLLASeyXmJiKWSTiZdNg3w9YhY2kCdzMxsANVyDuZ/gEkRsRBA0lakIa26A0xE3AJMqLBq9wp5g3QlW6XtTAem11sPMzMrTy1XkS3pDi7ZImBJSfUxM7NhomoPRtL78+wCSZcDF5Ou0vogK4elzMzMKuptiOx9hflHgXfn+X8A/sW8mZn1qmqAiYhDB7MiZmY2vNRyFdkWpKu+xhbzh2/Xb2ZmvajlKrLfke4b9nvgpXKrY2Zmw0UtAea5iDij9JqYmdmwUkuA+Z6kk4CrgOe7EyOioWfCmJnZ8FZLgPlX4GDgPawcIou8bGZmVlEtAebfgC2jcMt+MzOzvtTyS/5bgdFlV8TMzIaXWnowY4C7Jd3IK8/B+DJlMzOrqpYAc1LptTAzs2GnzwATEX8ajIqYmdnwUssv+Z9i5aOIVwdWA56OiHXLrJiZmbW2Wnowo4rLkg4Adi6tRmZmNizUchXZK0TE7/BvYMzMrA+1DJG9v7C4CulJlFElu5mZGVDbVWTF58KsAB4AppRSGzMzGzZqOQfj58KYmVm/9fbI5BN7KRcRcXIjO5a0KjAP+HtE7JefOzMD2AC4CTg4Il6QtAZwHjAeeBz4UEQ8kLdxHHAY8CLwuYiY3UidzMxs4PR2kv/pChOkD/RjBmDfRwB3FZa/CZweEeOAZXk/3ftbFhFbA6fnfEjaFjgQ2A6YDPwoBy0zM2sCVQNMRPxP9wScBawFHErqZWzZyE4ltQHvBX6Wl0W6Mm1mznIucECen5KXyet3z/mnADMi4vmIuB9YiC+fNjNrGoqofkGYpA2AI4GPkD7kvxcRyxreqTQT+AYwCjgK+BhwXe6lIGlz4IqI2F7SHcDkiOjM6+4DdgG+msucn9PPzmVm9tgdkqYB0wDGjBkzfsaMGXXVu6uri5EjR9ZVtlk1a5vmz6+/bFtbF52d9bdp/Pj6992Iam1utD21GOw2N+tx14jh3qZJkybNj4gJ/dpARFScgG8D95GGw0ZWy9ffCdgP+FGenwhcCmwMLCzk2Ry4Pc8vANoK6+4DNgR+CHy0kH428O997X/8+PFRr7lz59Zdtlk1a5ug/qm9fW5D5ZutzY22pxnb3KzHXSOGe5uAedHPz/veriL7IunuyV8GTkijUgAoxaW6bxXzdmB/SfsCawLrAt8FRksaERErgDbg4Zy/MwecTkkjgPWApYX0bsUyZmY2xHo7B7NKRKwVEaMiYt3CNKqB4EJEHBcRbRExlnSS/o8R8RFgLvCBnG0qcEmen5WXyev/mKPpLOBASWvkK9DGATfUWy8zMxtYtfzQcrAcA8yQdApwM2nIi/z3F5IWknouBwJExAJJFwN3kn4AenhEvDj41TYzs0qGNMBERAfQkecXUeEqsIh4DvhglfKnAqeWV0MzM6tXv292aWZmVgsHGDMzK4UDjJmZlcIBxszMSuEAY2ZmpXCAMTOzUjjAmJlZKRxgzMysFA4wZmZWCgcYMzMrhQOMmZmVwgHGzMxK4QBjZmalcIAxM7NSOMCYmVkpHGDMzKwUDjBmZlYKBxgzMyuFA4yZmZVi0AOMpM0lzZV0l6QFko7I6RtImiPp3vx3/ZwuSWdIWijpNkk7FbY1Nee/V9LUwW6LmZlVNxQ9mBXAFyPiX4BdgcMlbQscC1wdEeOAq/MywD7AuDxNA86EFJCAk4BdgJ2Bk7qDkpmZDb1BDzARsTgibsoOY71AAAAKRElEQVTzTwF3AZsBU4Bzc7ZzgQPy/BTgvEiuA0ZL2hTYG5gTEUsjYhkwB5g8iE0xM7NeKCKGbufSWOAaYHvgwYgYXVi3LCLWl3QpcFpE/CWnXw0cA0wE1oyIU3L6V4BnI6K9wn6mkXo/jBkzZvyMGTPqqm9XVxcjR46sq2yzatY2zZ9ff9m2ti46O+tv0/jx9e+7EdXa3Gh7ajHYbW7W464Rw71NkyZNmh8RE/q1gYgYkgkYCcwH3p+Xl/dYvyz/vQx4RyH9amA88CXgy4X0r5CG3nrd7/jx46Nec+fOrbtss2rWNkH9U3v73IbKN1ubG21PM7a5WY+7Rgz3NgHzop+f80NyFZmk1YBfAxdExG9y8qN56Iv8d0lO7wQ2LxRvAx7uJd3MzJrAUFxFJuBs4K6I+E5h1Syg+0qwqcAlhfRD8tVkuwJPRMRiYDawl6T188n9vXKamZk1gRFDsM+3AwcDt0u6JacdD5wGXCzpMOBB4IN53eXAvsBC4BngUICIWCrpZODGnO/rEbF0cJpgZmZ9GfQAE+lkvaqs3r1C/gAOr7Kt6cD0gaudmZkNFP+S38zMSuEAY2ZmpXCAMTOzUjjAmJlZKRxgzMysFA4wZmZWCgcYMzMrhQOMmZmVwgHGzJAGd5o/P/214W0obhVj1tT8wWc2MNyDMTOzUjjAmJlZKRxgzMysFA4wZmZWCgcYMzMrhQOMmZmVwgHGzMxK4QBjZmalcIDph+5fHw/2ZGbWilo+wEiaLOkeSQslHTvU9TGz2g3FFzZ/aRs8LR1gJK0K/BDYB9gWOEjStkNbKzMzgxYPMMDOwMKIWBQRLwAzgClDXKcBNxg3HWy2ycxanyJiqOtQN0kfACZHxH/l5YOBXSLiMz3yTQOm5cVtgHvq3OVGwGN1lm1WblPzG27tAbepVRTb9MaI2Lg/hVv9bsqVvuu+KmJGxFnAWQ3vTJoXERMa3U4zcZua33BrD7hNraLRNrX6EFknsHlhuQ14eIjqYmZmBa0eYG4ExknaQtLqwIHArCGuk5mZ0eJDZBGxQtJngNnAqsD0iFhQ4i4bHmZrQm5T8xtu7QG3qVU01KaWPslvZmbNq9WHyMzMrEk5wJiZWSkcYGrQqrejkTRd0hJJdxTSNpA0R9K9+e/6OV2SzshtvE3STkNX8+okbS5prqS7JC2QdEROb9l2SVpT0g2Sbs1t+lpO30LS9blNv8wXsiBpjby8MK8fO5T1r0bSqpJulnRpXm719jwg6XZJt0ial9Na9rgDkDRa0kxJd+f31G4D2SYHmD60+O1ozgEm90g7Frg6IsYBV+dlSO0bl6dpwJmDVMf+WgF8MSL+BdgVODz/P1q5Xc8D74mIHYAdgcmSdgW+CZye27QMOCznPwxYFhFbA6fnfM3oCOCuwnKrtwdgUkTsWPhtSCsfdwDfA66MiDcDO5D+XwPXpojw1MsE7AbMLiwfBxw31PXqR/3HAncUlu8BNs3zmwL35PmfAAdVytfME3AJsOdwaRewNnATsAvpF9QjcvrLxyHpqsnd8vyInE9DXfce7WjLH07vAS4l/Si6ZduT6/YAsFGPtJY97oB1gft7vtYD2Sb3YPq2GfBQYbkzp7WqMRGxGCD/3SSnt1w781DKW4HrafF25eGkW4AlwBzgPmB5RKzIWYr1frlNef0TwIaDW+M+fRc4GngpL29Ia7cH0l1CrpI0P99+Clr7uNsS+Afw8zyU+TNJ6zCAbXKA6VtNt6MZBlqqnZJGAr8GPh8RT/aWtUJa07UrIl6MiB1J3/x3Bv6lUrb8t6nbJGk/YElEzC8mV8jaEu0peHtE7EQaKjpc0rt6ydsKbRoB7AScGRFvBZ5m5XBYJf1ukwNM34bb7WgelbQpQP67JKe3TDslrUYKLhdExG9ycsu3CyAilgMdpPNLoyV1/xi6WO+X25TXrwcsHdya9urtwP6SHiDd4fw9pB5Nq7YHgIh4OP9dAvyW9EWglY+7TqAzIq7PyzNJAWfA2uQA07fhdjuaWcDUPD+VdA6jO/2QfKXIrsAT3d3kZiJJwNnAXRHxncKqlm2XpI0ljc7zawF7kE62zgU+kLP1bFN3Wz8A/DHyoHgziIjjIqItIsaS3i9/jIiP0KLtAZC0jqRR3fPAXsAdtPBxFxGPAA9J2iYn7Q7cyUC2aahPNLXCBOwL/C9pXPyEoa5PP+p9EbAY+Cfp28dhpLHtq4F7898Ncl6Rrpa7D7gdmDDU9a/SpneQuuW3Abfkad9WbhfwFuDm3KY7gBNz+pbADcBC4FfAGjl9zby8MK/fcqjb0EvbJgKXtnp7ct1vzdOC7s+BVj7ucj13BOblY+93wPoD2SbfKsbMzErhITIzMyuFA4yZmZXCAcbMzErhAGNmZqVwgDEzs1I4wNhriqQX891wF+S7Fx8paZW8boKkM3opO1bShwevti/vd8Nc51skPSLp74Xl1WvcRmfhTsC3SDq9n3X4i6Qd62uBvVa19COTzerwbKRbsiBpE+BC0i/HT4qIeaTfBFQzFvhwLjNoIuJx0u8VkPRVoCsi2uvY1Dsj3SnAbFC4B2OvWZFu+TEN+Ez+dfJErXx2ybsL3/Zvzr/iPg14Z077Qu7R/FnSTXn6f7nsREkdhedsXJDvQICkt0n6W+493SBpVL7R5bcl3Zifs/GJ/rRD0tGS7sjTZ/tZ9i+STst1uafQhrUl/SrXZwbpx5Bm/eIejL2mRcSiPES2SY9VRwGHR8Rf8401nyPdCPCoiNgP0ocwsGdEPCdpHOnOCd3PCXkrsB3pXk1/Bd4u6Qbgl8CHIuJGSesCz5LusPBERLxN0hrAXyVdFRH391V/STsDHyHdF2tV4AZJf4qI2ypk/7OkF/P89IjoHg5UROwsaX/gRNIzhD5DekbLWyS9ld57dmYVOcCYVb5L7F+B70i6APhNRHTmTkjRasAP8rmJF4E3FdbdEBGdAEq34R9Lug394oi4ESDyXaAl7QW8RVL3fbrWIz3Uqc8AA7wT+HVEPJO39TvS7XQqBZhqQ2TdNwydn+sJ8C7gW7meN0taUENdzF7BAcZe0yRtSQoOSyjcIj8iTpN0Gek+Z9dJ2qNC8S8Aj5KeBLgKqZfT7fnC/Iuk95qofHtzAZ+NiNn1NKGOMj1117W7nt18HylriM/B2GuWpI2BHwM/iB435ZO0VUTcHhHfJA0PvRl4ChhVyLYeqUfyEnAwaYiqN3cDr5f0tryPUUq3p58NfErpMQRIelO+Y28trgH+TdJaeShvCvDnGsv2td2P5PrsQBruM+sX92DstWatPGS1GrAC+AXwnQr5Pi9pEulb/Z3AFaSnM66QdCtwDvAj4NeSPki6Ff3Tve04Il6Q9CHg+0q35X+WdGv+n5GGpm7KFwP8AziglsZExA2SLiI9VgLSw6Nur5K9eA7m5og4tJdN/wA4V9JtpEc4+xyM9ZvvpmxmZqXwEJmZmZXCAcbMzErhAGNmZqVwgDEzs1I4wJiZWSkcYMzMrBQOMGZmVor/DxJtqafG0iXjAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "print (\"\\nMean distance from RuvC_3 to end of protein: %.0f std: %.0f\" % (ruvC3Dists.mean(), ruvC3Dists.std()))\n",
    "\n",
    "# the histogram of the data\n",
    "n, bins, patches = plt.hist(ruvC3Dists,facecolor='blue')\n",
    "plt.xlabel('Distance To End')\n",
    "plt.ylabel('Number of sequences')\n",
    "plt.title(\"Distance From RuvC_3 Domain to End of Sequence\")\n",
    "plt.grid(True)\n",
    "plt.savefig(\"images/%s_RuvC3_DomainLengths.png\" % gene)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f, (ax1, ax2) = plt.subplots(1, 2,figsize=(12, 4))\n",
    "ax1.hist(ruvC1Coords,facecolor='green')\n",
    "ax1.grid(True)\n",
    "ax2.hist(ruvC3Coords,facecolor='blue') \n",
    "plt.grid(True)\n",
    "plt.show();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table align='left'>\n",
       "    <tr style='background-color:#373D20;color:white'><td>No outlier domains</td><td>27001</td></tr><tr style='background-color:#BCBD8B;color:white'><td>Number of RuvC1 outliers</td><td>24</td></tr><tr style='background-color:#717744;color:white'><td>Number of RuvC3 outliers</td><td>1</td></tr><tr style='background-color:#766153;color:white'><td>Outlier Intersection</td><td>0</td></tr>\n",
       "    </table>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<table align='left'>\n",
      "    <tr style='background-color:#373D20;color:white'><td>No outlier domains</td><td>27001</td></tr><tr style='background-color:#BCBD8B;color:white'><td>Number of RuvC1 outliers</td><td>24</td></tr><tr style='background-color:#717744;color:white'><td>Number of RuvC3 outliers</td><td>1</td></tr><tr style='background-color:#766153;color:white'><td>Outlier Intersection</td><td>0</td></tr>\n",
      "    </table>\n",
      "\n",
      "\n",
      "Dumping pickles/Cas9_CorrectDomains.p\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "start_outliers = set(ruvC1Dists[ruvC1Dists > ruvC1Dists.mean()+8*ruvC1Dists.std()].index) #RuvC1 outlier\n",
    "end_outliers = set(ruvC3Dists[ruvC3Dists > ruvC3Dists.mean()+5*ruvC3Dists.std()].index) #RuvC3 outlier\n",
    "hasGoodDomains = hasAllDomains.difference(start_outliers.union(end_outliers))\n",
    "htmlString =\"\"\"<table align='left'>\n",
    "    \"\"\"\\\n",
    "    \"<tr style='background-color:#373D20;color:white'><td>No outlier domains</td><td>%i</td></tr>\" % (len(hasGoodDomains)) +\\\n",
    "    \"<tr style='background-color:#BCBD8B;color:white'><td>Number of RuvC1 outliers</td><td>%i</td></tr>\" % (len(start_outliers)) +\\\n",
    "    \"<tr style='background-color:#717744;color:white'><td>Number of RuvC3 outliers</td><td>%i</td></tr>\" % (len(end_outliers)) +\\\n",
    "    \"<tr style='background-color:#766153;color:white'><td>Outlier Intersection</td><td>%i</td></tr>\" % (len(start_outliers.intersection(end_outliers))) +\\\n",
    "    \"\"\"\n",
    "    </table>\n",
    "\"\"\"\n",
    "\n",
    "#\"<tr style='background-color:#766153;color:white'><td>Outlier Intersection</td><td>%i</td></tr>\" % (len(start_outliers.intersection(end_outliers))) +\\\n",
    "# htmlString += \"<br/><img src='nwks/%s.png' style='width:1000px;height:1000px;'>\" % (\"Cas9-Subtree1.V2\")\n",
    "display(HTML(htmlString))  \n",
    "print(htmlString)\n",
    "dump(hasGoodDomains,\"pickles/%s_CorrectDomains.p\" % gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove chrs that have more than 1 Cas9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27001\n",
      "26857\n"
     ]
    }
   ],
   "source": [
    "#Checked domains, removed outliers, now remove chrs that have more than 1 Cas9\n",
    "hasGoodDomains = load(open(\"pickles/%s_CorrectDomains.p\" % gene,\"rb\"))\n",
    "corrected = {}\n",
    "remove,baseMap = set(),{}\n",
    "print(len(hasGoodDomains))\n",
    "for orfName in hasGoodDomains:\n",
    "    baseCHR = orfName[:orfName.rfind(\"_\")]\n",
    "    if baseCHR in corrected:\n",
    "        try: \n",
    "            baseMap[baseCHR].add(orfName)\n",
    "            baseMap[baseCHR].add(corrected[baseCHR])\n",
    "        except: baseMap[baseCHR] =set([corrected[baseCHR],orfName])\n",
    "        remove.add(orfName)\n",
    "        remove.add(corrected[baseCHR])\n",
    "    corrected[baseCHR] = orfName\n",
    "\n",
    "hasGoodDomains = hasGoodDomains.difference(remove)\n",
    "print(len(hasGoodDomains))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cluster the sequences to choose a representative sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    26857  finished       2147  clusters\n"
     ]
    }
   ],
   "source": [
    "%%bash -s $gene\n",
    "cd-hit -i proteins/All_$1-Like-filtered.faa -M 0 -d 0 -c .90 -sc 1 -o proteins/$1-Like-clustered.faa >logs/$1_ClusterLog.log\n",
    "tail -n 8 logs/$1_ClusterLog.log > logs/clusterInfo\n",
    "head -n 1 logs/clusterInfo; rm logs/clusterInfo\n",
    "mv proteins/$1-Like-clustered.faa.clstr clusters/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved 26857 sequences to proteins/All_Cas9-Like-filtered.faa\n",
      "\n",
      "Dumping pickles/Cas9_Operons.p\n"
     ]
    }
   ],
   "source": [
    "casOperons = load(open(\"pickles/%s_Operons.p\" % gene,\"rb\"))\n",
    "casOperons.getRepSeqs(hasGoodDomains,\"proteins/All_%s-Like-filtered.faa\" %(gene),\"proteins/All_%s-Like.faa\" % (gene)) #TODO add clean to remove non good domains\n",
    "dump(casOperons, \"pickles/%s_Operons.p\" % gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create the file with just the chromosomes related to the representative proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getRepSeqs(hasGoodDomains,repSeqsFile,allCasRepsFile):\n",
    "    repSeqs = open(repSeqsFile,'w')\n",
    "    counter = 0\n",
    "    for rec in parse(allCasRepsFile,\"fasta\"):\n",
    "        if rec.id in hasGoodDomains: \n",
    "            counter +=1\n",
    "            write(rec,repSeqs,\"fasta\")\n",
    "    repSeqs.close()\n",
    "    print(\"Saved %i sequences to %s\" % (len(hasGoodDomains),repSeqsFile))\n",
    "    print(counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved 2147 sequences to assemblies/Cas9_Representative_Assemblies.fasta\n",
      "2147\n"
     ]
    }
   ],
   "source": [
    "casRelatedProteins = dict(fasta_index(\"proteins/%s-Like-clustered.faa\"%gene,'fasta'))\n",
    "getRepSeqs(casRelatedProteins.keys(),\"assemblies/%s_Representative_Assemblies.fasta\" % gene, \"assemblies/All_Cas9_Representative_Assemblies_copy.fasta\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Align the sequences and create a phylogenetic tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 24126726\n",
      "srun -A shadeash-colej mafft --nomemsave --thread 20 --maxiterate 1000 --reorder --treeout proteins/Cas9-Like-clustered.faa > alignments/Cas9-Like-clustered.aln\n",
      " \n",
      "scontrol show job $SLURM_JOB_ID\n"
     ]
    }
   ],
   "source": [
    "%%bash -s $gene\n",
    "sbatch ../scripts/hpc/Alignment.sb\n",
    "tail -n 3 ../scripts/hpc/Alignment.sb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading rep seqs\n",
      "Making reps\n",
      "\n",
      "Dumping pickles/Cas9_Operons.p\n"
     ]
    }
   ],
   "source": [
    "from types import MethodType\n",
    "repFile = \"assemblies/%s_Representative_Assemblies.fasta\" % (gene) #File already exists\n",
    "allRepAssemblies = \"assemblies/All_%s_Representative_Assemblies.fasta\" % (gene) #The new file we are creating\n",
    "print(\"Reading rep seqs\")\n",
    "# allNukSeqs = fasta_index(allRepAssemblies, \"fasta\")\n",
    "print(\"Making reps\")\n",
    "repSeqs = open(repFile,\"w\")\n",
    "casRelatedProteins = fasta_index(\"proteins/%s-Like-clustered.faa\" % gene, \"fasta\")\n",
    "for protID,rec in casRelatedProteins.items():\n",
    "    baseID = protID[:protID.rfind(\"_\")]\n",
    "    nuceRec = allNukSeqs[baseID]\n",
    "    nuceRec.id = protID\n",
    "    if not path.exists(\"assemblies/pseudoChromos/%s.fasta\" % (protID)):\n",
    "        with open(\"assemblies/pseudoChromos/%s.fasta\" % (protID),'w') as fh: write(nuceRec,fh,\"fasta\")\n",
    "    write(nuceRec,repSeqs,\"fasta\")\n",
    "    locus = casOperons[protID]\n",
    "    setattr(locus,'seq',nuceRec) #TODO comment this out on re-run\n",
    "    locus.seq = nuceRec\n",
    "repSeqs.close()\n",
    "dump(casOperons,\"pickles/%s_Operons.p\" % gene)    \n",
    "# corrected = set()\n",
    "# for id in seqs: corrected.add(id[:id.rfind(\"_\")])\n",
    "# casOperons.getRepSeqs(corrected,repFile,allRepAssemblies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "137\n"
     ]
    }
   ],
   "source": [
    "allNukSeqs = {}\n",
    "badIDs = {}\n",
    "for rec in parse(allRepAssemblies,\"fasta\"):\n",
    "    if rec.id in allNukSeqs:\n",
    "        badIDs[rec.id] = (allNukSeqs[rec.id].seq.upper() == rec.seq.upper())\n",
    "        continue\n",
    "    allNukSeqs[rec.id] = rec\n",
    "print(len(badIDs)) \n",
    "# sum(badIDs.values())\n",
    "badIDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Dumping pickles/Cas9_ClusteringResults.p\n",
      "\n",
      "Dumping pickles/Cas9_RevClusteringResults.p\n",
      "\n",
      "Dumping pickles/Cas9_ClusterPercs.p\n"
     ]
    }
   ],
   "source": [
    "allClusters, allClusterSeqIDs = {},{}\n",
    "clusterPercs = {}\n",
    "for line in open(\"clusters/%s-Like-clustered.faa.clstr\" % (gene)):\n",
    "    line = line.strip()\n",
    "    if \">Cluster\" in line: clusterID = line.replace(\">\",\"\")\n",
    "    else:\n",
    "        sequenceID = line[line.find(\">\")+1:line.find(\".\")] \n",
    "        seqID = line[line.find(\">\")+1:line.find(\"...\")] \n",
    "        perc = line.split(\" \")[-1]\n",
    "        aa_s = line.split(\"\\t\")[1]\n",
    "        aa_s = int(aa_s.split(\"aa,\")[0])\n",
    "        if \"%\" in perc: clusterPercs[seqID] = (aa_s,float(perc.replace(\"%\",\"\")))\n",
    "        try:allClusters[clusterID].add(seqID)\n",
    "        except:allClusters[clusterID]=set([seqID])\n",
    "        try:allClusterSeqIDs[seqID].add(clusterID)\n",
    "        except:allClusterSeqIDs[seqID]=set([clusterID])\n",
    "allClusters, allClusterSeqIDs\n",
    "dump(allClusters,\"pickles/%s_ClusteringResults.p\" % (gene))\n",
    "dump(allClusterSeqIDs,\"pickles/%s_RevClusteringResults.p\" % (gene))\n",
    "dump(clusterPercs,\"pickles/%s_ClusterPercs.p\" % (gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build the Cas9 Finding Report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Rho import *\n",
    "metadata = {}\n",
    "for line in open(\"tables/AllCas9Meta.txt\"):\n",
    "    seqID,description = line.strip().split(\"\\t\")\n",
    "    metadata[seqID]=description\n",
    "\n",
    "casOperons = load(open(\"pickles/%s_Operons.p\" % gene,\"rb\"))\n",
    "clusters = processClusterFile(\"clusters/%s-Like-clustered.faa.clstr\" % (gene))\n",
    "#seqCounter[seq], ruvC1Dists,ruvC3Dists, distBetween\n",
    "allAssemblies = load(open(\"pickles/All_%s_Representative_Assemblies.p\" % (gene),'rb'))\n",
    "allProts = dict(fasta_index(\"proteins/All_Cas9-Like.faa\",'fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5693 28855\n"
     ]
    }
   ],
   "source": [
    "report = open(\"tables/Cas9FindingReport.tsv\",'w')\n",
    "\n",
    "columns = ['Protein ID',\"Prot Length\",\"Chr Length\",\"Cluster\",\"Cluster Members\",'Systems with idenditcal protein',\n",
    "           \"RuvC I to Start\",\"RuvC III to End\",\"AAs Between I-III\",\"#CRISPR Arrays\",\"Consensus Repeats Sequence(s)\",\n",
    "           \"Consensus Length\",\"#Repeats in CRISPR Search\",\"#BLAST Repeats\",\"Metadata\"]\n",
    "\n",
    "report.write(\"\\t\".join(columns)+'\\n')\n",
    "skips = 0\n",
    "for protID in allAssemblies:\n",
    "    try:\n",
    "        operon = casOperons.operons[casOperons.seqMap[protID]]\n",
    "        crispr = operon.getCRISPR(protID)\n",
    "    \n",
    "        values = {\n",
    "            \"Protein ID\": protID,\n",
    "            \"Prot Length\": len(allProts[protID]),\n",
    "            \"Chr Length\": len(operon.seq.seq),\n",
    "            \"Cluster\": clusters.revMap[protID],\n",
    "            \"Cluster Members\": len(clusters[clusters.revMap[protID]].members),\n",
    "            \"Systems with idenditcal protein\": len(seqCounter[str(allProts[protID].seq).strip().upper()]),\n",
    "            \"RuvC I to Start\":ruvC1Dists[protID],\n",
    "            \"RuvC III to End\":ruvC3Dists[protID],\n",
    "            \"AAs Between I-III\":distBetween[protID],\n",
    "            \"#CRISPR Arrays\": len(crispr.consensusRepeats),\n",
    "            \"Consensus Repeats Sequence(s)\": \",\".join(crispr.consensusRepeats),\n",
    "            \"Consensus Length\": len(list(crispr.consensusRepeats)[0]),\n",
    "            \"#Repeats in CRISPR Search\":len(crispr.spacers),\n",
    "            \"#BLAST Repeats\":len([]), # will be crispr.numBlastRes\n",
    "            \"Metadata\":metadata[protID]\n",
    "        }\n",
    "    except: \n",
    "        skips+=1\n",
    "        continue\n",
    "    \n",
    "    for col in columns: report.write(str(values[col])+\"\\t\")\n",
    "    report.write('\\n')\n",
    "    \n",
    "print(skips,len(allAssemblies))\n",
    "report.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Yay Done!!!!!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO import index\n",
    "gene='Cas9'\n",
    "casRelatedProteins = index(\"proteins/%s-Like-clustered.faa\" % (gene),\"fasta\")\n",
    "fh = open(\"proteins/%s-Like-clustered_copy.faa\" % (gene),'w')\n",
    "for rec in casRelatedProteins.values():\n",
    "    rec.description = ''\n",
    "    write(rec,fh,'fasta')\n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mv proteins/Cas9-Like-clustered_copy.faa proteins/Cas9-Like-clustered.faa #To clean up the descriptions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "goodDomIDS = set()\n",
    "goodDomMap = {}\n",
    "hasGoodDomains = load(open(\"pickles/%s_CorrectDomains.p\" % gene, \"rb\"))\n",
    "for prot in hasGoodDomains: \n",
    "    baseID = prot[:prot.rfind(\"_\")] # remove the orf identifier\n",
    "    goodDomIDS.add(baseID)\n",
    "    try: goodDomMap[baseID].add(prot)\n",
    "    except: goodDomMap[baseID] = set([prot])\n",
    "print(\"Found %i Cas9-like proteins on the same sequence\" % (len(hasGoodDomains) - len(goodDomIDS)))\n",
    "print(\"Total proteins found %i in (%i == %i) sequences\" % (len(hasGoodDomains),len(goodDomMap),len(goodDomIDS)))\n",
    "dump(goodDomIDS,\"pickles/%s_GoodDomainIDS.p\" %(gene))\n",
    "dump(goodDomMap,\"pickles/%s_GoodDomMap.p\" %(gene))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Now we have assemblies with a crispr and a Cas9. Next we will copy all of the unique seqeuences over to the same place. This will allow the systems with the same protein but different DNA sequences to exist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblyDir = \"assemblies/assemblies_W_%s/\" % (gene)\n",
    "cas9Assemblies = listdir(assemblyDir)\n",
    "goodDomIDS = load(open(\"pickles/%s_GoodDomainIDS.p\" % (gene), \"rb\"))\n",
    "goodDomMap = load(open(\"pickles/%s_GoodDomMap.p\"    % (gene), \"rb\"))\n",
    "hmm_parser = load(open(\"pickles/%s_HMM_Parsing_Results.p\" % (gene), \"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assembliesWCrisprs = load(open(\"pickles/allAssemblyW_CRISPRs.p\",\"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getDB(aPath): \n",
    "    if \"genbank\" in aPath: return \"genbank\"\n",
    "    if \"refseq\"  in aPath: return \"refseq\"\n",
    "    return \"pat2\"\n",
    "hmm_parser = load(open(\"pickles/%s_HMM_Parsing_Results.p\" % (gene), \"rb\"))\n",
    "len(hmm_parser)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy unique nucleotide sequence\n",
    "from Bio.SeqIO import index, write\n",
    "from os import system\n",
    "# hmm_parser = load(open(\"pickles/%s_HMM_Parsing_Results.p\" % (gene), \"rb\"))\n",
    "hasGoodDomains = load(open(\"pickles/%s_CorrectDomains.p\" % gene, \"rb\"))\n",
    "# assembliesWCrisprs = load(open(\"pickles/allAssemblyW_CRISPRs.p\",\"rb\"))\n",
    "\n",
    "nukSeqHash, protAssemblyMap, assemblyProtMap = {}, {}, {}\n",
    "# BASEFILENAME eg GCF_0001232342.5.fasta = GCF_0001232342.5\n",
    "# protAssemblyMap = {CHROM_ORF#:BASEFILENAME}\n",
    "\n",
    "counter = 0\n",
    "#casFiltered = open(\"proteins/All_%s-Like-filtered.faa\" % gene, \"w\")\n",
    "crPath = \"/mnt/research/germs/shane/databases/crisprs/pilerCR\"\n",
    "for baseID, proteins in hmm_parser:\n",
    "    assemblyProtMap[baseID] = {}\n",
    "    protIDs = proteins.proteins.keys()\n",
    "    keepers = hasGoodDomains.intersection(protIDs)\n",
    "    if len(keepers) > 0:\n",
    "#         seqs = index(assembliesWCrisprs[baseID], \"fasta\")\n",
    "#         db = getDB(assembliesWCrisprs[baseID])\n",
    "        for protID in keepers:\n",
    "            counter += 1\n",
    "            protAssemblyMap[protID] = baseID\n",
    "            baseProtID = protID[:protID.rfind(\"_\")]\n",
    "            try: assemblyProtMap[baseID][baseProtID].add(protID)\n",
    "            except: assemblyProtMap[baseID] = {baseProtID:set([protID])}\n",
    "#             pchr = seqs[baseProtID]\n",
    "#             nukeSeq = str(pchr.seq).upper()\n",
    "#             if nukeSeq not in nukSeqHash:\n",
    "#                 if not path.exists(\"crisprs/%s.pcrout\" % (protID)):\n",
    "#                     system(\"ln -s %s/%s/%s.pcrout crisprs/%s.pcrout\" % (crPath,db,baseID,protID))\n",
    "#                     with open(\"assemblies/pseudoChromos/%s.fasta\" % protID, \"w\") as seqFH:\n",
    "#                         pchr.id = protID\n",
    "#                         write(pchr, seqFH, \"fasta\")\n",
    "#                 write(proteins.proteins[protID],casFiltered,\"fasta\")\n",
    "#                 nukSeqHash[nukeSeq] = set([protID])\n",
    "#             else: nukSeqHash[nukeSeq].add(protID)\n",
    "            if counter % 1000 == 0: print(counter/1000,end=' ')\n",
    "# casFiltered.close()\n",
    "# dump(nukSeqHash,\"pickles/%s_nukSeqHash.p\" % gene)\n",
    "# dump(protAssemblyMap,\"pickles/%s_protAssemblyMap.p\" % gene)\n",
    "dump(assemblyProtMap,\"pickles/%s_assemblyProMap.p\" % gene)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Write all proteins to a file ->Cluster proteins -> Align the representatives"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = \"Cas9\"\n",
    "clusterID = \"\"\n",
    "allClusters, allClusterSeqIDs = {},{}\n",
    "clusterPercs = {}\n",
    "for line in open(\"clusters/%s-Like-clustered.faa.clstr\" % (gene)):\n",
    "    line = line.strip()\n",
    "    if \">Cluster\" in line: clusterID = line.replace(\">\",\"\")\n",
    "    else:\n",
    "        seqID = line[line.find(\">\")+1:line.find(\"...\")] \n",
    "        perc = line.split(\" \")[-1]\n",
    "        aa_s = line.split(\"\\t\")[1]\n",
    "        aa_s = int(aa_s.split(\"aa,\")[0])\n",
    "        if \"%\" in perc: clusterPercs[seqID] = (aa_s,float(perc.replace(\"%\",\"\")))\n",
    "        try:allClusters[clusterID].add(seqID)\n",
    "        except:allClusters[clusterID]=set([seqID])\n",
    "        try:allClusterSeqIDs[seqID].add(clusterID)\n",
    "        except:allClusterSeqIDs[seqID]=set([clusterID])\n",
    "allClusters, allClusterSeqIDs\n",
    "dump(allClusters,\"pickles/%s_ClusteringResults.p\" % (gene))\n",
    "dump(allClusterSeqIDs,\"pickles/%s_RevClusteringResults.p\" % (gene))\n",
    "dump(clusterPercs,\"pickles/%s_ClusterPercs.p\" % (gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO import index,write\n",
    "from os import listdir\n",
    "gene = \"Cas9\"\n",
    "casAssemblies = index(\"data/proteins/%s-Like-clustered.faa\" % (gene), \"fasta\")\n",
    "casFH = open(\"data/assemblies/%s_Representative_Assemblies.fasta\" % (gene), \"w\")\n",
    "counter = 0\n",
    "for fname in listdir(\"data/assemblies/pseudoChromos/\"):\n",
    "    seqID = fname[:fname.find(\".fasta\")]\n",
    "    if seqID in casAssemblies: \n",
    "        asm = index(\"data/assemblies/pseudoChromos/\"+fname, \"fasta\")\n",
    "        write(asm[seqID],casFH,\"fasta\")\n",
    "        if counter % 100 == 0 : print(counter/100,end=\" \")\n",
    "        counter +=1\n",
    "casFH.close()\n",
    "counter "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dump(nukSeqHash,\"pickles/%s_nukSeqHash.p\" % gene)\n",
    "dump(protAssemblyMap,\"pickles/%s_protAssemblyMap.p\" % gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hmm_parser = HMM_Parser()\n",
    "hmmFiles = os.listdir(hmmResultsDir)\n",
    "counter, numWithCas9Like = 0, 0\n",
    "print (\"Working on the %s directory which contains %i files\" % (hmmResultsDir,len(hmmFiles)))\n",
    "for fileName in hmmFiles:\n",
    "    if \".hmmout\" not in fileName: continue\n",
    "    fileStats = os.stat(os.path.join(hmmResultsDir,fileName)).st_size  \n",
    "    counter+=1\n",
    "    if counter % 100000 == 0: print(counter,numWithCas9Like)\n",
    "    if fileStats < 1900: continue #Less than 1900 means no results were found\n",
    "    hmm_parser.parse(os.path.join(hmmResultsDir,fileName))\n",
    "    numWithCas9Like +=1\n",
    "dump(hmm_parser,\"pickles/%s_HMM_Parsing_Results.p\" % (gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "protCounter, fileCounter = 0,1\n",
    "protSeqs = set()\n",
    "IDs, dupes = {},{}\n",
    "casAll = open(\"proteins/All_%s-Like.faa\" % gene,\"w\")\n",
    "casUnique = open(\"proteins/Unique_%ss.faa\" % gene,\"w\")\n",
    "# fh = open(\"../scripts/hpc/hmmSearch/HMMSearch_0.sb\",\"w\")\n",
    "# header = \"\"\n",
    "# for line in open(\"../scripts/hpc/header.sb\"): header+=line\n",
    "# fh.write(header)\n",
    "\n",
    "for baseID,proteins in hmm_parser:\n",
    "    for protID, rec in proteins:\n",
    "        if len(rec.seq) > 2100: continue\n",
    "        if rec.id in IDs: \n",
    "            try: dupes[rec.id].add(baseID)\n",
    "            except: \n",
    "                dupes[rec.id] = set([baseID])\n",
    "                dupes[rec.id].add(IDs[rec.id])\n",
    "            continue\n",
    "        IDs[rec.id]=baseID\n",
    "        seq = str(rec.seq).upper()\n",
    "        protCounter +=1\n",
    "        if seq not in protSeqs:\n",
    "            \n",
    "            protSeqs.add(seq)\n",
    "            write(rec,casUnique,\"fasta\")\n",
    "        write(rec,casAll,\"fasta\")\n",
    "casUnique.close() \n",
    "casAll.close()\n",
    "print(\"Number of dupes: \",len(dupes))\n",
    "print (\"Writing %i proteins from %i assemblies to file: proteins/All_%s-Like.faa\" % (protCounter,numWithCas9Like,gene))\n",
    "print(\"Of the %s proteins, there are %i unique proteins\" % (protCounter,len(protSeqs)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO import index, write\n",
    "seqs = index(\"proteins/All_%s-Like.faa\" % (gene),\"fasta\")\n",
    "fh = open(\"proteins/multiple%ss.faa\" % (gene),\"w\")\n",
    "for baseID, orfList in baseMap.items():\n",
    "    orf1, orf2 = list(orfList)\n",
    "    seq1 = str(seqs[orf1].seq).upper()\n",
    "    seq2 = str(seqs[orf2].seq).upper()\n",
    "    if seq1 != seq2:\n",
    "        print(orf1,orf1)\n",
    "        write(seqs[orf1],fh,\"fasta\")\n",
    "        write(seqs[orf2],fh,\"fasta\")\n",
    "fh.close()       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO import index,write\n",
    "prots = index(\"proteins/All_%s-Like.faa\" % (gene), \"fasta\")\n",
    "fh = open(\"proteins/%s-Like.faa\" % (gene),'w')\n",
    "for fName in listdir(\"assemblies/pseudoChromos/\"):\n",
    "    orfID = fName.replace(\".fasta\",\"\")\n",
    "    prot = prots[orfID]\n",
    "    write(prot,fh,\"fasta\")\n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO import index\n",
    "fh = open(\"assemblies/%s_Representative_Assemblies.fasta\" % (gene),\"w\")\n",
    "prots = index(\"proteins/%s-Like-clustered.faa\" % (gene), \"fasta\" )\n",
    "for protName in prots.keys():\n",
    "    protAsm = index(\"assemblies/pseudoChromos/%s.fasta\" % (protName), \"fasta\")\n",
    "#     seqID = protName[:protName.rfind(\"_\")]\n",
    "#     rec = protAsm[seqID]\n",
    "#     rec.id = protName\n",
    "    write(protAsm[protName],fh,\"fasta\")\n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assembliesWCrisprs = load(open(\"pickles/allAssemblyCRISPRs.p\",\"rb\")) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import path, system\n",
    "#fh = open(\"MissingCRISPS.txt\",\"w\")\n",
    "#hmm_parser = load(open(\"pickles/%s_HMM_Parsing_Results.p\"  % (gene), \"rb\"))\n",
    "counter = 0    \n",
    "crisprDir = \"/mnt/research/germs/shane/databases/crisprs/pilerCR/\"  \n",
    "pcmd = \"pilercr -minid 0.85 -mincons 0.8 -noinfo -in %s -out /mnt/research/germs/shane/databases/crisprs/pilerCR/%s/%s.pcrout 2>/dev/null\" # % (assembly, outdir, assembly.replace(\".fasta\",\"\"))\n",
    "for fName, result in hmm_parser:\n",
    "    keepers = prots.intersection(result.proteins.keys())\n",
    "    for protID in keepers:\n",
    "        db = \"genbank\"\n",
    "        if \"refseq\" in assembliesWCrisprs[fName]: db = \"refseq\"\n",
    "        if \"PATRIC2\" in assembliesWCrisprs[fName]: db = \"pat2\"\n",
    "#         if not path.exists(\"crisprs/%s.pcrout\" % protID):\n",
    "#             system(pcmd % (assembliesWCrisprs[fName],db,fName))\n",
    "#             fh.write(protID + \"\\t\" + assembliesWCrisprs[fName] + \"\\n\")\n",
    "        system(\"unlink crisprs/%s.pcrout\" % protID)\n",
    "        system(\"ln -s /mnt/research/germs/shane/databases/crisprs/pilerCR/%s/%s.pcrout crisprs/%s.pcrout\" % (db,fName,protID))\n",
    "        counter +=1\n",
    "counter\n",
    "#fh.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Substitutions Substitute any assemblies were no blast results are found because the area repeat is too small and any that don't have a candidate repeat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hmm_parser = load(open(\"pickles/hmm_parsing_results.p\",\"rb\"))\n",
    "# assembliesWCrisprs = load(open(\"pickles/allAssemblyCRISPRs.p\",\"rb\")) \n",
    "from os import system,path\n",
    "counter = 0\n",
    "protMapper = {}\n",
    "for result in hmm_parser.results:\n",
    "    fullHmmPath = result.filePath\n",
    "    baseID = fullHmmPath.replace(\"hmm/results/\",\"\").replace(\".hmmout\",\"\")\n",
    "    if path.exists(\"assemblies/assemblies_W_Cas9s/%s.fasta\" % (baseID)):\n",
    "        counter+=1\n",
    "        if counter % 1000 ==0: print(counter,end=\" \")\n",
    "        continue\n",
    "    for rec in parse(assembliesWCrisprs[baseID],\"fasta\"):\n",
    "        if rec.id in goodDomIDS:\n",
    "            system(\"ln -s %s assemblies/assemblies_W_Cas9s/%s.fasta\" % (assembliesWCrisprs[baseID],baseID))\n",
    "            counter+=1\n",
    "            if counter % 1000 ==0: print(counter,end=\" \")\n",
    "            break\n",
    "\n",
    "#     for prot in result.proteins:\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO import index as fasta_index\n",
    "allCas9s = fasta_index(\"Cas9-Like-clustered.faa\", \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob, os\n",
    "outliers = start_outlierIDs.union(end_outlierIDs)\n",
    "files = glob.glob(\"assemblies/*.fasta\")\n",
    "for fileName in files:\n",
    "    fileName2 = fileName.replace(\"assemblies/\",\"\").replace(\".fasta\",\"\")\n",
    "#     if fileName2 in outliers: \n",
    "#         os.system(\"mv %s* assemblies/NoDomains/\" % (fileName.replace(\".fasta\",\"\")))\n",
    "        \n",
    "removeSeqs = set(hits.samples).difference(hasDomains)\n",
    "\n",
    "print len(hits.samples), len(removeSeqs), len(hits.samples) - len(removeSeqs)\n",
    "files = glob.glob(\"assemblies/*.fasta\")\n",
    "for fileName in files:\n",
    "    fileName2 = fileName.replace(\"assemblies/\",\"\").replace(\".fasta\",\"\")\n",
    "#     if fileName2 not in hits.samples: \n",
    "#         os.system(\"mv %s* assemblies/NoDomains/\" % (fileName.replace(\".fasta\",\"\")))\n",
    "print \"Total Samples:\",len(hits.samples)\n",
    "print \"\\tMissing Domain Samples:\",len(removeSeqs)\n",
    "print \"\\tDomain profile Samples:\",len(outliers)\n",
    "print \"After filtering:\", len(hits.samples) - (len(removeSeqs) + len(outliers))\n",
    "\n",
    "counter=0\n",
    "fh = open(\"Cas9-Like-clustered2.faa\",\"w\")\n",
    "from Bio.SeqIO import parse\n",
    "for rec in parse(\"Cas9-Like-clustered.faa\",\"fasta\"):\n",
    "    if os.path.exists(\"assemblies/%s.fa\" % (rec.id)):\n",
    "        counter += 1\n",
    "        fh.write(\">%s\\n%s\\n\" % (rec.id, str(rec.seq)))\n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cluster the results to remove duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash -s $gene\n",
    "cd-hit -i proteins/All_$1-Like.faa -c .9999999 -d 0 -M 0 -T 0 -o proteins/$1-Like-clustered.faa >logs/$1_ClusterLog.log\n",
    "tail -n 8 logs/$1_ClusterLog.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO import parse\n",
    "fh = open(\"CasRelatedAssemblies2.fasta\",\"w\")\n",
    "counter = 0\n",
    "for rec in parse(\"CasRelatedAssemblies.fasta\",\"fasta\"):\n",
    "    if os.path.exists(\"assemblies/%s.fa\" % (rec.id)):\n",
    "        counter += 1\n",
    "        fh.write(\">%s\\n%s\\n\" % (rec.id, str(rec.seq)))\n",
    "fh.close()\n",
    "print counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "casOperons = load(open(\"pickles/casOperonDataStructureW%s.p\" % gene,\"rb\"))\n",
    "protID = \"NZ_LT906463.1_ORF95365\"\n",
    "protID = \"FNMX01000011_ORF6660\"\n",
    "genomicAsmName = casOperons.revMap[protID] \n",
    "protOperon = casOperons.casOperons[genomicAsmName]\n",
    "protCRISPR = protOperon.crisprs[protID]\n",
    "\n",
    "protOperon.assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "files1 = set(glob.glob(\"assemblies/*.fasta\"))\n",
    "files2 = set(glob.glob(\"assemblies/*.fa\"))\n",
    "fastas,fas =set(),set()\n",
    "for f in files2: fas.add(f.replace(\".fa\",\"\"))\n",
    "for f in files1: fastas.add(f.replace(\".fasta\",\"\"))\n",
    "for f in fas.difference(fastas):\n",
    "    os.system(\"rm %s*\" % (f))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Remove missing domains and length filtering\n",
    "# Domain database pfam\n",
    "# wget ftp://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam31.0/Pfam-A.full.gz\n",
    "dom_counter = {}\n",
    "for line in open(\"/mnt/research/germs/shane/transActRNA/data/domains/Cas9-Like-clustered.faa.domtbl\"):\n",
    "    if \"#\" in line:continue\n",
    "    rec = line.strip().split(\"    \")\n",
    "#     print rec\n",
    "#     break\n",
    "    try:dom_counter[line.strip()]+=1\n",
    "    except:dom_counter[line.strip()]=1\n",
    "dom_counter = Counter(dom_counter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SearchIO import HmmerIO\n",
    "from Bio.SearchIO.HmmerIO import Hmmer3DomtabHmmhitParser as hmmParser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AACPSW010000030.1_ORF701.fasta\n",
      "alignments\n",
      "annotations\n",
      "assemblies\n",
      "blastDBs\n",
      "blastout\n",
      "Cas9SystemDetails.txt\n",
      "clusters\n",
      "conseqs0\n",
      "conseqs1\n",
      "conseqs10\n",
      "conseqs2\n",
      "conseqs3\n",
      "conseqs4\n",
      "conseqs5\n",
      "FindTracrsLog.txt\n",
      "GCA_005325615.1.hmmout\n",
      "hmm\n",
      "images\n",
      "InfernalResults.py\n",
      "LaunchStructureSearch.sh\n",
      "logs\n",
      "old\n",
      "pickles\n",
      "proteins\n",
      "__pycache__\n",
      "sequences\n",
      "tables\n",
      "test.txt\n",
      "tmp\n",
      "tracrRNA_SearchResults.txt\n",
      "tracrRNA_SearchResults.xlsx\n",
      "trees\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Program: hmmsearch (3.2.1)\n",
      "  Query: Cas9 (1039)\n",
      "         <unknown description>\n",
      " Target: sequences/orfs/GCA_005325615.1.orfs\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      2  AACPSW010000030.1_ORF701  length 991 aa, 2976 bp, from ...\n"
     ]
    }
   ],
   "source": [
    "res = HmmerIO.Hmmer3TextParser(open(\"GCA_005325615.1.hmmout\"))\n",
    "count = 0\n",
    "for val in res:\n",
    "#     print(val.split('\\n'))\n",
    "\n",
    "    print(val)\n",
    "    count+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "hit = val.hits[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hit(id='AACPSW010000030.1_ORF701', query_id='Cas9', 2 hsps)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__class__',\n",
       " '__delattr__',\n",
       " '__dict__',\n",
       " '__dir__',\n",
       " '__doc__',\n",
       " '__eq__',\n",
       " '__format__',\n",
       " '__ge__',\n",
       " '__getattribute__',\n",
       " '__gt__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__init_subclass__',\n",
       " '__iter__',\n",
       " '__le__',\n",
       " '__lt__',\n",
       " '__module__',\n",
       " '__ne__',\n",
       " '__new__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__subclasshook__',\n",
       " '__weakref__',\n",
       " '_parse_qresult',\n",
       " '_parse_row',\n",
       " 'handle',\n",
       " 'line']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'handle': <_io.TextIOWrapper name='GCA_005325615.1.hmmout' mode='r' encoding='UTF-8'>,\n",
       " 'line': '\\n'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res.__dict__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hasHNH, hasRuvC= set(), set()\n",
    "for x in hmmParser(open(\"/mnt/research/germs/shane/transActRNA/data/domains/DiverseCas9s1.domtbl\")):\n",
    "    if \"HNH\" == x.id:\n",
    "        print x.id, len(x.hits)\n",
    "        for hit in x.hits: hasHNH.add(hit.id)\n",
    "    elif \"RuvC\" == x.id:\n",
    "        print x.id, len(x.hits)\n",
    "        for hit in x.hits: hasRuvC.add(hit.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Copy unique nucleotide sequence\n",
    "from Bio.SeqIO import index\n",
    "nukSeqHash,protSeqHash = set(), set()\n",
    "alreadyGotIt,count = 0, 0\n",
    "for assembly in cas9Assemblies:\n",
    "    baseID = assembly[:-6]\n",
    "    allAssemblySeqs = index(assemblyDir+assembly,\"fasta\")\n",
    "    overlap = goodDomIDS.intersection(allAssemblySeqs.keys())\n",
    "    for recID in overlap:\n",
    "        seq = str(allAssemblySeqs[recID].seq).upper()\n",
    "        if seq in nukSeqHash and len(goodDomMap[recID]) == 1:\n",
    "            alreadyGotIt += 1\n",
    "            continue\n",
    "        nukSeqHash.add(seq)\n",
    "        #There may be more than 1 protein on the pseudochromosome, save both as separate files  \n",
    "        if len(goodDomMap[recID])>1: print(\"%i Cas9s on %s\" %(len(goodDomMap[recID]),recID))\n",
    "        for orfID in goodDomMap[recID]:\n",
    "            with open(\"assemblies/pseudoChromos/%s.fasta\" % (orfID),\"w\") as fh:\n",
    "                write(rec,fh,\"fasta\")\n",
    "                count +=1\n",
    "        if count % 1000 == 0: print(count,end=\" \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print \"Number of proteins with both:\", len(hasHNH.intersection(hasRuvC))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in hmmParser(open(\"/mnt/research/germs/shane/CRISPR_Cas/data/domains/DiverseCas9s1.domtbl\")):\n",
    "    if \"ruvc\" in x.id.lower():\n",
    "        print x.id, len(x.hits)\n",
    "        for hit in x.hits:\n",
    "            print \"\\t\",hit.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "definitions, descriptors, AC = {}, set(['AC','DE']), \"\"\n",
    "important = set([\"Cas9\",\"nuclease\",\"endonuclease\", \"HNH\", \"HnH\", \"RuvC\", \"ruvc\", \"RUVC\", \"HNH/ENDO\", \"CRISPR\"])\n",
    "for line in open(\"domains/Pfam-A.full\"):\n",
    "    rec = line.strip().split()\n",
    "    if len(rec)<3 or rec[1] not in descriptors: continue\n",
    "    if rec[1] == 'AC':AC = rec[2]\n",
    "    else: definitions[AC] = \" \".join(rec[2:])\n",
    "print \"Completed Loading %i definitions\" % len(definitions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hnhDomainIDs = set()\n",
    "for val in dom_counter.most_common(100):\n",
    "    id,count = val[0], val[1]\n",
    "    print \"%s\\t%i\\t%s\" % (id,count,definitions[id])\n",
    "#     if \"hnh\" in definitions[id].lower():\n",
    "#         hnhDomainIDs.add(id)\n",
    "#         print \"%s\\t%i\\t%s\" % (id,count,definitions[id])\n",
    "#     if \"ruvc\" in definitions[id].lower():\n",
    "#         print \"%s\\t%i\\t%s\" % (id,count,definitions[id])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def outliers(data,stdsAwayFromMean): return data[data > stdsAwayFromMean * data.std()]\n",
    "start_outliers = outliers(ruvC1Dists,2) \n",
    "len(start_outliers)\n",
    "\n",
    "from Bio.SeqIO import index\n",
    "\n",
    "cas9Like = index(\"proteins/All_%s-Like.faa\" % gene,\"fasta\")\n",
    "allLengths = {}\n",
    "for id,rec in cas9Like.items(): allLengths[id]=len(rec.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO import parse\n",
    "ids = set()\n",
    "for rec in parse(\"%s-Like-clustered.faa\" % gene,\"fasta\"): ids.add(rec.id[:-5])\n",
    "db_path = \"/mnt/research/germs/shane/databases/assemblies/NCBI/refseq\"\n",
    "for id in ids:\n",
    "    os.system(\"cp %s/%s assemblies/\" % (db_path,assembly_ids[id]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "from Bio.SeqIO import parse\n",
    "db_path = \"/mnt/research/germs/shane/databases/assemblies/NCBI/refseq\"\n",
    "files = os.listdir(db_path)\n",
    "assembly_ids = {}\n",
    "for fileName in files:\n",
    "    for rec in parse(os.path.join(db_path,fileName),\"fasta\"): assembly_ids[rec.id]=fileName\n",
    "print len(assembly_ids)\n",
    "pickle.dump(assembly_ids,open(\"/mnt/research/germs/shane/databases/assemblies/NCBI/refseq_manifest.p\",\"wb\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compare alignment methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO import index\n",
    "divCas9 = {} #index(geneProfile,\"fasta\")\n",
    "gapCount, baseCount = 0,0\n",
    "for rec in parse(\"alignments/Cas9.aln\",\"fasta\"): #default (more gaps less strict)\n",
    "    for base in rec.seq:\n",
    "        gapCount += int(base == \"-\")\n",
    "        baseCount += int(base != \"-\")\n",
    "    divCas9[rec.id]={'original':[gapCount,baseCount]}   \n",
    "gapCount, baseCount = 0,0\n",
    "for rec in parse(\"alignments/CCas9s.aln\",\"fasta\"): #local (Less gaps more strict)\n",
    "    for base in rec.seq:\n",
    "        gapCount += int(base == \"-\")\n",
    "        baseCount += int(base != \"-\")\n",
    "    divCas9[rec.id][\"local\"] = [gapCount,baseCount]\n",
    "    \n",
    "for id in divCas9:\n",
    "    print(\"%i/%i\\t%i/%i\\t%s\" % (divCas9[id][\"local\"][1],divCas9[id][\"original\"][1],divCas9[id][\"local\"][0],divCas9[id][\"original\"][0],id))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO import parse\n",
    "import os\n",
    "import pickle\n",
    "print \"Loading ncbi IDs\"\n",
    "assembly_ids = pickle.load(open(\"/mnt/research/germs/shane/databases/assemblies/NCBI/refseq_manifest.p\",\"rb\"))\n",
    "print \"Loading patric2 IDs\"\n",
    "notInNCBI, ids,fullID = [], [], {}\n",
    "for rec in parse(\"%s-Like-clustered.faa\" % gene,\"fasta\"): \n",
    "    fullID[rec.id[:-5]] = rec.id\n",
    "    ids.append(rec.id[:-5])\n",
    "for id in ids:\n",
    "    if id not in assembly_ids:\n",
    "        notInNCBI.append(id)\n",
    "        continue\n",
    "    os.system(\"cp /mnt/research/germs/shane/databases/assemblies/NCBI/refseq/%s /mnt/research/germs/shane/CRISPR_Cas/data/assemblies/%s.fasta\" % (assembly_ids[id],fullID[id]))\n",
    "    print id,assembly_ids[id]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print \"Loading PATRIC IDs\"\n",
    "assembly_ids = pickle.load(open(\"/mnt/research/germs/shane/databases/assemblies/PATRIC2/pat2_manifest.p\",\"rb\"))\n",
    "print \"Copying Files\"\n",
    "for id in notInNCBI:\n",
    "    os.system(\"cp /mnt/research/germs/shane/databases/assemblies/PATRIC2/fastas/%s /mnt/research/germs/shane/CRISPR_Cas/data/assemblies/%s.fasta\" % (assembly_ids[id],fullID[id]))\n",
    "    print id,assembly_ids[id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#assembly_ids = pickle.load(open(\"/mnt/research/germs/shane/databases/assemblies/NCBI/refseq_manifest.p\",\"rb\"))\n",
    "#assembly_ids = pickle.load(open(\"/mnt/research/germs/shane/databases/assemblies/PATRIC2/pat2_manifest.p\",\"rb\"))\n",
    "files = os.listdir(\"assemblies\")\n",
    "ids = set()\n",
    "for fileName in files:ids.add(fileName.replace(\".fasta\",\"\"))\n",
    "for rec in parse(\"%s-Like-clustered.faa\" % gene,\"fasta\"): \n",
    "    if rec.id not in ids:\n",
    "        print rec.id,assembly_ids[rec.id[:-5]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Teasing out on the same chr but uniq vs duplication event (ie found orfs)\n",
    "from Bio.SeqIO import index as fasta_index\n",
    "seqs = fasta_index(\"proteins/All_%s-Like.faa\" % (gene),\"fasta\")\n",
    "prots = {}\n",
    "print(len(remove))\n",
    "for orfID in remove: \n",
    "    seq = str(seqs[orfID].seq) \n",
    "    try: prots[seq].add(orfID)\n",
    "    except: prots[seq] = set([orfID])\n",
    "print(len(prots))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd scripts\n",
    "bash pilercrRun.sh &\n",
    "bash pilercrRunFW.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ete3 import Tree, TreeStyle\n",
    "# casTree=Tree(\"Cas9-Like-clustered.faa.tree\")\n",
    "# ts = TreeStyle()\n",
    "# ts.mode = \"c\"\n",
    "# ts.show_leaf_name = False\n",
    "# casTree.render(\"Cas9.png\",tree_style=ts)\n",
    "#t.write(format=1, outfile=\"new_tree.nw\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ete3 import Tree, TreeStyle\n",
    "rooted_tree = Tree( \"((A,B),(C,D));\" )\n",
    "ts = TreeStyle()\n",
    "ts.mode = \"c\"\n",
    "rooted_tree.render(\"%%inline\",tree_style=ts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandas import DataFrame, read_csv, Series\n",
    "outfmt6_cols = [\"qseqid\",\"sseqid\",\"pident\",\"length\",\"mismatch\",\"gapopen\",\"qstart\",\"qend\",\"sstart\",\"send\",\"evalue\",\"bitscore\"]\n",
    "blastResults = read_csv(\"data/DiverseCompare.blastout\",delimiter='\\t', names=outfmt6_cols) \n",
    "len(blastResults)\n",
    "blastResults\n",
    "len(blastResults[blastResults.pident>90])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(blastResults[blastResults.pident>98])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Move large number of hmm files to an old run dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import system\n",
    "files = os.listdir(\"hmm/results\")\n",
    "print(len(files))\n",
    "for fname in files:\n",
    "    mv_CMD = \"mv hmm/results/%s hmm/results/oldRuns/\"\n",
    "    if fname[-4:] == \"orfs\": system(\"rm hmm/results/\"+fname)\n",
    "    elif (fname[-4:] == \".out\"): system(mv_CMD % (fname))\n",
    "    else:\n",
    "        print(fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    if len(proteins.proteins) == 0:\n",
    "        proteins.__get_proteins__(hmmResultsDir+baseID+\".hmmout\")\n",
    "        nncount += 1\n",
    "    if len(proteins.proteins) == 0:\n",
    "        fh.write(getOrfsCMD % (assembliesWCrisprs[baseID],hmmResultsDir, baseID) + \"\\n\")\n",
    "        count+=1\n",
    "        if count % 50 == 0:\n",
    "            fh.close()\n",
    "            fh = open(\"../scripts/hpc/hmmSearch/HMMSearch_%i.sb\" % (fileCounter),\"w\")\n",
    "            fileCounter += 1\n",
    "            fh.write(header)\n",
    "            \n",
    "print(count,fileCounter,nncount)\n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hits = load(open(\"pickles/%s_HMM_DOMAIN_Search_Results.p\" % (gene),\"rb\"))\n",
    "\n",
    "for id in ['NZ_ATLK01000001.1', 'JDTS01000001','SEHM01000008.1', 'NZ_BBNY01000081.1', 'NZ_BBNR01000042.1',\n",
    "           'LOLQ01000035.1', 'NZ_KQ973595.1', 'FJCJ01000024.1', 'NZ_CP017601.1', 'NZ_QFKO01000012.1','ANHY01000014.1']:\n",
    "    for key in hits.samples.keys():\n",
    "        if id in key:  print (key, end=\"','\")\n",
    "\n",
    "for domName, dHit in hits.samples['ANHY01000014.1_ORF1375'].items():\n",
    "    print(domName,dHit.start,dHit.dend)\n",
    "\n",
    "Bifidobacterium_bombi {'NZ_ATLK01000001.1', 'JDTS01000001'}\n",
    "JejTia_pallidilutea {'SEHM01000008.1', 'NZ_BBNY01000081.1', 'NZ_BBNR01000042.1'}\n",
    "\n",
    "print(\"\\t\".join(['stds','numRemoved','meanLen','InHits','ANHY01000014.1_ORF1375']))\n",
    "for i in range(2,10):\n",
    "    newset = ruvC1Dists[ruvC1Dists > ruvC1Dists.mean()+i*ruvC1Dists.std()].index\n",
    "    print(\"\\t\".join([str(i),str(len(newset)), \"%.2f\" % (ruvC1Dists.mean()+i*ruvC1Dists.std()), str('ANHY01000014.1_ORF1375' in newset), str('JDTS01000001_ORF24522' in hasAllDomains)]))\n",
    "\n",
    "print(\"\\t\".join(['stds','numRemoved','meanLen','InHits','ANHY01000014.1_ORF1375']))\n",
    "for i in range(0,10): \n",
    "    newset = set(ruvC3Dists[ruvC3Dists > ruvC3Dists.mean()+i*ruvC3Dists.std()].index)\n",
    "    print(\"\\t\".join([str(i),str(len(newset)), \"%.2f\" % (ruvC3Dists.mean()+i*ruvC3Dists.std()), str('ANHY01000014.1_ORF1375' in newset), str('JDTS01000001_ORF24522' in hasAllDomains)]))\n",
    "    \n",
    "\n",
    "ruvC3Dists.describe()\n",
    "\n",
    "inBadDomains = set()\n",
    "for id in ['NZ_ATLK01000001.1_ORF30490','JDTS01000001_ORF24522','SEHM01000008.1_ORF5944','NZ_BBNY01000081.1_ORF323','NZ_BBNR01000042.1_ORF248','LOLQ01000035.1_ORF225','NZ_KQ973595.1_ORF2960','FJCJ01000024.1_ORF1896','NZ_CP017601.1_ORF13118','NZ_QFKO01000012.1_ORF2328','ANHY01000014.1_ORF1375']:\n",
    "    if (id in noRuvC1 or id in noRuvC2 or id in noRuvC3 or id in noHNH): inBadDomains.add(id)\n",
    "    print(id,id in noRuvC1, id in noRuvC2, id in noRuvC3, id in noHNH, id in start_outliers, id in end_outliers) #   = allSamples.difference(hits[\"HNH_Cas\"])\n",
    "inBadDomains"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
