{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The Known tracrRNA Universe\n",
    "#### From perviosuly published tracrRNAs, what are the corresponding BLAST and structural genomes in our cas9 Universe?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h5 style=\"color:blue\">Libraries and workspace setup</h5>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "import os, sys\n",
    "sys.path.append(\"scripts/\")\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.SeqIO import index as fasta_index\n",
    "from Bio.SeqIO import parse, write\n",
    "from collections import Counter\n",
    "from easyFunctions import *\n",
    "from HMMParser import *\n",
    "from InfernalResults import ProcessInfernal, TRACR_RNA\n",
    "from pandas import Series\n",
    "from random import randint\n",
    "\n",
    "r = lambda: randint(0,255)            \n",
    "def color():return '#%02X%02X%02X' % (r(),r(),r())\n",
    "def findBiggestCluster(seqID, clusters, clusterMap): return Counter(dict((k, clusterMap[k]) for k in (clusters))).most_common(1)[0][0] \n",
    "os.chdir(\"/mnt/research/germs/shane/transActRNA/data\")\n",
    "gene = \"Cas9\"\n",
    "# if os.path.exists(\"pickles/Step2.db\"): \n",
    "#     dill.load_session('pickles/Step2.db')\n",
    "#     print \"Loaded Session\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h5>Data</h5>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Raw\n",
    "knownTracrFile = \"sequences/PioneerCrisprs.fasta\"\n",
    "casRepAssembliesFile = \"assemblies/%s_Representative_Assemblies.fasta\" % (gene)\n",
    "allCasRepsFile = \"assemblies/All_%s_Representative_Assemblies.fasta\" % (gene)\n",
    "knownTracrRNAs = fasta_index(knownTracrFile,\"fasta\")\n",
    "casRepProteins = fasta_index(\"proteins/%s-Like-clustered.faa\" % (gene),\"fasta\")\n",
    "\n",
    "#minimum lengths of blast hits\n",
    "tracrLengths = {}\n",
    "for id,rec in knownTracrRNAs.items(): tracrLengths[id] = len(rec.seq)*0.6  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def processBLAST(blastFileName,tracrLengths):\n",
    "    blastResults = NCBIXML.parse(open(blastFileName,'r'))\n",
    "    tracrBLAST = {}\n",
    "    numRepr=set()\n",
    "    numFilteredByLength,hitCount = 0,0\n",
    "    for result in blastResults:\n",
    "        for hit in result.alignments:\n",
    "            for hsp in hit.hsps: \n",
    "                coord = Coordinate(min(hsp.sbjct_start,hsp.sbjct_end),max(hsp.sbjct_start,hsp.sbjct_end))\n",
    "                hitLen = len(coord)\n",
    "                if hitLen < tracrLengths[result.query]:\n",
    "                    numFilteredByLength += 1\n",
    "                    continue # Skip hits that may be sequencing artifacts or partial crRNA hits\n",
    "                try:\n",
    "                    numRepr.add(result.query)\n",
    "                    hitCount += 1\n",
    "                    tracrBLAST[hit.hit_def].append([result.query,coord])\n",
    "                except: tracrBLAST[hit.hit_def]=[[result.query,coord]]\n",
    "    return tracrBLAST, numRepr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4>BLAST the known tracrRNAs against the genomes of cas9 associated assemblies</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 46 of the 48 known tracrs\n"
     ]
    }
   ],
   "source": [
    "allKnownTracrBlastOut = \"blastout/PioneerTracrRNAsAll.blastout\"\n",
    "repsKnownTracrBlastOut = \"blastout/PioneerTracrRNAsReps.blastout\"\n",
    "\n",
    "#blastResults = BLAST_short(knownTracrFile,allCasRepsFile,repsKnownTracrBlastOut)\n",
    "# blastResults = BLAST_short(knownTracrFile,casRepAssembliesFile,repsKnownTracrBlastOut)\n",
    "tracrBLASTReps, numRepr = processBLAST(repsKnownTracrBlastOut, tracrLengths)\n",
    "tracrBLAST, numAll = processBLAST(allKnownTracrBlastOut, tracrLengths)\n",
    "# blastResults = NCBIXML.parse(open(knownTracrBlastOut,'r'))\n",
    "# tracrBLAST, tracrCVG = {},{}\n",
    "# lengthDist = []\n",
    "# numRepr=set()\n",
    "# numFilteredByLength,hitCount = 0,0\n",
    "# for result in blastResults:\n",
    "#     for hit in result.alignments:\n",
    "#         for hsp in hit.hsps: \n",
    "#             coord = Coordinate(min(hsp.sbjct_start,hsp.sbjct_end),max(hsp.sbjct_start,hsp.sbjct_end))\n",
    "#             hitLen = len(coord)\n",
    "#             try: tracrCVG[result.query] = max(tracrCVG[result.query],abs(hsp.sbjct_end-hsp.sbjct_start))\n",
    "#             except: tracrCVG[result.query] = hitLen\n",
    "#             if hitLen < tracrLengths[result.query]:\n",
    "#                 numFilteredByLength += 1\n",
    "#                 continue # Skip hits that may be sequencing artifacts or partial crRNA hits\n",
    "#             try:\n",
    "#                 numRepr.add(result.query)\n",
    "#                 lengthDist.append(hitLen)\n",
    "#                 hitCount += 1\n",
    "#                 tracrBLAST[hit.hit_def].append([result.query,coord])\n",
    "#             except: tracrBLAST[hit.hit_def]=[[result.query,coord]]\n",
    "# print(\"Found and recorded %i / %i hits\\nHits are represented in %i / %i Cas related Assemblies\" % (hitCount,hitCount+numFilteredByLength,len(tracrBLAST),len(casRepProteins)))\n",
    "print(\"Found %i of the %i known tracrs\" % (len(numRepr),len(knownTracrRNAs)))\n",
    "# dump(tracrBLAST,\"pickles/Pioneer_tracrBlast.p\")\n",
    "# dump(lengthDist,\"pickles/tracrLenDist.p\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "all = {}\n",
    "for res, val in tracrBLAST.items():\n",
    "    for id,coord in val:\n",
    "        try: all[id].add(res.split(\" \")[0])\n",
    "        except: all[id] = set([res.split(\" \")[0]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missingKeys = set(all.keys()).difference(numRepr)\n",
    "len(missingKeys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Legionella_pneumophila {'NZ_LOLP01000005.1', 'NZ_RBFS01000066.1', 'CP017457', 'FJCG01000006', 'NZ_KK074140.1', 'QWDX01000001.1', 'CP021260.1', 'JFIG01000037', 'NZ_QFHM01000015.1', 'CP021277', 'BBUJ01000003.1', 'FJBH01000013.1', 'AVOW01000039', 'NZ_CP021281.1', 'LOLE01000008', 'NZ_CP021279.1', 'LAXQ01000016.1', 'FJCH01000005', 'NZ_FJCI01000014.1', 'FJBY01000014.1', 'NZ_QWDU01000001.1', 'UGOU01000002.1', 'LOMA01000038', 'LBAA01000002', 'NZ_LAZZ01000018.1', 'NZ_QWDS01000002.1', 'FJLT01000003.1', 'FJCF01000016.1', 'AWES01000030', 'PQWY01000001', 'CP017602', 'CP016030', 'AVNJ01000012', 'FJEA01000015.1', 'RBFP01000056.1', 'AVOV01000026.1', 'NZ_CP021286.1', 'LCUP01000045', 'LBAM01000011.1', 'NZ_CP017458.1', 'FJDY01000015', 'CP021264', 'UGOO01000001.1', 'NZ_QWDW01000001.1', 'NZ_QWDY01000001.1', 'NZ_FJAQ01000015.1', 'LAVQ01000004.1', 'FJDZ01000002', 'NZ_CAFM01000006.1', 'FJBQ01000014', 'FJDU01000015.1', 'CP015927', 'NZ_FJAG01000013.1', 'QWDT01000001.1', 'CP016030.2', 'LOLL01000057.1', 'LBMS01000002.1', 'LBAT01000016', 'RBFC01000046.1', 'CR628336.1', 'CP015929', 'NZ_UGON01000002.1', 'CAAAHZ010000009.1', 'FJDQ01000016', 'LAXV01000018.1', 'FJBJ01000014.1', 'LOLB01000012.1', 'FJCA01000017.1', 'FJAR01000013.1', 'NZ_LOLN01000058.1', 'NZ_CDDZ01000036.1', 'FR687201', 'LT632614.1', 'LOLM01000019.1', 'LAXO01000017.1', 'RBFN01000061.1', 'FJAK01000003.1', 'FJFI01000015', 'NZ_JFIF01000026.1', 'NZ_LAXZ01000017.1', 'NZ_QWDR01000001.1', 'LOLO01000019.1', 'FJLQ01000003', 'LBHM01000008', 'FJLN01000005.1', 'FJDV01000014.1', 'QFJC01000015.1', 'NZ_FJBN01000013.1', 'LNYK01000030', 'PQWX01000004.1', 'FJAN01000015', 'NZ_LBAF01000019.1', 'JFIK01000015.1', 'LBAL01000006', 'CP021256', 'NZ_RBGS01000049.1', 'CP021258.1', 'CP021257', 'NZ_KQ973597.1', 'NZ_FJCE01000019.1', 'CP015931', 'CCAA010000026.1', 'NZ_FJLR01000002.1', 'NZ_RBFA01000044.1', 'NZ_QFKO01000012.1', 'NZ_UGOJ01000002.1', 'LBAS01000011.1', 'NZ_LR134176.1', 'FJLP01000003', 'QFKK01000013.1', 'FJMM01000003.1', 'LOLF01000038', 'NZ_FJFJ01000014.1', 'LBMR01000016', 'NZ_CP017601.1', 'NZ_CP011105.1', 'LAXU01000016', 'FJCJ01000024.1', 'RBFB01000035.1', 'LORH01000038', 'NZ_FJBV01000013.1', 'QFKT01000003.1', 'FJAV01000006', 'FJAP01000013', 'LOLG01000047.1', 'CP021261', 'LCUQ01000056.1', 'FJOE01000099.1', 'LR134332.1', 'NZ_RBFJ01000070.1', 'FJBA01000014', 'LAXW01000004', 'CP021263.1', 'FJCM01000005', 'NZ_KQ973595.1', 'NZ_FJMH01000006.1', 'LOLQ01000035.1', 'FJAZ01000012', 'FJBC01000013', 'NZ_LAXP01000016.1', 'NZ_FJDC01000029.1', 'NZ_FJCR01000004.1', 'QWDV01000001.1', 'QFHL01000003.1', 'FJCS01000004.1', 'NZ_LOLZ01000018.1', 'NZ_FJBF01000014.1', 'FJBL01000014', 'NZ_RBGQ01000056.1'}\n",
      "Bifidobacterium_bombi {'NZ_ATLK01000001.1', 'JDTS01000001'}\n"
     ]
    }
   ],
   "source": [
    "for key in missingKeys:\n",
    "    print(key,all[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">Caenispirillum_salinarum_84\n",
      "GGACCGGCACCTACAGGACAAGTGATCATACACGTGACAGCCGCCTCCCCCGCCCCAGTGGCCAAGGGGAGGCGGCTTTTCTTT\n",
      "\n",
      ">Bifidobacterium_bombi_81\n",
      "ATTGAATTTGAAATAAGGCTGTTCCTTCGTTAGTTCAGTCGATTGCTCCTCCGGTATTGCTTTTGCATGCCGGAGTTTTTT\n",
      "\n",
      ">Corynebacterium_camporealensis_79\n",
      "GGAACTGAACCTCAGTAATCACCGGCTCGTTTCCGGTGAAGAATGCTCCAGTCGGCGCCCACTGGGCGTCGACTTTTCT\n",
      "\n",
      ">Legionella_pneumophila_116\n",
      "TTCGATACCCTGAAATCAACAAAATTAAAGATTGAATCGTTTTTCTATGCTCGTCTTAATAGCGAGCATATAACGATTTCAAAATCTATCGTTACAGTGATTTAGAAACTCACACT\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i,key in enumerate(knownTracrRNAs.keys()):\n",
    "    if key not in tracrCVG: \n",
    "        print(\">\"+key+ \"_%i\" % (len(knownTracrRNAs[key].seq)))\n",
    "        print(str(knownTracrRNAs[key].seq)+\"\\n\") #.reverse_complement()\n",
    "#     print(\"%i\\t%i\\t%i\\t%i\" % ((i,tracrCVG[key],len(knownTracrRNAs[key].seq),len(knownTracrRNAs[key].seq)-tracrCVG[key])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set(knownTracrRNAs.keys()).difference(tracrCVG.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing = {\"Caenispirillum_salinarum\":['GCF_000315795.1','GCA_000315795.1'],\n",
    "           'Bifidobacterium_bombi':['GCF_000737845.1','GCA_000737845.1'],\n",
    "        \n",
    "'Corynebacterium_camporealensis':{'GCA_000766885.2': 'CP027001.1','GCA_000980815.1': 'CP011311.1'},\n",
    "'Legionella_pneumophila':{\n",
    "    'GCA_003004315.1': 'CP021286.1', \n",
    "    'GCA_003004275.1': 'CP021281.1',\n",
    "    'GCA_003003515.1': 'CP021256.1',\n",
    "    'GCA_001752825.1': 'CP015931.1',\n",
    "    'GCA_001752785.1': 'CP015929.1',\n",
    "    'GCA_001752745.1': 'CP015927.1',\n",
    "    'GCA_001677075.2': 'CP016030.2',\n",
    "    'GCA_001610735.1': 'CP011105.1'}}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene = 'Cas9'\n",
    "assembliesWCrisprs = load(open(\"pickles/allAssemblyW_CRISPRs.p\",\"rb\")) \n",
    "hmm_parser = load(open(\"pickles/%s_HMM_Parsing_Results.p\" % (gene), \"rb\"))\n",
    "assemblyProtMap = load(open(\"pickles/%s_assemblyProMap.p\" % (gene),'rb'))\n",
    "allClusterSeqIDs = load(open(\"pickles/%s_RevClusteringResults.p\" % (gene),'rb'))\n",
    "nukSeqHash = load(open(\"pickles/%s_nukSeqHash.p\" % gene,'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['FNUS01000008_ORF11136',\n",
       " 'NZ_KV793811.1_ORF3019',\n",
       " 'QJTD01000003.1_ORF18407',\n",
       " 'NIDL01000002.1_ORF135',\n",
       " 'GG740084.1_ORF30',\n",
       " 'NZ_HE999452.1_ORF30',\n",
       " 'CAJK01000259_ORF30',\n",
       " 'DF970668.1_ORF14579',\n",
       " 'BBZD01000017_ORF79',\n",
       " 'PFHY01000226.1_ORF106']"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(allClusterSeqIDs)[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caenispirillum_salinarum\n",
      "\tGCF_000315795.1 True\n",
      "\tGCA_000315795.1 True\n",
      "Bifidobacterium_bombi\n",
      "\tGCF_000737845.1 True\n",
      "\tGCA_000737845.1 True\n",
      "Corynebacterium_camporealensis\n",
      "\tGCA_000766885.2 True True CP027001.1_ORF23238 CP027001.1 False\n",
      "\tGCA_000980815.1 True True CP011311.1_ORF23235 CP011311.1 False\n",
      "Legionella_pneumophila\n",
      "\tGCA_003004315.1 True True CP021286.1_ORF11884 CP021286.1 False\n",
      "\tGCA_003004275.1 True True CP021281.1_ORF11885 CP021281.1 False\n",
      "\tGCA_003003515.1 True True CP021256.1_ORF11068 CP021256.1 False\n",
      "\tGCA_001752825.1 True True CP015931.1_ORF11289 CP015931.1 False\n",
      "\tGCA_001752785.1 True True CP015929.1_ORF11289 CP015929.1 False\n",
      "\tGCA_001752745.1 True True CP015927.1_ORF11289 CP015927.1 False\n",
      "\tGCA_001677075.2 True True CP016030.2_ORF11885 CP016030.2 False\n",
      "\tGCA_001610735.1 True True CP011105.1_ORF11092 CP011105.1 False\n"
     ]
    }
   ],
   "source": [
    "blastres = {}\n",
    "for tracrName,fList in missing.items():\n",
    "    blastres[tracrName]={}\n",
    "    print(tracrName)\n",
    "    for asm in fList:\n",
    "        if type(fList) == type({}):\n",
    "            #hmm_parser.results[asm].__get_proteins__(\"hmm/results/\"+asm+\".hmmout\")\n",
    "            for prot in hmm_parser.results[asm].proteins: \n",
    "                print(\"\\t\"+asm,asm in assembliesWCrisprs, asm in hmm_parser, prot, fList[asm],prot in allClusterSeqIDs )\n",
    "                seq = str(fasta_index(assembliesWCrisprs[asm],\"fasta\")[fList[asm]].seq).upper()\n",
    "                if seq in nukSeqHash:\n",
    "                    print(\"\\t\\t\",nukSeqHash[seq])\n",
    "        else: print(\"\\t\"+asm,asm in assembliesWCrisprs)\n",
    "        tmp = MakeFasta(\"%s.fasta\"%(asm))\n",
    "        tmp.write(asm,str(knownTracrRNAs[tracrName].seq))\n",
    "        tmp.close()\n",
    "        BLAST_short(\"%s.fasta\"%(asm),assembliesWCrisprs[asm], \"%s.blasout\"%(asm))\n",
    "        blastres[tracrName][asm] = NCBIXML.parse(open(\"%s.blasout\"%(asm),'r'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Caenispirillum_salinarum\n",
      "\t GCF_000315795.1\n",
      "\t\t GCF_000315795.1 NZ_ANHY01000014.1 Caenispirillum salinarum AK4 contig14, whole genome shotgun sequence [48346\t48429]+\n",
      "\t GCA_000315795.1\n",
      "\t\t GCA_000315795.1 ANHY01000014.1 Caenispirillum salinarum AK4 contig14, whole genome shotgun sequence [48346\t48429]+\n",
      "Bifidobacterium_bombi\n",
      "\t GCF_000737845.1\n",
      "\t\t GCF_000737845.1 NZ_ATLK01000001.1 Bifidobacterium bombi DSM 19703 DSM19703_C1, whole genome shotgun sequence [737030\t737110]+\n",
      "\t GCA_000737845.1\n",
      "\t\t GCA_000737845.1 ATLK01000001.1 Bifidobacterium bombi DSM 19703 DSM19703_C1, whole genome shotgun sequence [737030\t737110]+\n",
      "Corynebacterium_camporealensis\n",
      "\t GCA_000766885.2\n",
      "\t\t GCA_000766885.2 CP027001.1 Corynebacterium camporealensis strain CIP 105508 chromosome, complete genome [597198\t597276]+\n",
      "\t GCA_000980815.1\n",
      "\t\t GCA_000980815.1 CP011311.1 Corynebacterium camporealensis strain DSM 44610, complete genome [597241\t597319]+\n",
      "Legionella_pneumophila\n",
      "\t GCA_003004315.1\n",
      "\t\t GCA_003004315.1 CP021286.1 Legionella pneumophila subsp. pneumophila strain Albuquerque 1 (D-7474) chromosome, complete genome [189574\t189689]+\n",
      "\t GCA_003004275.1\n",
      "\t\t GCA_003004275.1 CP021281.1 Legionella pneumophila subsp. pneumophila strain Flint 2 (D-7477) chromosome, complete genome [189578\t189693]+\n",
      "\t GCA_003003515.1\n",
      "\t\t GCA_003003515.1 CP021256.1 Legionella pneumophila strain D-4954 chromosome, complete genome [176850\t176965]+\n",
      "\t GCA_001752825.1\n",
      "\t\t GCA_001752825.1 CP015931.1 Legionella pneumophila strain Philadelphia_4 chromosome, complete genome [180316\t180431]+\n",
      "\t GCA_001752785.1\n",
      "\t\t GCA_001752785.1 CP015929.1 Legionella pneumophila strain Philadelphia_2 chromosome, complete genome [180316\t180431]+\n",
      "\t GCA_001752745.1\n",
      "\t\t GCA_001752745.1 CP015927.1 Legionella pneumophila strain Philadelphia_1_ATCC, complete genome [180316\t180431]+\n",
      "\t GCA_001677075.2\n",
      "\t\t GCA_001677075.2 CP016030.2 Legionella pneumophila strain OLDA, complete genome [189578\t189693]+\n",
      "\t GCA_001610735.1\n",
      "\t\t GCA_001610735.1 CP011105.1 Legionella pneumophila strain L10-023, complete genome [178175\t178290]+\n"
     ]
    }
   ],
   "source": [
    "for tracr in blastres:\n",
    "    print(tracr)\n",
    "    for asm in blastres[tracr]:\n",
    "        print(\"\\t\",asm)\n",
    "        for res in blastres[tracr][asm]:\n",
    "            for hit in res.alignments:\n",
    "                for hsp in hit.hsps: \n",
    "                    coord = Coordinate(min(hsp.sbjct_start,hsp.sbjct_end),max(hsp.sbjct_start,hsp.sbjct_end))\n",
    "                    print(\"\\t\\t\",res.query,hit.hit_def, coord)\n",
    "#             print(\"\\t\\t\",str(parseSingleBLAST(res)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'alignments', 'application', 'blast_cutoff', 'database', 'database_length', 'database_letters', 'database_name', 'database_sequences', 'date', 'descriptions', 'dropoff_1st_pass', 'effective_database_length', 'effective_hsp_length', 'effective_query_length', 'effective_search_space', 'effective_search_space_used', 'expect', 'filter', 'frameshift', 'gap_penalties', 'gap_trigger', 'gap_x_dropoff', 'gap_x_dropoff_final', 'gapped', 'hsps_gapped', 'hsps_no_gap', 'hsps_prelim_gapped', 'hsps_prelim_gapped_attemped', 'ka_params', 'ka_params_gap', 'matrix', 'multiple_alignment', 'num_good_extends', 'num_hits', 'num_letters_in_database', 'num_seqs_better_e', 'num_sequences', 'num_sequences_in_database', 'posted_date', 'query', 'query_id', 'query_length', 'query_letters', 'reference', 'sc_match', 'sc_mismatch', 'threshold', 'version', 'window_size']\n",
      "\n",
      "GCA_001610735.1\n",
      "Score 116 (215 bits), expectation 5.5e-57, alignment length 116\n",
      "Query:       1 TTCGATACCCTGAAATCAACAAAATTAAAGATTGAATCGTTTTTC...ACT 116\n",
      "               |||||||||||||||||||||||||||||||||||||||||||||...|||\n",
      "Sbjct:  178290 TTCGATACCCTGAAATCAACAAAATTAAAGATTGAATCGTTTTTC...ACT 178175\n"
     ]
    }
   ],
   "source": [
    "print(dir(res))\n",
    "print()\n",
    "print(res.query)\n",
    "print(hsp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build the tree with Cas Related Assemblies with a BLAST Hit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Annotated 615 CRISPR-Cas Systems\n",
      "Systems were annotated with a total of 35 cluster groups\n",
      "\n",
      "Dumping pickles/Pioneer_tracrBlast.p\n"
     ]
    }
   ],
   "source": [
    "clusterMap = {} #Hits clustered by where the most common tracrRNA sequence\n",
    "tracrBLASTMap = {}\n",
    "for seqID,clusterList in tracrBLAST.items():\n",
    "    tracrBLASTMap[seqID]=set()\n",
    "    for clusterID in clusterList:\n",
    "        try: clusterMap[clusterID[0]].add(seqID)\n",
    "        except: clusterMap[clusterID[0]] = set([seqID])\n",
    "        tracrBLASTMap[seqID].add(clusterID[0])  \n",
    "\n",
    "blastTreeColors,colorMap = {},{}\n",
    "for seqID,clusters in tracrBLASTMap.items():\n",
    "    biggestCluster = findBiggestCluster(seqID, clusters, clusterMap)\n",
    "    try: blastTreeColors[seqID] = colorMap[biggestCluster]\n",
    "    except: \n",
    "        colorMap[biggestCluster] = color()\n",
    "        blastTreeColors[seqID] = colorMap[biggestCluster]\n",
    "print (\"Annotated %i CRISPR-Cas Systems\" % (len(blastTreeColors)))\n",
    "print (\"Systems were annotated with a total of %i cluster groups\" %(len(colorMap)))\n",
    "\n",
    "dump(blastTreeColors,\"pickles/Pioneer_tracrBlast.p\") #PaperTracrBLAST_HitsTreeColors.p\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Some stats on the blast hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYwAAAEZCAYAAACEkhK6AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAIABJREFUeJzt3XucHFWZ//HPNyTIPdxM4pIMCItcdhWIIlFRhsuiQQVvuIorBNBlFX54XQVXZaLrLqCIoquoixiQu6iJK7sgSwZEuZNIuBrkkoSQQSHhvpDA8/vjnE4qTU/mzGR6uifzfb9e/UrVqVPVT1dP+qlzquqUIgIzM7O+jGp1AGZmNjw4YZiZWREnDDMzK+KEYWZmRZwwzMysiBOGmZkVccIwG+Yk3S9pv0Hc3iRJT0jSIG3v+5L+JU/vI2nhYGw3b29vSXcN1vZszZwwhoH6HwRJH5D0mKQ3tzKu3gz2D1jB+20r6cX8I/eEpPskfb6uzgOSlkjasFJ2tKTZDbZ3n6TbC953SD9nfs+zJX1lLdY/QtKKyr76k6QfS9qxViciFkbEZtHHTVp5W7/t6z0j4mMR8bVq0VrE/6Kk7SvbvjYidhno9qx/nDCGGUlHAN8BpkZEn/9Z25Gk9ZqwrQDGRsRmwKHAlyTtX6kawHrAJ+s2sdqPl6S3AC8Htpf02sGKs838Pu+nscABwLPALZJ27ed2RB8//pIG+zfGdxq3kBPGMCLpH4GvAwdGxA25rHZ0fbikByU9IukLlXXWl/QtSQ9JWiTpdElj8rJuSe/O03vn7bwtz+8vaU6ePkLSbyV9Pbds/lSr1yDGc4AO4Ff5CPazlRiPkvQg8L+57sWSHpa0NMeya2U7G0g6LbcMlkq6RtLLettWbTWAiLgFuAPYvS68rwOfkbTZGnbzEcAvgcvy9IBIeoekOTn2ayW9urLsfkmfkfSHvPwCSetXln9O0uL8fR1dO6qW9FHgQ8Dn8r6dWXnLPXrbXm8iuT8ijgWuBrry+9f28ag8Py1/57UWyQcl7Qx8H3iDpCclPZbrni3pe5J+LelJoLNBq0iSTpT059yaO6yyYLakoyrzK1sxkq4mfce35VgOVV0Xl6Sd8zaWSpon6Z2VZWdL+q6k/8rrXyfplX3tJ1vFCWP4+DgwHdgvIuY0WP4mYEfSEeOXJe2Uy78IvB54DbBbnv5iXnY10Jmn3wz8Cdgnz78F6K5s//XAXcBWpB/esxoFGRGHAwuAd+RujW9UFr8F2Bl4a56/DNgBGAfcCpxXqXsasAcwBdgS+Bzw4hq2BTlhSJoC/A1wb114N+fP9M+NYlfqrnpfjuN84IOSRjequyaSJpP2z0dz7D8AZtUSdXYocCDwStL3Mi2v+zZSK2g/4K9J30cARMSPcmyn5n17SF/b64efk/4GaiLHsxHwbeCtuVXyRmBuRNwN/BNwXURsGhFbVtb9IPDViNgU+F2D95pA2i9/leP8oSpdYg3UPn/tb/PV+fNfUhfraOBXwP+QWonHA+fVbfsDwEnA5qS/92pXmfXBCWP4OAC4PiIa9a0H0BURz0fEbcAfSD8aAIcB0yPi0Yh4lJR0PpyXXc3qCeLfK/P75OU1D0bEj3O/9gxggqRxa4i3/oRpACdFxLMR8RxARPwkIp6JiOXAV4DdJG0qScCRwPERsSQfCV+f6zXcVn6/P0t6hvQj9b2ImMlLnQQcJ2mrBsveC/wfcDnwX6QurLev4TP25iPAmRFxc479XOA5UvKr+XZE9ETEMtKPXK01dChwdkTcHRH/R/q+SvS2vVKLST/ijbwAvFrSBvk9+jrJPDMirgeofD9VAXwpIpZHxDXAr4H39yPW3k7GvwHYOCJOiYgVETGb9D1+sFLn5xFxS0S8SEq+/d1PI5oTxvDxT8CrJDU8sgd6KtPPAJvk6b8iHfHXPJjLAK7L2xxHSjDnAJPyj+nrgWsq6y2pTUTEs6T/tJvQP4tqE5JGSTpZ0r2SlgH3k35Its6vlwH3lWyrFhap9bMx8FlSV8hLWgcRcQfpR+TEBts8HLg4/8g/D/yCgXVLbUvq+nosv5YCE1m132HN31f1KqKF9P4DWdXb9kptAzxWXxgRzwB/D3wMeFjSryqt1970dRXU0pwMa6p/k2vjFQ3e+0HSZ6tZUpkeyH4a0Zwwho9HgP2BN0v6Xj/WW0z6AavZNpfVfvhvAT4B3B4RK0hJ5NPAvRHxkh+QQr2dmKyWHwa8k9TFtjmwHemHUcBfSEf6O/TzPZR/7E8nHdF/vJd1u0jdRSt/SCRtQ+oG+od8XuVhUovjIEm9HXn3ZiHwtYjYMr+2iIhNIuKignUfJiWXmg5W/6zNOun7bqDhRRQR8ZuIOJDUlXQP8MM+Yukrxi1UuVqN9BkX5+mngY0qyyb0sa2qxcCkurIO4KF+bMPWwAljGImIJaQftbdK+mZl0ZqOQC8Avihpa0lbA18Czq0svwY4jlXdT9118wOxBNi+rqw+xk1JP+pLJW1M6g6r9VUHcDbwTUmvyK2RKZVzAI0+b33ZycDnG538jYg/AReR+rhrDif9GL6K1NraLU8vYvUujXrr55Pxtdd6wI+Af5L0egBJG0s6KH/OvlwMHJlP3m5E+r6qenjpvu2v2rmeUZK2k/QdUhdkV4M64yS9M8eyHHiK1EVVi2Vi3bmZ0vefLmmM0qXhbyd9boC5wHskbSjpr4Gj69Zt9LdVcwPwdL5oYLSkTuAdpP8DNgicMIaHlUdsEbGI1NJ4r6Sv1S9vMP+vpJO9tXMbN7P6ib6rSc3ya+rm+0oYazqKPJl0Wetjkj7dS/1zSF1lDwG3A7+vW/5ZYB5wE/Bo3mbt77XRe69WFhG/JnWxfLSXdb5COpKtlR8O/EdE/DkiHqm9SCes19Qt9WtS18az+d+T8lVaHwW+m68e+mPdNnrddxHxP8AZwOy8Xm2/1M4FnAX8Td63P+9re72YIukJ4PH8PpsAe0bEnQ1iHAV8hvQ9/YV0rqvWcruKdDXaEkmP9OP9HwaWkloE5wLHRMT8vOx0UmJaQjpo+Gndul3AOfnzv6+6IJ/jOhg4KMf6XeDDlW37kty1pGY+QEnSq0hHckE6qtieVUe4F5G6Rx4A3h8Rj+d1zgCmkpqm0yJibtMCNGtz+fLVecDL8olas5ZpagsjIv4YEXtExGTgtaQk8AvgBODKiNiJdJRyIoCkqcAOEbEjcAxwZjPjM2tHkt6Vu2u2AE4BZjlZWDsYyi6pA4A/RcRC4BDSpZnkf2vXkx9C6qog35g2VtL4IYzRrB0cA/wZmE/qnunt5L3ZkOr3TUlr4e9JN0MBjI+IHkgncivX82/D6pfFPZTLqpcMmq3TImJqq2Mwa2RIWhj5KoqDgdXuzGxUtUGZT1SZmbWBoWphTAVuiYi/5PkeSeMjokfSBNI9BpAuYaxeRz2RVddnryTJScTMbAAiYsDD1g/VOYwPsvq10LNYNdbNNGBmpfxwWDke0LJa11W9iBi2r5NOOqnlMTj+1scx0mJ3/K1/ra2mtzDyHZ0HAP9YKT4FuDiPSrmANH4OEXFZvsHpXtIVVUc2Oz4zMyvT9IQRafiJl9eVPUZKIo3qH9fsmMzMrP98p3cLdHZ2tjqEteL4W2c4xw6Of7hr6p3ezSIphmPcZmatJIkYBie9zcxsmHPCMDOzIk4YZmZWxAnDzMyKOGGYmVkRJwwzMyvihGFmZkWcMMzMrIgThpmZFXHCMDOzIk4YZmZWxAnDzMyKOGGYmVkRJwwzMyvihGFmZkWcMMzMrIgThpmZFXHCsKaZ0NGBpH6/JnR0tDp0M2vAj2i1ppEEs2f3f8V998Xfr9ng8yNazcxsSDhhmJlZkaYnDEljJV0i6S5Jd0jaS9IWkq6QdI+kyyWNrdQ/Q9J8SXMl7d7s+MzMrMxQtDC+DVwWEbsAuwF3AycAV0bETsBVwIkAkqYCO0TEjsAxwJlDEJ+ZmRVoasKQtCnw5og4GyAiVkTE48AhwIxcbUaeJ/97Tq57AzBW0vhmxmhmZmWa3cLYHviLpLMl3Srph5I2AsZHRA9ARCwBxuX62wALK+s/lMvMzKzFRg/B9icDx0bEzZJOJ3VH9XbNZKPLvRrW7erqWjnd2dlJZ2fnWgVqZrau6e7upru7e9C219T7MHJ30nURsX2e35uUMHYAOiOiR9IEYHZE7CLpzDx9Ua5/N7BPrTVS2a7vwxgGfB+GWXtp6/sw8g/9QkmvykX7A3cAs4BpuWwaMDNPzwIOB5A0BVhWnyzMzKw1mt0lBXA8cJ6kMcB9wJHAesDFko4CFgCHAkTEZZIOknQv8HSuayPNmDGpddJP4ydNYsmCBU0IyMzAQ4NYE61Nl5S7sswGX1t3SZmZ2brDCcPMzIo4YZiZWREnDDMzK+KEYWZmRZwwzMysiBOGmZkVccIwM7MiThhmZlbECcPMzIo4YZiZWREnDDMzK+KEYWZmRZwwzMysiBOGmZkVccIwM7MiThhmZlbECcPMzIo4YZiZWREnDDMzK+KEYWZmRZwwzMysiBOGmZkVaXrCkPSApD9ImiPpxly2haQrJN0j6XJJYyv1z5A0X9JcSbs3Oz4zMyszFC2MF4HOiNgjIl6fy04AroyInYCrgBMBJE0FdoiIHYFjgDOHID4zMyswFAlDDd7nEGBGnp6R52vl5wBExA3AWEnjhyBGMzPrw1AkjAAul3STpI/ksvER0QMQEUuAcbl8G2BhZd2HcpmZmbXY6CF4jzdGxBJJLweukHQPKYk0ogZlDet2dXWtnO7s7KSzs3MtwzQzW7d0d3fT3d09aNtTRG+/3YNP0knAU8BHSOc1eiRNAGZHxC6SzszTF+X6dwP71Fojle3EUMZtAyMJZs/u/4r77jvg9fx3YdY7SUREowPzIk3tkpK0kaRN8vTGwIHAPGAWMC1XmwbMzNOzgMNz/SnAsvpkYWZmrdHsLqnxwC8kRX6v8yLiCkk3AxdLOgpYABwKEBGXSTpI0r3A08CRTY7PzMwKNTVhRMT9wEvupYiIx4ADelnnuGbGZGZmA+M7vc3MrIgThpmZFXHCMDOzIk4YZmZWxAnDzMyKOGGYmVkRJwwzMyvihGFmZkWcMMzMrIgThpmZFXHCMDOzIk4YZmZWxAnDzMyKOGGYmVkRJwwzMyvihGFmZkWcMMzMrIgThpmZFXHCMDOzIkUJQ9LfNjsQMzNrb6UtjDMl3Sjp45I2b2pEZmbWlooSRkTsDXwImATcLOl8SX/X1MjMzKytFJ/DiIj5wBeBzwP7AGdIulvSe5oVnJmZtY/ScxivkXQ6cBewH/DOiNglT59esP4oSbdKmpXnt5N0vaR7JF0gaXQuX1/ShZLmS7pOUseAP5mZmQ2q0hbGd4Fbgd0i4tiIuBUgIhaTWh19+QRwZ2X+FOC0iNgJWAYcncuPBh6LiB2BbwGnFsZnZmZNVpowDgLOj4hnYWWLYSOAiDh3TStKmpjX/89K8X7ApXl6BvCuPH1Ingf4GbB/YXxmZtZkpQnjSmDDyvxGuazE6cA/AwEgaStgaUS8mJcvArbJ09sACwEi4gVgmaQtC9/HzMyaaHRhvQ0i4qnaTEQ8VWthrImktwM9ETFXUmetOL+qorJstU1Ulq2mq6tr5XRnZyednZ2NqpmZjVjd3d10d3cP2vZKE8bTkibXzl1Iei3wbMF6bwIOlnQQqYWyKencxFhJo3IrYyKwONdfRLp0d7Gk9YDNImJpow1XE4aZmb1U/cH09OnT12p7pV1SnwQukfRbSb8FLgKO62uliPhCRHRExPbAB4CrIuIfgNnAobnaEcDMPD0rz5OXX1UYn5mZNVlRCyMibpK0M7ATqZvo7ohYvhbvewJwoaSvAnOAs3L5WcC5kuYDj5KSjJmZtYHSLimAPYHt8jp7SCIizildOSKuBq7O0/cDezWo8xzw/n7EZGZmQ6QoYUg6F9gBmAu8kIsDKE4YZmY2vJW2MF4H7BoRDa9YMjOzdV/pSe/bgQnNDMTMzNpbaQtja+BOSTcCz9UKI+LgpkRlZmZtpzRhdDUzCDMza3+ll9VeLWlbYMeIuDLf5b1ec0MzM7N2Ujq8+UdJgwH+IBdtA/yyWUGZmVn7KT3pfSxpmI8nYOXDlMY1KygzM2s/pQnjuYh4vjaTH3jkS2zNzEaQ0oRxtaQvABvmZ3lfAvyqeWGZmVm7KU0YJwB/BuYBxwCXUfakPTMzW0eUXiX1IvCj/DIzsxGodCyp+2lwziIPW25mZiNAf8aSqtmA9KwKPzrVzGwEKTqHERGPVl4PRcS3gP2aHJtZ/4wZg6QBvSZ0dLQ6erO2V9olNbkyO4rU4ti0KRGZDdTy5TB79oBW7dl330EOxmzdU9oldVplegXwAH7QkZnZiFJ6lZQPv8zMRrjSLqlPr2l5RHxzcMIxM7N21Z+rpPYEZuX5dwLXAAubEZSZmbWf/jxAaXJEPAkgqQu4JCI+0qzAzMysvZQODdIBPF+Zfx7YbtCjMTOztlXawjgXuFHSL0h3fL8bOKdpUZmZWdspvXHva8CRwFJgGXBkRPxbX+tJepmkGyTNkTRP0km5fDtJ10u6R9IFebh0JK0v6UJJ8yVdJ8l3U5mZtYnSLimAjYAnIuLbwCJJr+xrhYh4Dtg3IvYAdgemStoLOAU4LSJ2IiWgo/MqRwOPRcSOwLeAU/sRn5mZNVHpI1pPAj4PnJiLxgA/LVk3Ip7Jky8jdYEFsC9waS6fAbwrTx+S5yE9Enb/kvcwM7PmK21hvBs4GHgaICIWUzg0iKRRkuYAS4DfAH8CluUh0wEWkZ4RTv53YX6PF4BlkjzIoZlZGyg96f18RISkAJC0cekb5MSwh6TNgF8AuzSqlv9VXbno5VGwXV1dK6c7Ozvp7OwsDcnMbETo7u6mu7t70LZXmjAulvQDYHNJHwWOop8PU4qIJyRdDUzJ2xmVk8lEYHGutgiYBCyWtB6wWUQsbbS9asIwM7OXqj+Ynj59+lptr/QqqW+QzilcCuwEfDkivtPXepK2ljQ2T28IHADcCcwmPVMD4AhgZp6elefJy68q+xhmZtZsfbYw8pH+5RFxAOkcRH+8ApghaRQpOV0UEZdJugu4UNJXgTnAWbn+WcC5kuYDjwIf6Of7mZlZk/SZMCLiBUnPSBobEY/3Z+MRMQ+Y3KD8fmCvBuXP4WHTzczaUuk5jP8D5kn6DflKKYCIOL4pUZmZWdspTRi/zi8zMxuh1pgwJHVExIKImLGmemZmtu7r6yqpX9YmJF26poq27prQ0YGkfr/MbN3SV5dU9X/99s0MxNpXz8KFMHt2/1fc10/2NVuX9NXCiF6mzcxshOmrhbGbpCdILY0N8zR5PiJis6ZGZ2ZmbWONCSMi1huqQMzMrL3153kYZmY2gjlhmJlZEScMMzMr4oRhZmZFnDDMzKyIE4aZmRVxwjAzsyJOGGZmVsQJw8zMijhhmJlZEScMMzMr4oRhBjBmzICe+TGho6PVkZsNmdJHtJqt25YvH9AzP3r8zA8bQdzCMDOzIk4YZmZWpKkJQ9JESVdJulPSPEnH5/ItJF0h6R5Jl0saW1nnDEnzJc2VtHsz4zMzs3LNbmGsAD4dEbsCbwCOlbQzcAJwZUTsBFwFnAggaSqwQ0TsCBwDnNnk+MzMrFBTE0ZELImIuXn6KeAuYCJwCDAjV5uR58n/npPr3wCMlTS+mTGamVmZITuHIWk7YHfgemB8RPRASirAuFxtG2BhZbWHcpmZmbXYkFxWK2kT4GfAJyLiKUnRW9UGZQ3rdnV1rZzu7Oyks7NzLaM0M1u3dHd3093dPWjba3rCkDSalCzOjYiZubhH0viI6JE0AXgkly8CJlVWnwgsbrTdasIwM7OXqj+Ynj59+lptbyi6pH4M3BkR366UzQKm5elpwMxK+eEAkqYAy2pdV2Zm1lpNbWFIehPwIWCepDmk7qUvAKcAF0s6ClgAHAoQEZdJOkjSvcDTwJHNjM/MzMo1NWFExO+A9XpZfEAv6xzXvIjMzGygfKe3mZkVccIYQSZ0dAxoRFYzM/BotSNKz8KFAxqRFY/Iama4hWFmZoWcMMzMrIgThpmZFXHCMDOzIk4YZmZWxAnDzMyKOGGYmVkRJwwzMyvihGFmZkWcMMzMrIgThpmZFXHCMDOzIk4YZmZWxAmjhQY63PiEjo5Wh241Y8b4O7QRw8Obt9BAhxvv8XDj7WP5cn+HNmK4hWFmZkWcMMzMrIgThpmZFXHCMDOzIk4Yw9EAr8yx4c9X1lkrNfUqKUlnAe8AeiLiNblsC+AiYFvgAeD9EfF4XnYGMBV4GpgWEXObGd+wNcArc/CVOcOer6yzVmp2C+Ns4K11ZScAV0bETsBVwIkAkqYCO0TEjsAxwJlNjs3MzPqhqQkjIq4FltYVHwLMyNMz8nyt/Jy83g3AWEnjmxmfmZmVa8U5jHER0QMQEUuAcbl8G2Bhpd5DuczMzNpAO93p3eisbPRWuaura+V0Z2cnnZ2dgx+Rmdkw1t3dTXd396BtrxUJo0fS+IjokTQBeCSXLwImVepNBBb3tpFqwjAzs5eqP5iePn36Wm1vKLqkxOqth1nAtDw9DZhZKT8cQNIUYFmt68rMzFqv2ZfVng90AltJWgCcBJwMXCLpKGABcChARFwm6SBJ95Iuqz2ymbGZmVn/NDVhRMRhvSw6oJf6xzUxHDMzWwu+09vMzIo4YZiZWREnDDMzK+KEYWZmRdrpxj2zkSOPOGw2nDhhmLWCRxy2YchdUmZmVsQJw8zMijhhDIKBPgXNzGw48TmMQTDQp6C5P9rMhhO3MMzMrIgThpmZFXHCMDOzIk4YZmZWxAnDzMyKOGGYmVkRJwwzMyvihGFmZkV84162YsUKli9f3u/1Ro1yzjWrN6GjI93Q2k/jJ01iyYIFTYjIBoMTRvbqPfdk/p139nvIjlEe4sPsJQY6+kGPRz9oa04Y2eLFi3nh/PNhq636td7Y447j+TvuaFJUZmbtw/0pZmZWpO0ShqS3Sbpb0h8lfb7V8ZitE/IT/vr7mtDR0erIrY20VZeUpFHAd4H9gcXATZJmRsTdrY1skM2dC7vv3uooBs7xDz8DfMJfz4EHDu5Q/H3t+wE+unaoTpZ3d3fT2dnZ9PdpV22VMIDXA/Mj4kEASRcChwBOGO3E8Y8cg/0o2b72/UAT2xCdLB/pCaPduqS2AarX4i3KZWZm1mLt1sJo1BaNoXjj9ceMYdNTT0Xrr9+v9Z598MEmRWRmxQbYlQUD784aifeaKGJIfo+LSJoCdEXE2/L8CUBExCl19donaDOzYSQiBnxSqt0SxnrAPaST3g8DNwIfjIi7WhqYmZm1V5dURLwg6TjgCtL5lbOcLMzM2kNbtTDMzKx9tdtVUg1JGiXpVkmz8vx2kq6XdI+kCyS1VUupStJYSZdIukvSHZL2krSFpCty/JdLGtvqOHsj6VOSbpd0m6TzJK3fzvtf0lmSeiTdVinrdX9LOkPSfElzJbX8Wtte4j81//3MlXSppM0qy07M8d8l6cDWRL1Ko/gryz4r6UVJW1bK2n7/5/L/l28onifp5Ep52+9/SbtJuk7SHEk3Stqzsqx/+z8i2v4FfAr4KTArz18EHJqnvw8c0+oY1xD7T4Aj8/RoYCxwCvC5XPZ54ORWx9lL7H8F3AesX9nvR7Tz/gf2BnYHbquUNdzfwFTg13l6L+D6No3/AGBUnj4Z+Pc8vSswJ/9dbQfcS+41aKf4c/lE4H+A+4Eth9n+7yR1k4/O81vnf3cZDvsfuBw4sLLPZ+fpg/q7/9u+hSFpIumD/WeleD/g0jw9A3j3UMdVQtKmwJsj4myAiFgREY+TbkackavNAN7VohBLrAdsnFsRG5LuwN+XNt3/EXEtsLSuuH5/H1IpPyevdwMwVtL4oYizN43ij4grI+LFPHs96ccX4GDgwvx39QAwn3Tza8v0sv8BTgf+ua5sWOx/4GOkg4wVuc5fcvkhDI/9/yLpQBVgc+ChPH0w/dz/bZ8wWPWHFgCStgKWVv4DLSIdCbej7YG/SDo7d6n9UNJGwPiI6AGIiCXAy1saZS8iYjFwGrCA9Ef2OHArsGyY7P+acXX7e1wur79R9CHa/0bRo4DL8vSwiF/SO4GFETGvbtGwiB94FfCW3A07W9Jrc/lwif9TwDckLQBOBU7M5f2Ov60ThqS3Az0RMZdVN/WpMl3TrmfuRwOTgf+IiMnA08AJtG+8q5G0OekoaltSUtiY1KStNyw+TwMtu1F0ICT9C7A8Ii6oFTWo1lbxS9oQ+BfgpEaLG5S1VfzZaGDziJgCfA64JJcPl/g/BnwiIjpIyePHubzf8bd1wgDeBBws6T7gAlJX1LdITada7BNJ3STtaBHpyOrmPH8pKYH01Jp+kiYAj7Qovr4cANwXEY9FxAvAL4A3ApsPk/1f09v+XgRMqtRr288i6QhS1+xhleLhEP8OpP79P0i6nxTjrZLGMTzih3QU/nOAiLgJeCH3dCwCqsP5tmv8R0TELwEi4mdA7aR3v/d/WyeMiPhCRHRExPbAB4CrIuIfgNnAobnaEcDMVsW4JrkbZKGkV+Wi/YE7gFnAtFzWtvGTuqKmSNpAadyFWvztvv/rW6HV/T2NVfHOAg6HlaMMLKt1XbXYavFLehvpyPbgiHiuUm8W8IF85dorgb8m3ezaaivjj4jbI2JCRGwfEa8k/UjtERGPMEz2P/BL0t8++f/y+hHxKCn+v2/n/Z89JGkfAEn7k861wED2fyvP6Pfz7P8+rLpK6pXADcAfSVfsjGl1fGuIezfgJmAu6ShlLLAlcCXprvbfkJq7LY+1l/hPAu4CbiOdMB7TzvsfOJ90lPQcKeEdCWzR2/4mDad/L/AHYHKbxj8feJB0/uhW4HuV+ifm+O8iXwnTbvHXLb+PfJXUMNr/o4FzgXnAzcA+w2n/k3oFbiZd0XUdKWEPaP/7xj0zMyvS1l1SZmbWPpwwzMysiBOGmZkVccIwM7MiThhmZlbECcPMzIo4YdiQyMNaf70y/xlJXx7gts6W9J4BrntEvtu7Nv9DSTv3Uu87A3mPuu1cuzbxDTZJD0j6Q37NljSpsqzP7yivd16z4rP25oRhQ+U54D3VZyFNt0j4AAAEVklEQVS0yDQqA6xFxD9GxN291F3rm5QiYu9+rjKNXgaAqwzHMiD5bv0Xgc6I2A24GvhSpcoav6OcWEUaiG/DtYnFhicnDBsqK4AfAp+uX1DfYpD0ZGX6c0oPb5oj6d8arDtZUrekmyT9d2XMqN3zQ2NqDx3aXNJ7gdcBP82jB2+Qj7In53WOVHrI0mzSOGa199ha0s8k3ZBfb8jl++S4bpV0i6SNG8T3ZKXubK16mNa5Deo2iu9+SV+SdA3wPkkfUXoIzpy8rQ3yuuMk/Tx/3jmSpkjaVtKdkv5D0i2kcYOqw0Zcx+ojDff6HWWHkYbDvoI0NLaNNK2+ld2vkfECngA2IT1AZ1PgM8CX87KzgfdU6+Z/pwLXAi/L85tX65OGbPgdsFUufz/pOfCQhjrYO09PB76Zp2ez+tAIs0kDQk4gDb+xJWn4k2uBM3Kd84A35ulJwJ15ehbwhjy9EfkhR/WfO/+7D+k5Ba8g/WD/vrbNuvpX1cV3P/DZyvwWlemvAsfm6QuB4/O08j7elpQE9qzbXu0BRqcDHyn5jvLye/Ln/ztgZqv/pvwa+lfbPFrT1n0R8ZSkGcAngGcLVtkfODvygHsRsaxu+U7A3wK/yd0to4DFSo8wHRvpYTKQxsC6uLJeo2Gd9yI9iewxAEkXATvmZQcAu+T3ANgktyZ+B5ye+/R/HhEP1W+0zo0R8XDe/lzSKK6/r6vTaPj+iyrTr5b0r6QH4WxMepoapJGcPwwQEQE8mbuWHow0wmrV7HyepIc09PhKvX1Hkl4HPBIRCyUtBn4saWykB4LZCOEuKRtq3waOJv3Y1axg9b/F9fO/Ys3nEQTcHhGTI2KPiNgtIho9r2NtCZiS32OPSCMoPx0Rp5A+y4bA9ZVRiXtTHWn2BSg+YHu6Mv0T4OMR8RrgK8AGuby3/fR0g7JO0rDcd5BaKfUafUeHATsrPWrgXlIL5L1l4du6wgnDhkptuOulpKP9oyvLHiD13SPpXaQuIUh95UfVTrBK2qJum/cAL89DMyNptKRdI+IJYKmk2nmID5NO8AI8CWzWIL4bgH0kbSFpDKuGb6/FcfzKDyLtlv/dPiLuiIhTSaOBvuRqKxq3ZtbkiV7iq9kEWJJj/FCl/H+Bj+e4Rik9Hri391dutX0KOFzpQVkr69Z/R7ll9T7g1bFqmPJ3sfqzOWwEcMKwoVI9Aj4N2KpS9iPSj/X1pGciPw0QEZeTzhPcLOlWUp/6ym1FxHLSD9kpuYtnDvCGXGca6bGUc0lDzH8ll/8EOLN2UrmyrSVAF+mZ2VcAt1Ti/QTwunxJ6e3AMbn8k5LmSZoDPAP8dx+fu6R8RqP4Kr5EeubC5aQhtWs+Cewr6TZS8tq1l/dZOZ8/8/nAsQ3q1r4jgLcAi3L9mmtI3XQtfQa3DS0Pb25mZkXcwjAzsyJOGGZmVsQJw8zMijhhmJlZEScMMzMr4oRhZmZFnDDMzKyIE4aZmRX5/w+rSE0Tg46UAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x2b6e92a19e50>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#tracrRNA BLAST Hit Length Distribution\n",
    "dist = Series(lengthDist)\n",
    "ax = dist.plot.hist(bins=20,color='c');\n",
    "ax.set_title(\"Known tracrRNA Length Distribution\");\n",
    "ax.set_xlabel(\"Nucleotides in tracrRNA\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    83.000000\n",
      "mean      8.397590\n",
      "std      14.400999\n",
      "min       1.000000\n",
      "25%       1.000000\n",
      "50%       3.000000\n",
      "75%       7.500000\n",
      "max      81.000000\n",
      "dtype: float64\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYMAAAEACAYAAABRQBpkAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAFHVJREFUeJzt3X+UZ3V93/Hna2FdhpWsRHG3igvZJCDayi91UWOz6GIgqYA9JQmNPeCPnp5TDLSepiw2J0yPp23IKSW0m3NSLSEbFUXWKJsWC3LgS2xag4TdDML+MCsKxO5Eo2CRyLDsu39878RxmZ39DjPfe7+7PB/nzNnvvfO93/va79z5vuZ+7r3fb6oKSdIL25KuA0iSumcZSJIsA0mSZSBJwjKQJGEZSJIYchkkOSnJ1iT3N/8+keTyJMcmuSPJziS3J1kxzBySpLmlresMkiwBHgPWAh8A/rqqfivJlcCxVbWhlSCSpOdoc5hoPbC7qh4FLgA2NfM3ARe2mEOStJ82y+CXgJua2yurahKgqvYAx7WYQ5K0n1bKIMlS4HzglmaW74EhSSPkyJbWcx7wZ1X17WZ6MsnKqppMsgr4q9kWSmJpSNLzUFWZz/3bGia6GPjkjOktwKXN7UuAWw+0YFWN1NfVV1/deQYzHV65zGSmxf56PoZeBknG6B88/sMZs68Bzkmys/nebw47hyTpwIY+TFRVf8N+B4ir6jv0S0CSNAK8Anme1q1b13WE5zDT4EYxl5kGY6bhau2is+cjSY1yPkkaRUmoET2ALEkaYZaBJMkykCRZBpIkLANJEpaBJAnLQJKEZSBJwjKQJNHeW1gv2GvPPJOvbt/e+nrPfec72XLzza2vV5LadMiUwV/s2MEzH/84jI21uNK/4MGPfrS99UlSRw6ZMgD6RdBmGSxb1t66JKlDHjOQJFkGkiTLQJKEZSBJwjKQJGEZSJKwDCRJWAaSJCwDSRItlEGSFUluSbI9yYNJ1iY5NskdSXYmuT3JimHnkCQdWBt7BtcDt1XVKcCpwA5gA3BnVZ0M3AVc1UIOSdIBDLUMkhwDvLWqbgSoqr1V9QRwAbCpudsm4MJh5pAkzW3YewZrgG8nuTHJ/Uk+kuRoYGVVTQJU1R7guCHnkCTNYdjvWnokcAZwWVXdl+Q6+kNENegDjI+PA7B3agomJmDt2mHklKRDVq/Xo9frLegxUjXw6/L8HzxZCfyfqlrTTP8M/TL4SWBdVU0mWQXc3RxT2H/5ms63bPlypjZvbvctrHftYs3GjeyemGhvnZK0QEmoqsxnmaEOEzVDQY8mOamZ9XbgQWALcGkz7xLg1mHmkCTNrY0Pt7kc+ESSpcDXgPcARwCfTvJe4BHgohZySJIOYOhlUFV/Drxhlm+tH/a6JUmD8QpkSZJlIEmyDCRJWAaSJCwDSRKWgSQJy0CShGUgScIykCRhGUiSsAwkSVgGkiQsA0kSloEkCctAkoRlIEnCMpAkYRlIkrAMJElYBpIkLANJEpaBJAnLQJIEHDnsFST5OvAEsA94pqremORY4GbgBODrwC9W1RPDziJJml0bewb7gHVVdXpVvbGZtwG4s6pOBu4CrmohhyTpANoog8yynguATc3tTcCFLeSQJB1AG2VQwO1Jvpzk/c28lVU1CVBVe4DjWsghSTqAoR8zAN5cVXuSHAfckWQn/YIYyPj4OAB7p6ZgYgLWrh1OSkk6RPV6PXq93oIeI1UDvy4vWJKrgSeB99M/jjCZZBVwd1WdMsv9azrfsuXLmdq8GcbGWsvLrl2s2biR3RMT7a1TkhYoCVWV+Swz1GGiJEcneXFzeznwDuABYAtwaXO3S4Bbh5lDkjS3YQ8TrQQ+m6SadX2iqu5Ich/w6STvBR4BLhpyDknSHIZaBlX1MHDaLPO/A6wf5rolSYPzCmRJkmUgSbIMJElYBpIkLANJEpaBJAnLQJKEZSBJwjKQJGEZSJKwDCRJWAaSJCwDSRKWgSQJy0CShGUgScIykCRhGUiSsAwkSQxYBkn+7rCDSJK6M+iewe8muTfJP0/ykqEmkiS1bqAyqKqfAX4FeBVwX5Kbkpwz1GSSpNYMfMygqr4K/DpwJfCzwH9OsiPJPxxWOElSOwY9ZvC6JNcB24G3Ae+sqlOa29cNsPySJPcn2dJMn5jkS0l2JvlkkiMX8H+QJC3QoHsGG4H7gVOr6rKquh+gqr5Jf2/hYK4AHpoxfQ1wbVWdDDwOvG/wyJKkxTZoGfw8cFNV/Q387V/6RwNU1cfmWjDJ8c3y/23G7LcBn2lubwLeNZ/QkqTFNWgZ3AmMzZg+upk3iOuAXwMKIMlLge9W1b7m+48BrxjwsSRJQzDoWP1RVfXk9ERVPTm9ZzCXJL8ATFbVtiTrpmc3XzPVgR5jfHwcgL1TUzAxAWvXDhhZkl4Yer0evV5vQY+RqgO+Dv/wTsmfAL86fawgyZnAxqp600GW+/fAu4G99PcsjgE+B7wDWFVV+5KcBVxdVefNsnxN51u2fDlTmzfD2Nj+dxueXbtYs3Ejuycm2lunJC1QEqpq/z+65zToMNG/AG5J8sUkXwRuBj5wsIWq6kNVtbqq1gC/DNxVVe8G7gYuau52CXDrfEJLkhbXQMNEVfXlJK8GTqY/xLOjqp5ZwHo3AJ9K8mFgK3DDAh5LkrRA8zm//w3Aic0ypze7IX8w6MJVdQ9wT3P7YcDBf0kaEQOVQZKPAT8JbAOebWYXMHAZSJJG16B7Bq8HXlODHG2WJB1yBj2A/BVg1TCDSJK6M+iewcuAh5LcCzw9PbOqzh9KKklSqwYtg/FhhpAkdWvQU0vvSXIC8NNVdWdz9fERw40mSWrLoG9h/U+BzcB/bWa9kv6VxJKkw8CgB5AvA94CfA/+9oNuXj6sUJKkdg1aBk9X1dT0RPNhNJ5mKkmHiUHL4J4kHwLGms8+vgX4o+HFkiS1adAy2AB8C3gA+GfAbQz2CWeSpEPAoGcT7QM+2nxJkg4zg7430cPMcoygeWtqSdIhbj7vTTTtKPqfRfDjix9HktSFgY4ZVNVfz/j6y6r6bfofai9JOgwMOkx0xozJJfT3FI4ZSiJJUusGHSa6dsbtvcDXgV9c9DSSpE4MejbR2cMOIknqzqDDRB+c6/tV9Z8WJ44kqQvzOZvoDcCWZvqdwB8Djw4jlCSpXfP5cJszqur/ASQZB26pqvcPK5gkqT2Dvh3FamBqxvQUcOKip5EkdWLQPYOPAfcm+Sz9K5HfBfzB0FJJklo16NlE/y7J54G3NrPeU1VbD7ZckmX0jy28qFnX5qr6t0lOBD4FHAvcD/yTqto7//iSpMUw6DARwNHA96rqeuCxJD9xsAWq6mng7Ko6HTgNOC/JWuAa4NqqOhl4HHjf/KNLkhbLoB97eTVwJXBVM2sp8PFBlq2qp5qby+jvHRRwNvCZZv4m+sNOkqSODLpn8C7gfOD7AFX1TQZ8O4okS5JsBfYAXwB2A483b4sN8BjwivmEliQtrkEPIE9VVSUpgCTLB11B86J/epIfAz4LnDLb3Q60/Pj4OAB7p6ZgYgLWrh101ZL0gtDr9ej1egt6jFQd/KOMk/wr4KeBc4D/ALwXuKmq/su8Vpb8BvAU8K+BVVW1L8lZwNVVdd4s96/pfMuWL2dq82YYG5vPKhdm1y7WbNzI7omJ9tYpSQuUhKrKfJYZ9C2s/yOwmf44/8nAbwxSBElelmRFc3sMWA88BNxN/zMRAC4Bbp1PaEnS4jroMFGSI4Dbq2o9/TH/+fg7wKYkS+gXz81VdVuS7cCnknwY2ArcMM/HlSQtooOWQVU9m+SpJCuq6on5PHhVPQCcMcv8hwEH/yVpRAx6APkHwANJvkBzRhFAVV0+lFSSpFYNWgb/o/mSJB2G5iyDJKur6pGq2tRWIElS+w52NtHnpm8k+cxcd5QkHboOVgYzz1NdM8wgkqTuHKwM6gC3JUmHkYMdQD41yffo7yGMNbdppquqfmyo6SRJrZizDKrqiLaCSJK6M5/PM5AkHaYsA0mSZSBJsgwkSVgGkiQsA0kSloEkCctAkoRlIEnCMpAkYRlIkrAMJElYBpIkLANJEpaBJIkhl0GS45PcleShJA8kubyZf2ySO5LsTHJ7khXDzCFJmtuw9wz2Ah+sqtcAbwIuS/JqYANwZ1WdDNwFXDXkHJKkOQy1DKpqT1Vta24/CWwHjgcuADY1d9sEXDjMHJKkubV2zCDJicBpwJeAlVU1Cf3CAI5rK4ck6bnm/AzkxZLkxcBm4IqqejJJDbrs+Pg4AHunpmBiAtauHU5ISTpE9Xo9er3egh4jVQO/Lj+/FSRHAv8d+HxVXd/M2w6sq6rJJKuAu6vqlFmWrel8y5YvZ2rzZhgbG2reH7FrF2s2bmT3xER765SkBUpCVWU+y7QxTPR7wEPTRdDYAlza3L4EuLWFHJKkAxjqMFGStwC/AjyQZCtQwIeAa4BPJ3kv8Ahw0TBzSJLmNtQyqKo/AY44wLfXD3PdkqTBeQWyJMkykCRZBpIkLANJEpaBJAnLQJKEZSBJwjKQJGEZSJKwDCRJWAaSJCwDSRKWgSQJy0CShGUgScIykCRhGUiSsAwkSVgGkiQsA0kSloEkCctAkoRlIEliyGWQ5IYkk0kmZsw7NskdSXYmuT3JimFmkCQd3LD3DG4Efm6/eRuAO6vqZOAu4KohZ5AkHcRQy6Cq/hfw3f1mXwBsam5vAi4cZgZJ0sF1cczg5VU1CVBVe4DjOsggSZrhyK4DHMz4+DgAe6emYGIC1q7tNpAkjZher0ev11vQY6SqFifNgVaQnAD8UVW9rpneDqyrqskkq4C7q+qUAyxb0/mWLV/O1ObNMDY21Lw/Ytcu1mzcyO6JiYPfV5JGRBKqKvNZpo1hojRf07YAlza3LwFubSGDJGkOwz619CbgfwMnJXkkyXuA3wTOSbITWN9MS5I6NNRjBlX1jw/wrfXDXK8kaX68AlmSZBlIkiwDSRKWgSSJQ+Cis659Y/duknmdrrsoVr7qVex55JHW1yvphckyOIhnn3oK7r679fVOnn126+uU9MLlMJEkyTKQJFkGkiQsA0kSloEkCctAkoRlIEnCMpAk4UVno2vpUq98ltQay2BUPfOMVz5Lao3DRJIky0AvXKtWryZJ61+rVq/u+r8uPYfDRHrBmnz0UYfipIZ7BpIky0CSZBlIkuiwDJKcm2RHkl1JruwqhySpozJIsgTYCPwc8Frg4iSv7iLLvG3b1nWC5xrBTL1er+sIsxrFXGYajJmGq6uzid4IfLWqvgGQ5FPABcCOjvIMbts2OO20rlP8qMXM1NGVz0uOOop9P/hB6+sdBb1ej3Xr1nUd40csZqZVq1f3z9xqUVfb06F8BX9XZfBKYObW8Rj9glDXFuvK59//fbj00oHvvu/ss9s5zXNmLk/xbMWincI7j22qk+2JQ/u04a7KYLY/PWuuBY5cupSjPvxhOOKIIUV6rmeffJLvt7Y2SepOquZ8DR7OSpOzgPGqOreZ3gBUVV2z3/3aDydJh4Gqmtd4b1dlcASwE3g78H+Be4GLq2p762EkSd0ME1XVs0k+ANxB/4ymGywCSepOJ3sGkqTRMpJXII/KBWlJbkgymWRixrxjk9yRZGeS25OsaDnT8UnuSvJQkgeSXN51riTLkvxpkq1Npqub+Scm+VKT6ZNJWt8TTbIkyf1JtoxCpiRfT/LnzXN1bzOv621qRZJbkmxP8mCStSOQ6aTmObq/+feJJJePQK5/meQrSSaSfCLJi0Zgm7qi+b1b0OvByJXBiF2QdmOTY6YNwJ1VdTJwF3BVy5n2Ah+sqtcAbwIua56fznJV1dPA2VV1OnAacF6StcA1wLVNpseB97WVaYYrgIdmTHedaR+wrqpOr6rp06m73qauB26rqlOAU+lf79Nppqra1TxHZwBnAt8HPttlriSvAH4VOKOqXkd/mP1iOtymkry2Wd/r6f/u/YMkP8XzeZ6qaqS+gLOAz8+Y3gBc2WGeE4CJGdM7gJXN7VXAjo6fr88B60clF3A0cB/960b+Clgy4+f6P1vOcjzwBWAdsKWZ962OMz0MvHS/eZ397IBjgN2zzB+J7alZ/zuAL3adC3gF8A3gWPpFsAU4p8vtHPhHwEdmTP868GvA9vk+TyO3Z8DsF6S9sqMss3l5VU0CVNUe4LiugiQ5kf5fA1+i/4PvLFczHLMV2EP/BXg38HhV7Wvu8hj9X6Y2XUf/F6OajC8FvttxpgJuT/LlJO9v5nX5s1sDfDvJjc2QzEeSHN1xpv39EnBTc7uzXFX1TeBa4BHgL4EngPvpdjv/CvD3m2Gho4GfB17F83ieRrEM5n1B2gtRkhcDm4ErqupJOn6Oqmpf9YeJjqe/V3DKbHdrK0+SXwAmq2obP9ymwnO3r7aftzdX1evp/9JeluStHWSY6UjgDOB3qj8k8336e+Mj8TuXZClwPnBLM6uzXEleQv9tc06g/4K/HDhvlru2lrGqdtAfproTuA3YRn8oed5GsQweA2Z+LuDxwDc7yjKbySQrAZKsor+L2KrmANVm4GNVdeuo5AKoqu8B99DfXX5JcwwI2v85vgU4P8nXgE8CbwN+G1jRYabpv9Koqm/RH+J7I93+7B4DHq2q+5rpz9Avh5HYnui/2P5ZVX27me4y13rga1X1nap6lv4xjDfT7XZOVd1YVWdW1Trgu8AunsfzNIpl8GXgp5KckORFwC/TH5vryv5/TW4BLm1uXwLcuv8CLfg94KGqun7GvM5yJXnZ9NkKScbo/9I8BNwNXNRFpqr6UFWtrqo19Lehu6rq3V1mSnJ0s0dHkuX0x8IfoMOfXTOU8GiSk5pZbwce7DLTfi6mX+bTusz1CHBWkqOShB8+V51tUwBJjmv+XQ28i/7zNf/nqa0DHfM8KHIu/SuUvwps6DDHTfRb/mn6G8J76B88urPJ9wXgJS1negvwLP3dwa30xyzPBX68q1zA32tybAMmgH/TzP8J4E/p/6VyM7C0o5/jz/LDA8idZWrWPf1ze2B62+7yZ9es/1T6f4RtA/4QWNF1pibXGP0D/sfMmNf1c3U1/YOzE8AmYGnX2znwx/SPHWylf6ba83qevOhMkjSSw0SSpJZZBpIky0CSZBlIkrAMJElYBpIkLANJEpaBJAn4/2TgV3HnduGSAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x2b6eb6c9bad0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "blastClusterColorCounter = Counter(blastTreeColors.values())\n",
    "dist = Series(blastClusterColorCounter)\n",
    "dist.plot.hist(color='c');\n",
    "print dist.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of tracrRNAs that did not cross hit in 1 systems: 25\n",
      "Number of tracrRNAs that did not cross hit in 2 systems: 14\n",
      "Number of tracrRNAs that did not cross hit in 3 systems: 7\n",
      "Number of tracrRNAs that did not cross hit in 4 systems: 6\n",
      "Number of tracrRNAs that did not cross hit in 5 systems: 4\n",
      "Number of tracrRNAs that did not cross hit in 6 systems: 1\n",
      "Number of tracrRNAs that did not cross hit in 7 systems: 5\n",
      "Number of tracrRNAs that did not cross hit in 8 systems: 3\n",
      "Number of tracrRNAs that did not cross hit in 9 systems: 1\n"
     ]
    }
   ],
   "source": [
    "for i in range(1,10):\n",
    "    print \"Number of tracrRNAs that did not cross hit in %i systems: %i\" % (i,len(dist[dist==i]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BLAST results are known, now prepare for structural search"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4>Create a file for all of the tracrRNA hits.</h>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add hits to a fasta file\n",
    "tracrSeqs = {}\n",
    "fh = open(\"sequences/PaperTracrBLAST_Hits.fa\",\"w\")\n",
    "index = 0\n",
    "for id in tracrBLAST:\n",
    "    for queryID, coord in tracrBLAST[id]:\n",
    "        seq = casRelatedAssemblies[id][min(coord.start,coord.end):max(coord.start,coord.end)]\n",
    "        seq.id += \"_%i\" % index\n",
    "        index+=1\n",
    "        write(seq,fh,\"fasta\")\n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4>Cluster all BLAST hits to form consensus sequences</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================================================================\n",
      "Program: CD-HIT, V4.7 (+OpenMP), Jul 11 2017, 18:04:58\n",
      "Command: cd-hit-est -i sequences/PaperTracrBLAST_Hits.fa -o\n",
      "         clusters/PaperTracrBLAST_Hits.fa -M 0 -d 0 -n 7 -c .9\n",
      "         -T 0 -sc 1\n",
      "\n",
      "Started: Fri Jan 11 15:50:40 2019\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "total number of CPUs in the system is 28\n",
      "Actual number of CPUs to be used: 28\n",
      "\n",
      "total seq: 2966\n",
      "longest and shortest : 167 and 49\n",
      "Total letters: 272089\n",
      "Sequences have been sorted\n",
      "\n",
      "Approximated minimal memory consumption:\n",
      "Sequence        : 0M\n",
      "Buffer          : 28 X 12M = 339M\n",
      "Table           : 2 X 0M = 0M\n",
      "Miscellaneous   : 0M\n",
      "Total           : 340M\n",
      "\n",
      "Table limit with the given memory limit:\n",
      "Max number of representatives: 449393\n",
      "Max number of word counting entries: 10850988\n",
      "\n",
      "# comparing sequences from          0  to         98\n",
      "---------- new table with       37 representatives\n",
      "# comparing sequences from         98  to        193\n",
      "...................---------- new table with        7 representatives\n",
      "# comparing sequences from        193  to        285\n",
      ".................---------- new table with       21 representatives\n",
      "# comparing sequences from        285  to        374\n",
      ".............---------- new table with       14 representatives\n",
      "# comparing sequences from        374  to        460\n",
      "99.7%---------- new table with        7 representatives\n",
      "# comparing sequences from        460  to        543\n",
      ".......---------- new table with       13 representatives\n",
      "# comparing sequences from        543  to        623\n",
      "99.2%---------- new table with        8 representatives\n",
      "# comparing sequences from        623  to        701\n",
      "...---------- new table with       11 representatives\n",
      "# comparing sequences from        701  to        776\n",
      "99.6%---------- new table with        4 representatives\n",
      "# comparing sequences from        776  to        849\n",
      "..---------- new table with        7 representatives\n",
      "# comparing sequences from        849  to        919\n",
      "99.6%---------- new table with        2 representatives\n",
      "# comparing sequences from        919  to        987\n",
      "..---------- new table with        3 representatives\n",
      "# comparing sequences from        987  to       1052\n",
      "..---------- new table with        6 representatives\n",
      "# comparing sequences from       1052  to       1115\n",
      "99.6%---------- new table with        2 representatives\n",
      "# comparing sequences from       1115  to       1176\n",
      "---------- new table with        1 representatives\n",
      "# comparing sequences from       1176  to       1235\n",
      "99.7%---------- new table with        5 representatives\n",
      "# comparing sequences from       1235  to       1292\n",
      "---------- new table with        2 representatives\n",
      "# comparing sequences from       1292  to       1347\n",
      "99.8%---------- new table with        2 representatives\n",
      "# comparing sequences from       1347  to       1400\n",
      "---------- new table with        1 representatives\n",
      "# comparing sequences from       1400  to       1452\n",
      "---------- new table with        2 representatives\n",
      "# comparing sequences from       1452  to       1502\n",
      ".---------- new table with        2 representatives\n",
      "# comparing sequences from       1502  to       1550\n",
      ".---------- new table with        1 representatives\n",
      "# comparing sequences from       1550  to       1597\n",
      "---------- new table with        2 representatives\n",
      "# comparing sequences from       1597  to       1642\n",
      "---------- new table with        1 representatives\n",
      "# comparing sequences from       1642  to       1686\n",
      "79.4%---------- new table with        2 representatives\n",
      "# comparing sequences from       1686  to       1728\n",
      "---------- new table with        0 representatives\n",
      "# comparing sequences from       1728  to       1769\n",
      "---------- new table with        2 representatives\n",
      "# comparing sequences from       1769  to       1808\n",
      ".---------- new table with        2 representatives\n",
      "# comparing sequences from       1808  to       1846\n",
      "---------- new table with        0 representatives\n",
      "# comparing sequences from       1846  to       1883\n",
      "---------- new table with        0 representatives\n",
      "# comparing sequences from       1883  to       1919\n",
      "---------- new table with        1 representatives\n",
      "# comparing sequences from       1919  to       1953\n",
      "---------- new table with        0 representatives\n",
      "# comparing sequences from       1953  to       1986\n",
      "---------- new table with        1 representatives\n",
      "# comparing sequences from       1986  to       2966\n",
      "........---------- new table with      102 representatives\n",
      "\n",
      "     2966  finished        271  clusters\n",
      "\n",
      "Apprixmated maximum memory consumption: 340M\n",
      "writing new database\n",
      "writing clustering information\n",
      "program completed !\n",
      "\n",
      "Total CPU time 6.06\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#Cluster the tracrRNA sequence hits\n",
    "cd-hit-est -i sequences/PaperTracrBLAST_Hits.fa -o clusters/PaperTracrBLAST_Hits.fa -M 0 -d 0 -n 7 -c .9 -T 0 -sc 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4>Take consensus clusters to divide blast results by closest probable structure</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tTotal number of clusters: 171\n",
      "\tNumber of nodes covered: 2866\n",
      "2866\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYwAAAEACAYAAACgS0HpAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAE1RJREFUeJzt3X+QXeV93/H3ZyUwyDY/Y6TaMsg4LrbTJJg2BtfJRLaJbfAY4k5I7Mgt/hGPZ0xqt8ykCJKJ4I82VabUP8bDOEkpI1MZF/DEkMUpoMFLx51g8IAMsQCRxo7AlDWpQdSOhcXq2z/uWbSsfvBotXfPlfb9mjmjc5577j3f+2jvfvac555zUlVIkvRixvouQJJ0aDAwJElNDAxJUhMDQ5LUxMCQJDUxMCRJTYYaGEmuTjKZ5P4ZbX+c5MEkm5N8JckxMx67NMkj3ePvHGZtkqQDM+w9jGuAd81quw34uao6HXgEuBQgyRuB3wTeAJwDXJUkQ65PktRoqIFRVd8AnprVtqmqdnWLdwEru/nzgC9X1XNV9T0GYfLmYdYnSWrX9xjGR4CvdfOvAh6d8dj3uzZJ0gjoLTCS/D6ws6qum27ay2pet0SSRsTSPjaa5ELgXODtM5ofA149Y3kl8Pg+nm+QSNIcVNWcx4YXYg8jzNh7SPJu4N8B51XVszPWuxl4f5Ijk7wG+Fng7n296D333MOSJUewdOnRPUxHsXz5KVRV79O6det6r2FUJvvCvrAv9j8drKHuYST5ErAaODHJNmAdcBlwJHB79yWou6rqE1W1Jcn1wBZgJ/CJ2s87nJyc5GUv+zW2b79lmG9hH3by5JPLetiuJPVnqIFRVb+9l+Zr9rP+HwF/NLyKJElz1fe3pHSQVq9e3XcJI8O+2M2+2M2+mD+Zj+NaCy1JjY+Ps2bNVb0dkhobW8bU1M4eti1Jc5OEGvFBb0nSYcDAkCQ1MTAkSU0MDElSEwNDktTEwJAkNTEwJElNDAxJUhMDQ5LUxMCQJDUxMCRJTQwMSVITA0OS1MTAkCQ1MTAkSU0MDElSEwNDktTEwJAkNTEwJElNDAxJUhMDQ5LUxMCQJDUxMCRJTQwMSVITA0OS1MTAkCQ1GWpgJLk6yWSS+2e0HZ/ktiQPJ7k1ybEzHvtckkeSbE5y+jBrkyQdmGHvYVwDvGtW21pgU1WdBtwBXAqQ5BzgtVX1OuDjwBeGXJsk6QAMNTCq6hvAU7Oazwc2dPMbuuXp9i92z/smcGyS5cOsT5LUro8xjJOqahKgqp4ATuraXwU8OmO973dtkqQRMEqD3tlLWy14FZKkvVrawzYnkyyvqskkK4AfdO2PAa+esd5K4PF9vcjGjRvZsWMrcDmwupskSdMmJiaYmJiYt9dL1XD/iE+yCviLqvr5bnk98MOqWp9kLXBcVa1Nci5wUVW9J8lZwGeq6qx9vGaNj4+zZs1VbN9+y1Dr37udjI0tY2pqZw/blqS5SUJV7e1oTpOh7mEk+RKDP/1PTLINWAf8R+CGJB8BtgEXAFTV15Kcm+RvgB8DHx5mbZKkAzPUwKiq397HQ2fvY/3fHWI5kqSDMEqD3pKkEWZgSJKaGBiSpCYGhiSpiYEhSWpiYEiSmhgYkqQmBoYkqYmBIUlqYmBIkpoYGJKkJgaGJKmJgSFJamJgSJKaGBiSpCYGhiSpiYEhSWpiYEiSmhgYkqQmBoYkqYmBIUlqYmBIkpoYGJKkJgaGJKmJgSFJamJgSJKaGBiSpCYGhiSpiYEhSWrSW2Ak+bdJ/jrJ/Uk2JjkyyaokdyV5OMl1SZb2VZ8k6YV6CYwkrwT+NXBGVf0CsBT4ALAeuLKqTgOeBj7aR32SpD31eUhqCfDSbi/iaOBx4G3AV7rHNwDv66k2SdIsvQRGVT0OXAlsA74PbAfuBZ6uql3dao8Br+yjPknSnnoZI0hyHHA+cAqDsLgBOGcvq9a+XmPjxo3s2LEVuBxY3U2SpGkTExNMTEzM2+ulap+/k4cmyW8A76qqj3XL/xJ4C/AbwIqq2pXkLGBdVe0RJElqfHycNWuuYvv2Wxa09oGdjI0tY2pqZw/blqS5SUJVZa7P72sMYxtwVpKjkgR4B/Ad4OvABd06FwI39VSfJGmWvsYw7gZuBO4Dvg0E+FNgLXBxkq3ACcDVfdQnSdpTb+c5VNUVwBWzmr8LnNlDOZKkF+GZ3pKkJgaGJKmJgSFJamJgSJKaGBiSpCYGhiSpiYEhSWpiYEiSmjQFRpJ/MuxCJEmjrXUP4wtJ7k7yie5Ks5KkRaYpMKrql4E1wKuBbyX5UpJfG2plkqSR0jyGUVWPAH8AXAL8KvC5JA8l+RfDKk6SNDpaxzB+IcmngQeBtwPvrao3dPOfHmJ9kqQR0Xq12s8DfwZcVlU/mW6sqseT/MFQKpMkjZTWwDgX+ElVTQEkGQOOqqp/qKprh1adJGlktI5hbAKOnrG8rGuTJC0SrYFxVFX9aHqhm182nJIkSaOoNTB+nOSM6YUk/xT4yX7WlyQdZlrHMP4NcEOSx7vlfwT81nBKkiSNoqbAqKp7krweOA0I8FBV7RxqZZKkkdK6hwHwS8Cq7jlvSkJVfXEoVUmSRk5TYCS5FngtsBmY6poLMDAkaZFo3cP4Z8Abq6qGWYwkaXS1fkvqr4EVwyxEkjTaWvcwfgbYkuRu4Nnpxqo6byhVSZJGTmtgXD7MIiRJo6/1a7V3JjkFeF1VbUqyDFgy3NIkSaOk9fLmHwNuBP6ka3oV8NVhFSVJGj2tg94XAW8FnoHnb6Z00sFsOMmxSW5I8mCS7yQ5M8nxSW5L8nCSW5McezDbkCTNn9bAeLaqfjq9kGQpg/MwDsZnga91N2L6ReAhYC2wqapOA+4ALj3IbUiS5klrYNyZ5DLg6O5e3jcAfzHXjSZ5OfArVXUNQFU9V1XbgfOBDd1qG4Bfn+s2JEnzqzUw1gJPAg8AHwe+xuD+3nN1KvD3Sa5Jcm+SP+0G0pdX1SRAVT0BvOIgtiFJmket35LaxeAWrX82j9s9A7ioqr7V3S98LQdwmGvjxo3s2LGVwTd+V3eTJGnaxMQEExMT8/Z6abnaR5Lvspdf5lV16pw2miwH/mr6+Ul+mUFgvBZYXVWTSVYAX+/GOGY/v8bHx1mz5iq2b79lLiUcpJ2MjS1jasoL9ko6dHQXjc1cn38g15KadhRwAXDCXDfaBcKjSf5xVW0F3gF8p5s+BKwHLgRumus2JEnzq/WQ1P+d1fSZJN8A/vAgtv1JYGOSI4C/BT7M4GTA65N8BNjGIJgkSSOg9fLmZ8xYHGOwx/Hyg9lwVX2bwT02Zjv7YF5XkjQcrYekrpwx/xzwPeA3570aSdLIaj0k9bZhFyJJGm2th6Qu3t/jVfWf56ccSdKoOpBvSf0ScHO3/F7gfwKPDqMoSdLoOZAbKJ1RVf8PIMnlwA1V9TvDKkySNFpaLw1yMvDTGcs/BVbNezWSpJHVuodxLXB3kj9ncMb3+4AvDq0qSdLIaf2W1L9P8pfAr3RNH66q+4ZXliRp1LQekgJYBjxTVZ8FHkvymiHVJEkaQa23aF0HXMLuGxodAfy3YRUlSRo9rXsY7wPOA34MUFWPc5CXBpEkHVpaA+OnNbgOegEkeenwSpIkjaLWwLg+yZ8AxyX5GLCJ+buZkiTpEND6Lan/1N3L+xngNOAPq+r2oVYmSRopLxoYSZYAt1bV2YAhIUmL1IsekqqqKeAfkhy7APVIkkZU65neO4AHktxO900pgKr65FCqkiSNnNbAuKWbJEmL1H4DI8nJVbWtqjYsVEGSpNH0YmMYX52eSfKVIdciSRphLxYYmTF/6jALkSSNthcLjNrHvCRpkXmxQe9fTPIMgz2No7t5uuWqqmOGWp0kaWTsNzCqaslCFSJJGm0Hcj8MSdIiZmBIkpoYGJKkJgaGJKlJr4GRZCzJvUlu7pZXJbkrycNJrkvSeukSSdKQ9b2H8Slgy4zl9cCVVXUa8DTw0V6qkiTtobfASLISOBf4LzOa3w5MX4JkA4N7iUuSRkCfexifBn6P3fcJPxF4qqp2dY8/Bryyp9okSbP0MkaQ5D3AZFVtTrJ6upkXXrsK9nM5ko0bN7Jjx1bgcmB1N0mSpk1MTDAxMTFvr5eqhb9EVJL/AHwQeA44Gng5gyvjvhNYUVW7kpwFrKuqc/by/BofH2fNmqvYvr2P23TsZGxsGVNTO3vYtiTNTRKqavYf5s16OSRVVZdV1clVdSrwfuCOqvog8HXggm61C4Gb+qhPkrSnvr8lNdta4OIkW4ETgKt7rkeS1On9PIequhO4s5v/LnBmvxVJkvZm1PYwJEkjysCQJDUxMCRJTQwMSVITA0OS1MTAkCQ1MTAkSU0MDElSEwNDktTEwJAkNTEwJElNDAxJUhMDQ5LUxMCQJDUxMCRJTQwMSVITA0OS1MTAkCQ1MTAkSU0MDElSEwNDktTEwJAkNTEwJElNDAxJUhMDQ5LUxMCQJDUxMCRJTQwMSVKTXgIjycokdyTZkuSBJJ/s2o9PcluSh5PcmuTYPuqTJO2prz2M54CLq+qNwFuAi5K8HlgLbKqq04A7gEt7qk+SNEsvgVFVT1TV5m7+R8CDwErgfGBDt9oG4Nf7qE+StKfexzCSrAJOB+4CllfVJAxCBXhFf5VJkmbqNTCSvAy4EfhUt6dRfdYjSdq3pX1tOMlSBmFxbVXd1DVPJlleVZNJVgA/2NfzN27cyI4dW4HLgdXdJEmaNjExwcTExLy9Xqr6+aM+yReBv6+qi2e0rQd+WFXrk1wCHF9Va/fy3BofH2fNmqvYvv2WBax62k7GxpYxNbWzh21L0twkoaoy1+f3soeR5K3AGuCBJPcxOBR1GbAeuD7JR4BtwAV91CdJ2lMvgVFV/wtYso+Hz17IWiRJbXr/lpQk6dBgYEiSmhgYkqQmBoYkqYmBIUlqYmBIkpoYGJKkJgaGJKmJgSFJamJgSJKaGBiSpCYGhiSpiYEhSWpiYEiSmhgYkqQmBoYkqYmBIUlqYmBIkpr0covWw8GuXUtJ5nwv9YO2fPkpPPHE93rbvqTFx8CYsx1A9bb1ycn+wkrS4uQhKUlSEwNDktTEwJAkNTEwJElNDAxJUhMDQ5LUxMCQJDUxMCRJTUYyMJK8O8lDSbYmuaTveiRJIxgYScaAzwPvAn4O+ECS1/db1eiamJjou4SRYV/sZl/sZl/Mn5ELDODNwCNV9XdVtRP4MnB+zzWNLD8Mu9kXu9kXu9kX82cUA+NVwKMzlh/r2vQCLyEJV1xxBUkWdFqxYlXfb743K1asWvD+tt/75f/5bqN48cG9XVVvj6v8HXHEEezYcQ/HHPPeBShptl0880wPm32BZxl0y+XdtHAW84UPJyf/jr4uOrmY+71P/p/vlqr+rri6N0nOAi6vqnd3y2uBqqr1M9YZraIl6RBRVXNOoVEMjCXAw8A7gP8D3A18oKoe7LUwSVrkRu6QVFVNJfld4DYGYyxXGxaS1L+R28OQJI2mUfyW1H4ttpP6klydZDLJ/TPajk9yW5KHk9ya5NgZj30uySNJNic5vZ+q51+SlUnuSLIlyQNJPtm1L8a+eEmSbya5r+uLdV37qiR3dX1xXZKlXfuRSb7c9cVfJTm533cw/5KMJbk3yc3d8qLsiyTfS/Lt7mfj7q5t3j4jh1RgZHGe1HcNg/c701pgU1WdBtwBXAqQ5BzgtVX1OuDjwBcWstAhew64uKreCLwFuKj7v190fVFVzwJvq6o3AacD5yQ5E1gPXNn1xdPAR7unfBT4YdcXnwH+uIeyh+1TwJYZy4u1L3YBq6vqTVX15q5t/j4jVXXITMBZwF/OWF4LXNJ3XQvwvk8B7p+x/BCwvJtfATzYzX8B+K0Z6z04vd7hNgFfBc5e7H0BLAO+xeCE1x8AY137858V4H8AZ3bzS4An+657nvtgJXA7sBq4uWt7cpH2xXeBE2e1zdtn5JDaw8CT+qadVFWTAFX1BHBS1z67f77PYdg/SVYx+Mv6LgY/4IuuL7pDMPcBTzD4Zfm/gaerale3yszPxvN9UVVTwNNJTljgkofp08Dv0Z0skeRE4KlF2hcF3JrkniS/07XN22dk5L4l9SKaTupbxA77/knyMuBG4FNV9aP9nJNzWPdF98vwTUmOAf4ceMPeVuv+nd0X4TDpiyTvASaranOS1dPN7PmeD/u+6PzzqnoiySuA25I8zL7f3wF/Rg61PYzHgJmDVCuBx3uqpU+TSZYDJFnB4FAEDPrn1TPWO6z6pxu4vBG4tqpu6poXZV9Mq6pngDsZHHY5rhvngxe+3+f7ojvP6Ziqemqhax2StwLnJflb4Drg7QzGJo5dhH0xvQdBVT3J4LDtm5nHz8ihFhj3AD+b5JQkRwLvB27uuaaFMPsvppuBD3XzHwJumtH+r+D5M+afnt4VPUz8V2BLVX12Rtui64skPzP9TZckRzMYy9kCfB24oFvtQl7YFxd28xcwGPg8LFTVZVV1clWdyuD3wR1V9UEWYV8kWdbtgZPkpcA7gQeYz89I34M0cxjUeTeDM8EfAdb2Xc8CvN8vMUj9Z4FtwIeB44FNXT/cDhw3Y/3PA38DfBs4o+/657Ef3gpMAZuB+4B7u5+FExZhX/x89/43A/cDv9+1vwb4JrAV+O/AEV37S4Dru8/MXcCqvt/DkPrlV9k96L3o+qJ7z9Ofjwemfz/O52fEE/ckSU0OtUNSkqSeGBiSpCYGhiSpiYEhSWpiYEiSmhgYkqQmBoYkqYmBIUlq8v8BDOOfFmcgFWsAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x2b6eb6fc56d0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "allClusters = processClusterFile(\"clusters/PaperTracrBLAST_Hits.fa.clstr\")\n",
    "TreeColors, clusterDist = generateTreeColors(allClusters,cutoff=2)\n",
    "clusterDist = Series(clusterDist)\n",
    "clusterDist[clusterDist>2].plot.hist();"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4>Build Consensus Structure Fasta files</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "conseqs0/Cluster_79.fasta ( 5 / 6 )[86,88] 3.32665998663\n",
      "conseqs0/Cluster_78.fasta ( 5 / 6 )[104,105] 1.47196014439\n",
      "conseqs0/Cluster_71.fasta ( 5 / 6 )[61,93] 25.1853131805\n",
      "conseqs0/Cluster_70.fasta ( 4 / 7 )[86,89] 22.5007936368\n",
      "conseqs0/Cluster_73.fasta ( 5 / 6 )[66,79] 7.74596669241\n",
      "conseqs0/Cluster_72.fasta ( 5 / 6 )[105,105] 8.9814623902\n",
      "conseqs0/Cluster_75.fasta ( 4 / 6 )[68,68] 6.19677335393\n",
      "conseqs0/Cluster_74.fasta ( 5 / 6 )[70,73] 11.9777571635\n",
      "conseqs0/Cluster_77.fasta ( 5 / 6 )[79,97] 19.1276414298\n",
      "conseqs0/Cluster_76.fasta ( 4 / 6 )[92,93] 2.87518115371\n",
      "conseqs0/Cluster_162.fasta ( 2 / 2 )[94,97] 2.12132034356\n",
      "conseqs0/Cluster_163.fasta ( 2 / 2 )[70,72] 1.41421356237\n",
      "conseqs0/Cluster_170.fasta ( 2 / 2 )[94,95] 0.707106781187\n",
      "conseqs0/Cluster_48.fasta ( 9 / 12 )[59,69] 6.65149789464\n",
      "conseqs0/Cluster_49.fasta ( 9 / 12 )[61,87] 29.013580937\n",
      "conseqs0/Cluster_44.fasta ( 8 / 13 )[73,99] 16.0032048072\n",
      "conseqs0/Cluster_45.fasta ( 8 / 12 )[89,96] 5.84846522214\n",
      "conseqs0/Cluster_46.fasta ( 11 / 12 )[90,99] 5.59220236336\n",
      "conseqs0/Cluster_47.fasta ( 8 / 12 )[73,82] 7.01243484278\n",
      "conseqs0/Cluster_40.fasta ( 11 / 15 )[55,72] 9.51540105003\n",
      "conseqs0/Cluster_41.fasta ( 11 / 14 )[89,95] 3.03821810125\n",
      "conseqs0/Cluster_42.fasta ( 9 / 13 )[89,90] 1.06818801764\n",
      "conseqs0/Cluster_43.fasta ( 12 / 13 )[76,84] 6.34883393291\n",
      "conseqs0/Cluster_139.fasta ( 2 / 2 )[79,85] 4.24264068712\n",
      "conseqs0/Cluster_138.fasta ( 2 / 2 )[85,166] 57.2756492761\n",
      "conseqs0/Cluster_111.fasta ( 2 / 3 )[103,105] 16.1967074843\n",
      "conseqs0/Cluster_131.fasta ( 2 / 2 )[61,79] 12.7279220614\n",
      "conseqs0/Cluster_130.fasta ( 2 / 2 )[74,74] 0.0\n",
      "conseqs0/Cluster_133.fasta ( 2 / 2 )[60,80] 14.1421356237\n",
      "conseqs0/Cluster_132.fasta ( 2 / 2 )[66,93] 19.091883092\n",
      "conseqs0/Cluster_135.fasta ( 2 / 2 )[82,87] 3.53553390593\n",
      "conseqs0/Cluster_134.fasta ( 2 / 2 )[87,87] 0.0\n",
      "conseqs0/Cluster_137.fasta ( 2 / 2 )[72,77] 3.53553390593\n",
      "conseqs0/Cluster_136.fasta ( 2 / 2 )[60,167] 75.660425587\n",
      "conseqs0/Cluster_159.fasta ( 2 / 2 )[66,66] 0.0\n",
      "conseqs0/Cluster_158.fasta ( 2 / 2 )[66,66] 0.0\n",
      "conseqs0/Cluster_153.fasta ( 2 / 2 )[80,104] 16.9705627485\n",
      "conseqs0/Cluster_152.fasta ( 2 / 2 )[63,63] 0.0\n",
      "conseqs0/Cluster_151.fasta ( 2 / 2 )[131,139] 5.65685424949\n",
      "conseqs0/Cluster_150.fasta ( 2 / 2 )[62,62] 0.0\n",
      "conseqs0/Cluster_157.fasta ( 2 / 2 )[99,105] 4.24264068712\n",
      "conseqs0/Cluster_156.fasta ( 2 / 2 )[105,108] 2.12132034356\n",
      "conseqs0/Cluster_155.fasta ( 2 / 2 )[67,67] 0.0\n",
      "conseqs0/Cluster_154.fasta ( 2 / 2 )[95,137] 29.6984848098\n",
      "conseqs0/Cluster_53.fasta ( 7 / 10 )[74,107] 17.0411267233\n",
      "conseqs0/Cluster_52.fasta ( 7 / 11 )[92,94] 5.19265215123\n",
      "conseqs0/Cluster_51.fasta ( 8 / 11 )[88,91] 3.38445644891\n",
      "conseqs0/Cluster_50.fasta ( 8 / 11 )[69,73] 2.82842712475\n",
      "conseqs0/Cluster_57.fasta ( 8 / 10 )[61,61] 6.32455532034\n",
      "conseqs0/Cluster_56.fasta ( 5 / 10 )[64,64] 10.0576118217\n",
      "conseqs0/Cluster_55.fasta ( 8 / 10 )[109,117] 5.45791983167\n",
      "conseqs0/Cluster_54.fasta ( 7 / 10 )[66,86] 23.6570402958\n",
      "conseqs0/Cluster_166.fasta ( 2 / 2 )[93,93] 0.0\n",
      "conseqs0/Cluster_59.fasta ( 6 / 9 )[88,95] 5.38000413052\n",
      "conseqs0/Cluster_58.fasta ( 8 / 10 )[96,137] 21.5468481841\n",
      "conseqs0/Cluster_128.fasta ( 2 / 3 )[82,84] 11.5902257671\n",
      "conseqs0/Cluster_129.fasta ( 2 / 3 )[105,105] 0.57735026919\n",
      "conseqs0/Cluster_126.fasta ( 2 / 3 )[95,110] 24.00694344\n",
      "conseqs0/Cluster_127.fasta ( 2 / 3 )[84,89] 12.0968315411\n",
      "conseqs0/Cluster_124.fasta ( 2 / 3 )[88,89] 11.2694276696\n",
      "conseqs0/Cluster_125.fasta ( 2 / 3 )[118,125] 14.0118997047\n",
      "conseqs0/Cluster_122.fasta ( 2 / 3 )[81,81] 2.30940107676\n",
      "conseqs0/Cluster_123.fasta ( 2 / 3 )[102,102] 1.73205080757\n",
      "conseqs0/Cluster_120.fasta ( 2 / 3 )[91,91] 1.73205080757\n",
      "conseqs0/Cluster_121.fasta ( 2 / 3 )[75,79] 19.1572440607\n",
      "conseqs0/Cluster_148.fasta ( 2 / 2 )[128,128] 0.0\n",
      "conseqs0/Cluster_149.fasta ( 2 / 2 )[64,64] 0.0\n",
      "conseqs0/Cluster_167.fasta ( 2 / 2 )[105,120] 10.6066017178\n",
      "conseqs0/Cluster_140.fasta ( 2 / 2 )[52,81] 20.5060966544\n",
      "conseqs0/Cluster_141.fasta ( 2 / 2 )[81,83] 1.41421356237\n",
      "conseqs0/Cluster_142.fasta ( 2 / 2 )[141,143] 1.41421356237\n",
      "conseqs0/Cluster_143.fasta ( 2 / 2 )[77,77] 0.0\n",
      "conseqs0/Cluster_144.fasta ( 2 / 2 )[83,84] 0.707106781187\n",
      "conseqs0/Cluster_145.fasta ( 2 / 2 )[71,77] 4.24264068712\n",
      "conseqs0/Cluster_146.fasta ( 2 / 2 )[93,93] 0.0\n",
      "conseqs0/Cluster_147.fasta ( 2 / 2 )[91,102] 7.77817459305\n",
      "conseqs0/Cluster_26.fasta ( 20 / 24 )[83,92] 6.29340003956\n",
      "conseqs0/Cluster_27.fasta ( 15 / 24 )[80,95] 11.3712069644\n",
      "conseqs0/Cluster_24.fasta ( 25 / 27 )[84,94] 8.41972314816\n",
      "conseqs0/Cluster_25.fasta ( 18 / 27 )[68,91] 30.102436698\n",
      "conseqs0/Cluster_22.fasta ( 18 / 31 )[91,93] 15.2323933264\n",
      "conseqs0/Cluster_23.fasta ( 25 / 28 )[88,100] 6.70337118062\n",
      "conseqs0/Cluster_20.fasta ( 17 / 35 )[68,96] 22.1413715956\n",
      "conseqs0/Cluster_21.fasta ( 26 / 34 )[85,92] 9.02566157381\n",
      "conseqs0/Cluster_28.fasta ( 14 / 22 )[81,104] 13.159774074\n",
      "conseqs0/Cluster_29.fasta ( 11 / 22 )[93,105] 21.9191746298\n",
      "conseqs0/Cluster_117.fasta ( 2 / 3 )[119,126] 15.1327459504\n",
      "conseqs0/Cluster_116.fasta ( 2 / 3 )[93,101] 26.2868788562\n",
      "conseqs0/Cluster_115.fasta ( 2 / 3 )[76,82] 14.7422295917\n",
      "conseqs0/Cluster_114.fasta ( 2 / 3 )[107,110] 5.68624070308\n",
      "conseqs0/Cluster_113.fasta ( 2 / 3 )[120,127] 11.7898261226\n",
      "conseqs0/Cluster_112.fasta ( 2 / 3 )[64,70] 8.62167810425\n",
      "conseqs0/Cluster_35.fasta ( 15 / 18 )[80,95] 8.48451107045\n",
      "conseqs0/Cluster_34.fasta ( 10 / 18 )[63,64] 6.7174536103\n",
      "conseqs0/Cluster_37.fasta ( 7 / 17 )[93,110] 12.0650808729\n",
      "conseqs0/Cluster_36.fasta ( 16 / 18 )[78,79] 5.22437724969\n",
      "conseqs0/Cluster_31.fasta ( 11 / 19 )[94,115] 15.2388392675\n",
      "conseqs0/Cluster_30.fasta ( 18 / 22 )[63,83] 10.4288975813\n",
      "conseqs0/Cluster_33.fasta ( 8 / 18 )[82,85] 3.95935229403\n",
      "conseqs0/Cluster_32.fasta ( 10 / 18 )[97,104] 6.13811620843\n",
      "conseqs0/Cluster_110.fasta ( 4 / 4 )[77,105] 15.8823801743\n",
      "conseqs0/Cluster_39.fasta ( 12 / 17 )[81,86] 4.93412487026\n",
      "conseqs0/Cluster_38.fasta ( 10 / 17 )[80,106] 14.6490814809\n",
      "conseqs0/Cluster_80.fasta ( 5 / 6 )[95,95] 2.44948974278\n",
      "conseqs0/Cluster_81.fasta ( 4 / 6 )[84,93] 14.9855485941\n",
      "conseqs0/Cluster_82.fasta ( 6 / 6 )[96,103] 3.50713558335\n",
      "conseqs0/Cluster_83.fasta ( 5 / 6 )[113,140] 18.0739222823\n",
      "conseqs0/Cluster_84.fasta ( 4 / 5 )[64,70] 4.43846820423\n",
      "conseqs0/Cluster_85.fasta ( 3 / 5 )[86,88] 12.0706255016\n",
      "conseqs0/Cluster_86.fasta ( 4 / 5 )[76,82] 11.1445053726\n",
      "conseqs0/Cluster_87.fasta ( 4 / 5 )[84,88] 9.39680796867\n",
      "conseqs0/Cluster_88.fasta ( 4 / 5 )[82,91] 6.64830805544\n",
      "conseqs0/Cluster_89.fasta ( 4 / 5 )[97,97] 0.4472135955\n",
      "conseqs0/Cluster_168.fasta ( 2 / 2 )[97,99] 1.41421356237\n",
      "conseqs0/Cluster_169.fasta ( 2 / 2 )[59,59] 0.0\n",
      "conseqs0/Cluster_119.fasta ( 2 / 3 )[82,87] 5.50757054729\n",
      "conseqs0/Cluster_118.fasta ( 2 / 3 )[65,67] 8.14452781525\n",
      "conseqs0/Cluster_97.fasta ( 3 / 5 )[79,119] 27.3367883995\n",
      "conseqs0/Cluster_96.fasta ( 4 / 5 )[75,86] 8.92748564827\n",
      "conseqs0/Cluster_95.fasta ( 4 / 5 )[109,111] 1.788854382\n",
      "conseqs0/Cluster_94.fasta ( 3 / 5 )[76,86] 12.549900398\n",
      "conseqs0/Cluster_93.fasta ( 2 / 5 )[90,90] 7.52994023881\n",
      "conseqs0/Cluster_92.fasta ( 4 / 5 )[64,69] 2.79284800875\n",
      "conseqs0/Cluster_91.fasta ( 4 / 5 )[87,88] 0.836660026534\n",
      "conseqs0/Cluster_90.fasta ( 3 / 5 )[88,90] 1.92353840617\n",
      "conseqs0/Cluster_160.fasta ( 2 / 2 )[94,124] 21.2132034356\n",
      "conseqs0/Cluster_99.fasta ( 4 / 4 )[115,115] 0.0\n",
      "conseqs0/Cluster_98.fasta ( 4 / 5 )[89,89] 8.94427191\n",
      "conseqs0/Cluster_161.fasta ( 2 / 2 )[57,57] 0.0\n",
      "conseqs0/Cluster_19.fasta ( 19 / 35 )[80,103] 20.7446662174\n",
      "conseqs0/Cluster_18.fasta ( 31 / 39 )[71,89] 10.774275857\n",
      "conseqs0/Cluster_17.fasta ( 22 / 40 )[60,77] 9.19029588542\n",
      "conseqs0/Cluster_16.fasta ( 32 / 41 )[105,107] 10.6091883784\n",
      "conseqs0/Cluster_15.fasta ( 39 / 43 )[101,113] 6.48748689072\n",
      "conseqs0/Cluster_14.fasta ( 27 / 43 )[83,121] 21.5975471311\n",
      "conseqs0/Cluster_13.fasta ( 19 / 44 )[59,74] 10.7624749388\n",
      "conseqs0/Cluster_12.fasta ( 39 / 49 )[85,90] 2.90114919759\n",
      "conseqs0/Cluster_11.fasta ( 27 / 49 )[82,106] 14.6077395384\n",
      "conseqs0/Cluster_10.fasta ( 39 / 54 )[120,123] 8.6888926421\n",
      "conseqs0/Cluster_164.fasta ( 2 / 2 )[110,139] 20.5060966544\n",
      "conseqs0/Cluster_165.fasta ( 2 / 2 )[119,120] 0.707106781187\n",
      "conseqs0/Cluster_104.fasta ( 4 / 4 )[98,98] 0.0\n",
      "conseqs0/Cluster_105.fasta ( 4 / 4 )[68,80] 6.92820323028\n",
      "conseqs0/Cluster_106.fasta ( 3 / 4 )[65,65] 0.5\n",
      "conseqs0/Cluster_107.fasta ( 3 / 4 )[86,88] 2.51661147842\n",
      "conseqs0/Cluster_100.fasta ( 4 / 4 )[76,108] 18.4752086141\n",
      "conseqs0/Cluster_101.fasta ( 2 / 4 )[71,74] 7.45542308212\n",
      "conseqs0/Cluster_102.fasta ( 2 / 4 )[84,92] 29.2389580298\n",
      "conseqs0/Cluster_103.fasta ( 2 / 4 )[76,82] 5.85234995536\n",
      "conseqs0/Cluster_108.fasta ( 3 / 4 )[69,70] 2.21735578261\n",
      "conseqs0/Cluster_109.fasta ( 3 / 4 )[59,62] 9.74679434481\n",
      "conseqs0/Cluster_68.fasta ( 3 / 7 )[82,82] 21.3374252028\n",
      "conseqs0/Cluster_69.fasta ( 4 / 7 )[59,61] 11.2546286774\n",
      "conseqs0/Cluster_62.fasta ( 7 / 8 )[85,95] 9.05439908868\n",
      "conseqs0/Cluster_63.fasta ( 8 / 8 )[70,70] 0.0\n",
      "conseqs0/Cluster_60.fasta ( 8 / 9 )[80,81] 0.726483157257\n",
      "conseqs0/Cluster_61.fasta ( 7 / 8 )[72,89] 9.22632104362\n",
      "conseqs0/Cluster_66.fasta ( 6 / 7 )[65,68] 2.44948974278\n",
      "conseqs0/Cluster_67.fasta ( 6 / 7 )[68,73] 12.8507920823\n",
      "conseqs0/Cluster_64.fasta ( 7 / 8 )[83,90] 4.79396942597\n",
      "conseqs0/Cluster_65.fasta ( 4 / 7 )[88,94] 7.50238057457\n",
      "conseqs0/Cluster_9.fasta ( 50 / 56 )[118,121] 5.63983973246\n",
      "conseqs0/Cluster_8.fasta ( 35 / 59 )[97,110] 7.30463452638\n",
      "conseqs0/Cluster_7.fasta ( 43 / 60 )[77,91] 7.37630578309\n",
      "conseqs0/Cluster_6.fasta ( 54 / 78 )[87,103] 11.5729598605\n",
      "conseqs0/Cluster_5.fasta ( 56 / 82 )[99,104] 3.15226363655\n",
      "conseqs0/Cluster_4.fasta ( 63 / 82 )[80,94] 11.6870321668\n",
      "conseqs0/Cluster_3.fasta ( 81 / 96 )[83,107] 14.4584098031\n",
      "conseqs0/Cluster_2.fasta ( 64 / 98 )[91,101] 5.64554567671\n",
      "conseqs0/Cluster_1.fasta ( 152 / 239 )[93,101] 5.33571126506\n",
      "conseqs0/Cluster_0.fasta ( 385 / 470 )[85,96] 5.87766373643\n"
     ]
    }
   ],
   "source": [
    "tracrSeqDict = fasta_index(\"sequences/PaperTracrBLAST_Hits.fa\",\"fasta\")\n",
    "\n",
    "for cluster, seqIDList in allClusters.iteritems():\n",
    "    clusterFileName = \"conseqs0/\"+cluster.replace(\" \",\"_\")+\".fasta\"\n",
    "    fh = open(clusterFileName,\"wb\")\n",
    "    clusterStats = []\n",
    "    for seqID in seqIDList: clusterStats.append(len(tracrSeqDict[seqID]))\n",
    "    clusterStats = Series(clusterStats)\n",
    "    keepSeqs = clusterStats[~((clusterStats-clusterStats.mean()).abs() > clusterStats.std())]\n",
    "    minSeq, maxSeq = keepSeqs.min(), keepSeqs.max()\n",
    "    index=0\n",
    "    for seqID in seqIDList: \n",
    "        seqLen = len(tracrSeqDict[seqID])\n",
    "        if seqLen >= minSeq and seqLen <= maxSeq:\n",
    "            index+=1\n",
    "            fh.write(\">%s\\n%s\\n\" % (seqID,str(tracrSeqDict[seqID].seq)))\n",
    "    fh.close()\n",
    "    if index <= 1:\n",
    "        os.system(\"rm %s\" % (clusterFileName))\n",
    "        continue\n",
    "    print clusterFileName,\"(\",index,\"/\",len(seqIDList), \")[%i,%i]\" % (minSeq,maxSeq),clusterStats.std() \n",
    "    os.system(\"cd-hit-est -i %s -o %s_cluster -M 0 -d 0 -c .90 -T 0 -sc 1\" % (clusterFileName,clusterFileName))\n",
    "    clusteredResults = processClusterFile(\"%s_cluster.clstr\" % (clusterFileName))\n",
    "    fh = open(clusterFileName,\"wb\")\n",
    "    index = 0\n",
    "    for seqID,direction in clusteredResults.filter():\n",
    "        index +=1\n",
    "        if direction: fh.write(\">%s\\n%s\\n\" % (seqID,str(tracrSeqDict[seqID].seq).upper()))\n",
    "        else: fh.write(\">%s\\n%s\\n\" % (seqID,str(tracrSeqDict[seqID].seq.reverse_complement()).upper()))\n",
    "    fh.close()\n",
    "    if index <= 1: os.system(\"rm %s*\" % (clusterFileName))\n",
    "    else: os.system(\"rm %s_cluster*\" % (clusterFileName))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perform Infernal Structural Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Submitted batch job 6767393\n",
      "Submitted batch job 6767394\n",
      "Submitted batch job 6767395\n",
      "Submitted batch job 6767396\n",
      "Submitted batch job 6767401\n",
      "Submitted batch job 6767402\n",
      "Submitted batch job 6767404\n",
      "Submitted batch job 6767405\n",
      "Submitted batch job 6767406\n",
      "Submitted batch job 6767407\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for i in 1 2 3 4 5 6 7 8 9 10\n",
    "do\n",
    "    sbatch ../scripts/hpc/StructureSearch.sb\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4>Process results</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty Output File:  Cluster_58.out 0\n",
      "Empty Output File:  Cluster_89.out 0\n",
      "Empty Output File:  Cluster_55.out 0\n",
      "Empty Output File:  Cluster_75.out 0\n",
      "Empty Output File:  Cluster_125.out 0\n",
      "Empty Output File:  Cluster_155.out 0\n",
      "Empty Output File:  Cluster_20.out 0\n",
      "Empty Output File:  Cluster_142.out 0\n",
      "Dumping /mnt/research/germs/shane/transActRNA/data/pickles/InfernalResults00.p Number of results: 977\n"
     ]
    }
   ],
   "source": [
    "infernalResults = ProcessInfernal(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4>Build the dictionary to annotate sequences with a BLAST hit, a structural hit, both, or none</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Dumping pickles/PaperAllBlastTreeColors.p\n",
      "\n",
      "# in both: 667\n",
      "# with no blast hit: 324\n",
      "# with no struct hit: 30\n",
      "# with no hit: 437\n",
      "\n",
      "Dumping pickles/PaperCombinedTreeColors.p\n"
     ]
    }
   ],
   "source": [
    "both = set(infernalResults.seqTracrs.keys()).intersection(tracrBLAST.keys())\n",
    "onlyStruct = set(infernalResults.seqTracrs.keys()).difference(tracrBLAST.keys())\n",
    "onlyBlast = set(tracrBLAST.keys()).difference(infernalResults.seqTracrs.keys())\n",
    "\n",
    "treeColors = {}\n",
    "for key in tracrBLAST.keys(): treeColors[key] = \"#9B7E46\"\n",
    "dump(treeColors,\"pickles/PaperAllBlastTreeColors.p\")\n",
    "    \n",
    "print \"\\n# in both:\", len(both)\n",
    "print \"# with no blast hit:\", len(onlyStruct)\n",
    "print \"# with no struct hit:\", len(onlyBlast)\n",
    "treeColors = {}\n",
    "bothCounter, onlyS, onlyB,noHit =0,0,0,0\n",
    "for id in casRelatedAssemblies:\n",
    "    if   id in both:       treeColors[id] = \"#70857A\"; bothCounter +=1\n",
    "    elif id in onlyStruct: treeColors[id] = \"#F4B266\"; onlyS +=1\n",
    "    elif id in onlyBlast:  treeColors[id] = \"#9B7E46\"; onlyB +=1\n",
    "    else: noHit += 1\n",
    "\n",
    "print \"# with no hit:\",noHit\n",
    "dump(treeColors,\"pickles/PaperCombinedTreeColors.p\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1458"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "667+324+30+437"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4>Process and combine infernal results with blast to create consust strucutre clusters</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1458\n",
      "Number of Strucutre hits: 991\n",
      "Found in 46 clusters\n",
      "\n",
      "Dumping pickles/KnownTracrStuctureHitColors.p\n"
     ]
    }
   ],
   "source": [
    "print len(casRelatedAssemblies)\n",
    "\n",
    "clusterMap = {}\n",
    "for cluster, seqIDS in infernalResults.structMapping.iteritems():\n",
    "    clusterMap[cluster] = len(seqIDS)\n",
    "\n",
    "infernalMap,colorMap = {},{}\n",
    "for seqID in infernalResults:\n",
    "    clusterName = findBiggestCluster(seqID, infernalResults[seqID],clusterMap)\n",
    "    try: infernalMap[seqID]=colorMap[clusterName]\n",
    "    except:\n",
    "        colorMap[clusterName] = color()\n",
    "        infernalMap[seqID]=colorMap[clusterName]\n",
    "         \n",
    "print \"Number of Strucutre hits:\", len (infernalMap)\n",
    "print \"Found in %i clusters\"% len(colorMap)\n",
    "dump(infernalMap,\"pickles/KnownTracrStuctureHitColors.p\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write systems that don't have a structural or blast or neither to a new file for the next step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ready to do a tracrRNA search in 791 cas9-related genomic assemblies\n"
     ]
    }
   ],
   "source": [
    "fh = open(\"sequences/CasRelatedAssemblies_NoTracr.fa\" ,\"wb\")\n",
    "counter = 0\n",
    "for id, rec in casRelatedAssemblies.iteritems():\n",
    "    if id not in both:\n",
    "        write(rec,fh,\"fasta\")\n",
    "        counter+=1\n",
    "fh.close()\n",
    "print \"Ready to do a tracrRNA search in %i cas9-related genomic assemblies\" % (counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h4>Development Artifacts, not important</h4>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are a total of 0 duplication values\n",
      "The new file has a total of 220 tracrRNAs\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqIO import parse\n",
    "seq_rec = {}\n",
    "counter =0\n",
    "for rec in parse(\"KnownTracrRNAs.fa\",\"fasta\"):\n",
    "    if rec.id not in seq_rec:\n",
    "        seq_rec[rec.id]=str(rec.seq)\n",
    "    else:\n",
    "        #print rec.id, \n",
    "        if (seq_rec[rec.id]!=str(rec.seq)):\n",
    "            counter +=1\n",
    "            seq_rec[rec.id+\"_DUP%i\" % counter]=str(rec.seq)\n",
    "            print rec.id\n",
    "            print seq_rec[rec.id]\n",
    "            print str(rec.seq)\n",
    "            print\n",
    "            \n",
    "fh = open(\"KnownTracrRNAs.fasta\",\"w\")\n",
    "for id,seq in seq_rec.iteritems():\n",
    "    fh.write(\">%s\\n%s\\n\"%(id,seq))\n",
    "fh.close()\n",
    "print \"There are a total of %i duplication values\" % counter\n",
    "print \"The new file has a total of %i tracrRNAs\" %(len(seq_rec))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "casRelatedAssembliesFile = \"sequences/CasRelatedAssemblies.fa\"\n",
    "casRelatedAssemblies = fasta_index(casRelatedAssembliesFile,\"fasta\")\n",
    "cas9Proteins = fasta_index(\"proteins/Cas9-Like-clustered-2.faa\",\"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO import write\n",
    "fh = open(\"CasRelatedAssemblies.fa\",\"w\")\n",
    "for id in cas9Proteins: write(casRelatedAssemblies[id],fh,\"fasta\")\n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "fh = open(\"sequences/PioneerCrisprs.txt\")\n",
    "fh.readline()\n",
    "fasta = open(\"sequences/PioneerCrisprs.fasta\",\"w\")\n",
    "seqIds = set()\n",
    "for line in fh:\n",
    "    rec = line.strip().split('\\t')\n",
    "    rec[0] = rec[0].replace(\"sp.\",\"sp\")\n",
    "    id = rec[0].replace(\" \",\"_\")\n",
    "    if rec[0] == \"\":break\n",
    "    count=0\n",
    "    while id in seqIds:\n",
    "        count+=1\n",
    "        id = rec[0].replace(\" \",\"_\") + \"_%i\" % (count)\n",
    "    fasta.write(\">%s\\n%s\\n\" % (id,rec[1]))\n",
    "    seqIds.add(id)\n",
    "fasta.close()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
